Index: u-boot-txmp-2550/board/karo/stm32mp1/stm32mp1.c
===================================================================
--- u-boot-txmp-2550.orig/board/karo/stm32mp1/stm32mp1.c
+++ u-boot-txmp-2550/board/karo/stm32mp1/stm32mp1.c
@@ -674,20 +674,6 @@ static inline void rand_init(void)
 
 int board_late_init(void)
 {
-#if CONFIG_IS_ENABLED(ENV_VARS_UBOOT_RUNTIME_CONFIG)
-	const void *fdt_compat;
-	ofnode root = ofnode_path("/");
-
-	if (ofnode_valid(root))
-		fdt_compat = ofnode_read_string(root, "compatible");
-
-	if (fdt_compat) {
-		if (strncmp(fdt_compat, "karo,", 5) != 0)
-			env_set("board_name", fdt_compat);
-		else
-			env_set("board_name", fdt_compat + 5);
-	}
-#endif
 	karo_env_cleanup();
 
 	if (ctrlc()) {
