diff --git a/board/karo/txmp/txmp.c b/board/karo/txmp/txmp.c
index 92f1d1d1df5..a5f6d90c6a2 100644
--- a/board/karo/txmp/txmp.c
+++ b/board/karo/txmp/txmp.c
@@ -651,6 +651,7 @@ int ft_board_setup(void *blob, struct bd_info *bd)
 	karo_fixup_mtdparts(blob, nodes, ARRAY_SIZE(nodes));
 #endif
 
+#if 0
 	/* Update DT if coprocessor started */
 	node = ofnode_path("/m4");
 	if (ofnode_valid(node)) {
@@ -675,6 +676,7 @@ int ft_board_setup(void *blob, struct bd_info *bd)
 				    "early-booted");
 		}
 	}
+#endif
 
 	karo_fixup_lcd_panel(env_get("videomode"));
 
