diff --git a/board/karo/txmp/txmp.c b/board/karo/txmp/txmp.c
index a5f6d90c6a2..3779af01d89 100644
--- a/board/karo/txmp/txmp.c
+++ b/board/karo/txmp/txmp.c
@@ -34,6 +34,7 @@
 #include <asm/arch/stm32.h>
 #include <asm/arch/stm32mp1_smc.h>
 #include <asm/arch/sys_proto.h>
+#include <dm/ofnode.h>
 #include <jffs2/load_kernel.h>
 #include <linux/delay.h>
 #include <linux/if_ether.h>
@@ -389,13 +390,13 @@ static struct udevice *led_dev;
 
 static int txmp_get_led(struct udevice **dev, char *led_string)
 {
-	char *led_name;
+	const char *led_name;
 	int ret;
 
 	if (led_state == LED_STATE_DISABLED)
 		return -ENODEV;
 
-	led_name = fdtdec_get_config_string(gd->fdt_blob, led_string);
+	led_name = ofnode_conf_read_str(led_string);
 	if (!led_name) {
 		debug("%s: could not find %s config string\n",
 		      __func__, led_string);
@@ -451,11 +452,6 @@ static void txmp_setup_led(void)
 		pr_err("No boot-led defined\n");
 		led_state = LED_STATE_DISABLED;
 	}
-	ret = led_default_state();
-	if (ret) {
-		pr_err("Failed to initialize default LED state: %d\n", ret);
-		led_state = LED_STATE_DISABLED;
-	}
 }
 #else
 static inline void txmp_setup_led(void)
