diff --git a/include/configs/txmp.h b/include/configs/txmp.h
index 52cf120cb84..7c1f416708f 100644
--- a/include/configs/txmp.h
+++ b/include/configs/txmp.h
@@ -57,15 +57,6 @@
 #define CONFIG_SYS_NONCACHED_MEMORY	(1 * SZ_1M)	/* 1M */
 #endif
 
-#ifdef CONFIG_DM_VIDEO
-#define CONFIG_VIDEO_LOGO
-#define CONFIG_VIDEO_BMP_LOGO
-#define CONFIG_VIDEO_BMP_RLE8
-#define CONFIG_BMP_16BPP
-#define CONFIG_BMP_24BPP
-#define CONFIG_BMP_32BPP
-#endif
-
 #define CONFIG_SYS_MMCSD_FS_BOOT_PARTITION	1
 #define BOOT_PART_STR		__stringify(CONFIG_SYS_MMCSD_FS_BOOT_PARTITION)
 
