Index: u-boot/drivers/serial/serial_stm32.c
===================================================================
--- u-boot.orig/drivers/serial/serial_stm32.c
+++ u-boot/drivers/serial/serial_stm32.c
@@ -197,10 +197,10 @@ static int stm32_serial_probe(struct ude
 	}
 
 	plat->clock_rate = clk_get_rate(&clk);
-	if (!plat->clock_rate) {
+	if (!plat->clock_rate || IS_ERR_VALUE(plat->clock_rate)) {
 		clk_disable(&clk);
-		return -EINVAL;
-	};
+		return plat->clock_rate ?: -EINVAL;
+	}
 
 	_stm32_serial_init(plat->base, plat->uart_info);
 
