diff --git a/board/karo/common/fdt.c b/board/karo/common/fdt.c
index 39a8e2df19..50638ca377 100644
--- a/board/karo/common/fdt.c
+++ b/board/karo/common/fdt.c
@@ -19,12 +19,6 @@
 #include <linux/libfdt.h>
 #include "karo.h"
 
-#ifdef CONFIG_MAX_DTB_SIZE
-#define MAX_DTB_SIZE	CONFIG_MAX_DTB_SIZE
-#else
-#define MAX_DTB_SIZE	SZ_64K
-#endif
-
 DECLARE_GLOBAL_DATA_PTR;
 
 static void karo_set_fdtsize(void *fdt)
@@ -71,7 +65,7 @@ static void *karo_fdt_load_dtb(unsigned long fdtaddr)
 		if (fs_set_blk_dev(bootdev, bootpart, FS_TYPE_ANY))
 			return NULL;
 
-		ret = fs_read(dtbfile, fdtaddr, 0, MAX_DTB_SIZE, &fdtsize);
+		ret = fs_read(dtbfile, fdtaddr, 0, 0, &fdtsize);
 		if (ret) {
 			printf("Failed to load dtb from '%s': %d\n",
 			       dtbfile, ret);
