commit 1288f4ef61fbff9f7c0e315d809bd491106c3ec2
Author: Markus Bauer <mb@karo-electronics.de>
Date:   Wed Apr 14 09:44:48 2021 +0000

    Fit PLL2_Q for GPU to 533 MHz

diff --git a/arch/arm/dts/stm32mp15-qsmp-u-boot.dtsi b/arch/arm/dts/stm32mp15-qsmp-u-boot.dtsi
index 8f3fb8efbf..3ed5fe1f4e 100644
--- a/arch/arm/dts/stm32mp15-qsmp-u-boot.dtsi
+++ b/arch/arm/dts/stm32mp15-qsmp-u-boot.dtsi
@@ -80,7 +80,7 @@
 	st,pll@1 {
 		u-boot,dm-pre-reloc;
 		reg = <1>;
-		cfg = <2 65 1 1 0 5>;
+		cfg = <2 65 1 0 0 5>;
 		frac = <5120>;
 	};
 
diff --git a/arch/arm/dts/stm32mp15-txmp-u-boot.dtsi b/arch/arm/dts/stm32mp15-txmp-u-boot.dtsi
index b7b452eedc..fd4ab1f142 100644
--- a/arch/arm/dts/stm32mp15-txmp-u-boot.dtsi
+++ b/arch/arm/dts/stm32mp15-txmp-u-boot.dtsi
@@ -80,7 +80,7 @@
 	st,pll@1 {
 		u-boot,dm-pre-reloc;
 		reg = <1>;
-		cfg = <2 65 1 1 0 5>;
+		cfg = <2 65 1 0 0 5>;
 		frac = <5120>;
 	};
 
