Index: u-boot/cmd/gpt.c
===================================================================
--- u-boot.orig/cmd/gpt.c
+++ u-boot/cmd/gpt.c
@@ -850,18 +850,18 @@ U_BOOT_CMD(gpt, CONFIG_SYS_MAXARGS, 1, d
 	"GUID Partition Table",
 	"<command> <interface> <dev> <partitions_list>\n"
 	" - GUID partition table restoration and validity check\n"
-	" Restore or verify GPT information on a device connected\n"
-	" to interface\n"
+	"   Restore or verify GPT information on a device connected\n"
+	"   to interface\n"
 	" Example usage:\n"
-	" gpt write mmc 0 $partitions\n"
-	" gpt verify mmc 0 $partitions\n"
+	"   gpt write mmc 0 $partitions\n"
+	"   gpt verify mmc 0 $partitions\n"
 	" gpt guid <interface> <dev>\n"
 	"    - print disk GUID\n"
 	" gpt guid <interface> <dev> <varname>\n"
 	"    - set environment variable to disk GUID\n"
 	" Example usage:\n"
-	" gpt guid mmc 0\n"
-	" gpt guid mmc 0 varname\n"
+	"   gpt guid mmc 0\n"
+	"   gpt guid mmc 0 varname\n"
 #ifdef CONFIG_CMD_GPT_RENAME
 	"gpt partition renaming commands:\n"
 	" gpt read <interface> <dev>\n"
@@ -872,7 +872,7 @@ U_BOOT_CMD(gpt, CONFIG_SYS_MAXARGS, 1, d
 	" gpt rename <interface> <dev> <part> <name>\n"
 	"    - rename the specified partition\n"
 	" Example usage:\n"
-	" gpt swap mmc 0 foo bar\n"
-	" gpt rename mmc 0 3 foo\n"
+	"   gpt swap mmc 0 foo bar\n"
+	"   gpt rename mmc 0 3 foo\n"
 #endif
 );
