Index: u-boot/drivers/clk/clk_stm32mp1.c
===================================================================
--- u-boot.orig/drivers/clk/clk_stm32mp1.c
+++ u-boot/drivers/clk/clk_stm32mp1.c
@@ -718,7 +718,7 @@ char * const stm32mp1_clk_parent_name[_P
 	[_PCLK3] = "PCLK3",
 	[_PCLK4] = "PCLK4",
 	[_PCLK5] = "PCLK5",
-	[_HCLK6] = "KCLK6",
+	[_HCLK6] = "HCLK6",
 	[_HCLK2] = "HCLK2",
 	[_CK_PER] = "CK_PER",
 	[_CK_MPU] = "CK_MPU",
