diff --git a/lib/image-sparse.c b/lib/image-sparse.c
index 5ec0f94ab3e..188a6d2713a 100644
--- a/lib/image-sparse.c
+++ b/lib/image-sparse.c
@@ -50,7 +50,7 @@
 
 static void default_log(const char *ignored, char *response) {}
 
-static lbaint_t write_sparse_chunk_raw(struct sparse_storage *info,
+static long write_sparse_chunk_raw(struct sparse_storage *info,
 				       lbaint_t blk, lbaint_t blkcnt,
 				       void *data,
 				       char *response)
@@ -111,7 +111,7 @@ int write_sparse_image(struct sparse_storage *info,
 {
 	lbaint_t blk;
 	lbaint_t blkcnt;
-	lbaint_t blks;
+	long blks;
 	uint64_t bytes_written = 0;
 	unsigned int chunk;
 	unsigned int offset;
