Index: u-boot-txmp-1570/arch/arm/dts/Makefile
===================================================================
--- u-boot-txmp-1570.orig/arch/arm/dts/Makefile
+++ u-boot-txmp-1570/arch/arm/dts/Makefile
@@ -1199,6 +1199,7 @@ dtb-$(CONFIG_STM32MP15X) += \
 	stm32mp157c-ev1.dtb \
 	stm32mp157c-odyssey.dtb \
 	stm32mp157c-qsmp-1570.dtb \
+	stm32mp153a-txmp-1530.dtb \
 	stm32mp157c-txmp-1570.dtb \
 	stm32mp157c-txmp-1571.dtb \
 	stm32mp15xx-dhcom-drc02.dtb \
@@ -1312,6 +1313,8 @@ dtb-$(CONFIG_TARGET_CORSTONE1000) += cor
 
 include $(srctree)/scripts/Makefile.dts
 
+dtb-y += stm32mp157c-txmp-1570-mb7.dtb
+
 targets += $(dtb-y)
 
 # Add any required device tree compiler flags here
Index: u-boot-txmp-1570/board/karo/stm32mp1/Kconfig
===================================================================
--- u-boot-txmp-1570.orig/board/karo/stm32mp1/Kconfig
+++ u-boot-txmp-1570/board/karo/stm32mp1/Kconfig
@@ -22,6 +22,16 @@ source "board/st/common/Kconfig"
 
 choice
 	prompt "Ka-Ro STM32 module variant"
+	default KARO_TXMP_1570
+
+config KARO_TXMP_1530
+	bool "Ka-Ro TXMP-1530 module"
+	depends on STM32MP15X
+	select KARO_STM32
+	imply KARO_STM32_EMMC
+	help
+	  Enable support for the Ka-Ro electronics TXMP-1530 module featuring
+	  an STM32MP153A SOC, 512MiB of SDRAM and 4GiB or 8GiB eMMC.
 
 config KARO_TXMP_1570
 	bool "Ka-Ro TXMP-1570 module"
Index: u-boot-txmp-1570/board/karo/stm32mp1/stm32mp1.c
===================================================================
--- u-boot-txmp-1570.orig/board/karo/stm32mp1/stm32mp1.c
+++ u-boot-txmp-1570/board/karo/stm32mp1/stm32mp1.c
@@ -259,7 +259,9 @@ int checkboard(void)
 	else
 		mode = "basic";
 
-#if defined(CONFIG_KARO_TXMP_1570)
+#if defined(CONFIG_KARO_TXMP_1530)
+	printf("Board: TXMP-1530 in %s mode", mode);
+#elif defined(CONFIG_KARO_TXMP_1570)
 	printf("Board: TXMP-1570 in %s mode", mode);
 #elif defined(CONFIG_KARO_TXMP_1571)
 	printf("Board: TXMP-1571 in %s mode", mode);
Index: u-boot-txmp-1570/include/configs/txmp.h
===================================================================
--- u-boot-txmp-1570.orig/include/configs/txmp.h
+++ u-boot-txmp-1570/include/configs/txmp.h
@@ -61,6 +61,8 @@
 #define SOC_FAMILY "stm32mp153a"
 #elif defined(CONFIG_KARO_QSMP_1570)
 #define SOC_FAMILY "stm32mp157c"
+#elif defined(CONFIG_KARO_TXMP_1530)
+#define SOC_FAMILY "stm32mp153a"
 #elif defined(CONFIG_KARO_TXMP_1570)
 #define SOC_FAMILY "stm32mp157c"
 #elif defined(CONFIG_KARO_TXMP_1571)
