diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index c026243ed8d..c669e89d75c 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -1187,7 +1187,6 @@ dtb-$(CONFIG_STM32MP15X) += \
 	stm32mp151a-qsmp-1510.dtb \
 	stm32mp151a-qsmp-1510-qsbase4.dtb \
 	stm32mp153a-qsmp-1530.dtb \
-	stm32mp157a-avenger96.dtb \
 	stm32mp157a-dk1.dtb \
 	stm32mp157a-ed1.dtb \
 	stm32mp157a-ev1.dtb \
