diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 37f0c42749b..94d26943fa7 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -946,6 +946,8 @@ dtb-$(CONFIG_ARCH_STI) += stih410-b2260.dtb
 
 dtb-$(CONFIG_STM32MP15x) += \
 	stm32mp151a-qsmp-1510.dtb \
+	stm32mp151a-qsmp-1510-qsbase2.dtb \
+	stm32mp151a-qsmp-1510-qsbase4.dtb \
 	stm32mp153a-qsmp-1530.dtb \
 	stm32mp157a-dk1.dtb \
 	stm32mp157a-avenger96.dtb \
