Index: u-boot/drivers/firmware/psci.c
===================================================================
--- u-boot.orig/drivers/firmware/psci.c
+++ u-boot/drivers/firmware/psci.c
@@ -21,8 +21,15 @@
 
 #define PSCI_METHOD_HVC 1
 #define PSCI_METHOD_SMC 2
+#define PSCI_METHOD_NONE 3
 
-int __efi_runtime_data psci_method;
+/*
+ * Since this variable is written to prior to relocation, we need to
+ * make sure it doesn't reside in the .bss section (which overlays the
+ * .rel.dyn section).
+ * Initializing it to a non-zero value accomplishes this.
+ */
+static int __efi_runtime_data psci_method = PSCI_METHOD_NONE;
 
 unsigned long __efi_runtime invoke_psci_fn
 		(unsigned long function_id, unsigned long arg0,
