diff --git a/board/karo/txmp/Kconfig b/board/karo/stm32mp1/Kconfig
similarity index 78%
rename from board/karo/txmp/Kconfig
rename to board/karo/stm32mp1/Kconfig
index 44ca4d2182d..017f25c788b 100644
--- a/board/karo/txmp/Kconfig
+++ b/board/karo/stm32mp1/Kconfig
@@ -1,6 +1,6 @@
 source "board/karo/common/Kconfig"
 
-if TARGET_KARO_TXMP
+if TARGET_KARO_STM32MP1
 
 config SYS_BOARD
 	default "txmp"
@@ -21,10 +21,12 @@ config KARO_STM32
 choice
 	prompt "Ka-Ro STM32 module variant"
 
+if STM32MP15X
+
 config KARO_TXMP_1570
 	bool "Ka-Ro TXMP-1570 module"
 	select KARO_STM32
-	imply TXMP_EMMC
+	imply KARO_STM32_EMMC
 	help
 	  Enable support for the Ka-Ro electronics TXMP-1570 module featuring
 	  an STM32MP157C SOC, 512MiB of SDRAM and 4GiB or 8GiB eMMC.
@@ -32,7 +34,7 @@ config KARO_TXMP_1570
 config KARO_TXMP_1571
 	bool "Ka-Ro TXMP-1571 module"
 	select KARO_STM32
-	imply TXMP_EMMC
+	imply KARO_STM32_EMMC
 	help
 	  TXMP-1570 with corrected pinout compatible with the rest of the
 	  TX module family
@@ -40,7 +42,7 @@ config KARO_TXMP_1571
 config KARO_QSMP_1510
 	bool "Ka-Ro QSMP-1510 solder-in module"
 	select KARO_STM32
-	select TXMP_NAND
+	select KARO_STM32_NAND
 	help
 	  Enable support for the Ka-Ro electronics QSMP-1510 module featuring
 	  an STM32MP151A SOC, 256MiB of SDRAM and 128MiB NAND flash on a
@@ -49,7 +51,7 @@ config KARO_QSMP_1510
 config KARO_QSMP_1530
 	bool "Ka-Ro QSMP-1530 solder-in module"
 	select KARO_STM32
-	imply TXMP_EMMC
+	imply KARO_STM32_EMMC
 	help
 	  Enable support for the Ka-Ro electronics QSMP-1530 module featuring
 	  an STM32MP153A SOC, 256MiB of SDRAM and 4GiB eMMC on a solder-in
@@ -58,27 +60,42 @@ config KARO_QSMP_1530
 config KARO_QSMP_1570
 	bool "Ka-Ro QSMP-1570 solder-in module"
 	select KARO_STM32
-	imply TXMP_EMMC
+	imply KARO_STM32_EMMC
 	help
 	  Enable support for the Ka-Ro electronics QSMP-1570 module featuring
 	  an STM32MP157C SOC, 512MiB of SDRAM and 4GiB eMMC on a solder-in
 	  module.
 
+endif
+
+if STM32MP13X
+
+config KARO_QSMP_1351
+	bool "Ka-Ro QSMP-1351 solder-in module"
+	select KARO_STM32
+	imply KARO_STM32_EMMC
+	help
+	  Enable support for the Ka-Ro electronics QSMP-1351 module featuring
+	  an STM32MP135C SOC, 256MiB of SDRAM and 4GiB eMMC on a solder-in
+	  module.
+
+endif
+
 endchoice
 
 choice
 	prompt "Ka-Ro TXMP/QSMP boot device"
-	default TXMP_EMMC
+	default KARO_STM32_EMMC
 
-config TXMP_EMMC
+config KARO_STM32_EMMC
 	bool "eMMC device"
 
-config TXMP_NAND
+config KARO_STM32_NAND
 	bool "SPI NAND device"
 	depends on STM32_QSPI
 	select DM_MTD
 
-config TXMP_SPI
+config KARO_STM32_SPI
 	bool "SPI NOR device"
 	depends on STM32_QSPI
 	select DM_MTD
diff --git a/board/karo/txmp/Makefile b/board/karo/stm32mp1/Makefile
similarity index 100%
rename from board/karo/txmp/Makefile
rename to board/karo/stm32mp1/Makefile
diff --git a/board/karo/txmp/board_mtdparts.c b/board/karo/stm32mp1/board_mtdparts.c
similarity index 100%
rename from board/karo/txmp/board_mtdparts.c
rename to board/karo/stm32mp1/board_mtdparts.c
diff --git a/board/karo/txmp/config.mk b/board/karo/stm32mp1/config.mk
similarity index 100%
rename from board/karo/txmp/config.mk
rename to board/karo/stm32mp1/config.mk
diff --git a/board/karo/txmp/dbg-uart.c b/board/karo/stm32mp1/dbg-uart.c
similarity index 100%
rename from board/karo/txmp/dbg-uart.c
rename to board/karo/stm32mp1/dbg-uart.c
diff --git a/board/karo/txmp/dfu.c b/board/karo/stm32mp1/dfu.c
similarity index 100%
rename from board/karo/txmp/dfu.c
rename to board/karo/stm32mp1/dfu.c
diff --git a/board/karo/txmp/qsmp-1510_env.txt b/board/karo/stm32mp1/qsmp-1510_env.txt
similarity index 100%
rename from board/karo/txmp/qsmp-1510_env.txt
rename to board/karo/stm32mp1/qsmp-1510_env.txt
diff --git a/board/karo/txmp/qsmp-1530_env.txt b/board/karo/stm32mp1/qsmp-1530_env.txt
similarity index 100%
rename from board/karo/txmp/qsmp-1530_env.txt
rename to board/karo/stm32mp1/qsmp-1530_env.txt
diff --git a/board/karo/txmp/qsmp-1570_env.txt b/board/karo/stm32mp1/qsmp-1570_env.txt
similarity index 100%
rename from board/karo/txmp/qsmp-1570_env.txt
rename to board/karo/stm32mp1/qsmp-1570_env.txt
diff --git a/board/karo/txmp/txmp-1570_env.txt b/board/karo/stm32mp1/txmp-1570_env.txt
similarity index 100%
rename from board/karo/txmp/txmp-1570_env.txt
rename to board/karo/stm32mp1/txmp-1570_env.txt
diff --git a/board/karo/txmp/txmp-1571_env.txt b/board/karo/stm32mp1/txmp-1571_env.txt
similarity index 100%
rename from board/karo/txmp/txmp-1571_env.txt
rename to board/karo/stm32mp1/txmp-1571_env.txt
diff --git a/board/karo/txmp/txmp.c b/board/karo/stm32mp1/txmp.c
similarity index 99%
rename from board/karo/txmp/txmp.c
rename to board/karo/stm32mp1/txmp.c
index be466e2aa41..a0ff3b6cfc4 100644
--- a/board/karo/txmp/txmp.c
+++ b/board/karo/stm32mp1/txmp.c
@@ -131,6 +131,8 @@ int checkboard(void)
 	printf("Board: QSMP-1530 in %s mode", mode);
 #elif defined(CONFIG_KARO_QSMP_1570)
 	printf("Board: QSMP-1570 in %s mode", mode);
+#elif defined(CONFIG_KARO_QSMP_1351)
+	printf("Board: QSMP-1351 in %s mode", mode);
 #else
 #error Unsupported Board type
 #endif
