Index: u-boot/drivers/video/stm32/stm32_ltdc.c
===================================================================
--- u-boot.orig/drivers/video/stm32/stm32_ltdc.c
+++ u-boot/drivers/video/stm32/stm32_ltdc.c
@@ -252,7 +252,8 @@ static void stm32_ltdc_set_mode(struct s
 		val |= GCR_HSPOL;
 	if (timings->flags & DISPLAY_FLAGS_VSYNC_HIGH)
 		val |= GCR_VSPOL;
-	if (timings->flags & DISPLAY_FLAGS_DE_HIGH)
+	if (timings->flags & DISPLAY_FLAGS_DE_LOW)
+		/* DEPOL: '_NOT_ data enable polarity'! */
 		val |= GCR_DEPOL;
 	if (timings->flags & DISPLAY_FLAGS_PIXDATA_NEGEDGE)
 		val |= GCR_PCPOL;
@@ -375,10 +376,10 @@ static int stm32_ltdc_probe(struct udevi
 
 	ret = clk_set_rate(&pclk, timings.pixelclock.typ);
 	if (ret)
-		dev_warn(dev, "fail to set pixel clock %d hz\n",
+		dev_warn(dev, "failed to set pixel clock to %d Hz\n",
 			 timings.pixelclock.typ);
 
-	debug("%s: Set pixel clock req %d hz get %ld hz\n", __func__,
+	debug("%s: Set pixel clock req %d hz got %ld Hz\n", __func__,
 	      timings.pixelclock.typ, clk_get_rate(&pclk));
 
 	ret = reset_get_by_index(dev, 0, &rst);
@@ -392,8 +393,10 @@ static int stm32_ltdc_probe(struct udevi
 
 	if (IS_ENABLED(CONFIG_VIDEO_BRIDGE)) {
 		ret = uclass_get_device(UCLASS_VIDEO_BRIDGE, 0, &bridge);
-		if (ret)
-			debug("No video bridge, or no backlight on bridge\n");
+		if (ret) {
+			dev_err(dev, "No video bridge, or no backlight on bridge\n");
+			return ret;
+		}
 
 		if (bridge) {
 			ret = video_bridge_attach(bridge);
