diff --git a/board/karo/stm32mp1/Kconfig b/board/karo/stm32mp1/Kconfig
index 92711fc1b99..812061189de 100644
--- a/board/karo/stm32mp1/Kconfig
+++ b/board/karo/stm32mp1/Kconfig
@@ -23,10 +23,9 @@ source "board/st/common/Kconfig"
 choice
 	prompt "Ka-Ro STM32 module variant"
 
-if STM32MP15X
-
 config KARO_TXMP_1570
 	bool "Ka-Ro TXMP-1570 module"
+	depends on STM32MP15X
 	select KARO_STM32
 	imply KARO_STM32_EMMC
 	help
@@ -35,6 +34,7 @@ config KARO_TXMP_1570
 
 config KARO_TXMP_1571
 	bool "Ka-Ro TXMP-1571 module"
+	depends on STM32MP15X
 	select KARO_STM32
 	imply KARO_STM32_EMMC
 	help
@@ -43,6 +43,7 @@ config KARO_TXMP_1571
 
 config KARO_QSMP_1510
 	bool "Ka-Ro QSMP-1510 solder-in module"
+	depends on STM32MP15X
 	select KARO_STM32
 	select KARO_STM32_NAND
 	help
@@ -52,6 +53,7 @@ config KARO_QSMP_1510
 
 config KARO_QSMP_1530
 	bool "Ka-Ro QSMP-1530 solder-in module"
+	depends on STM32MP15X
 	select KARO_STM32
 	imply KARO_STM32_EMMC
 	help
@@ -61,6 +63,7 @@ config KARO_QSMP_1530
 
 config KARO_QSMP_1570
 	bool "Ka-Ro QSMP-1570 solder-in module"
+	depends on STM32MP15X
 	select KARO_STM32
 	imply KARO_STM32_EMMC
 	help
@@ -68,12 +71,9 @@ config KARO_QSMP_1570
 	  an STM32MP157C SOC, 512MiB of SDRAM and 4GiB eMMC on a solder-in
 	  module.
 
-endif
-
-if STM32MP13X
-
 config KARO_QSMP_1351
 	bool "Ka-Ro QSMP-1351 solder-in module"
+	depends on STM32MP13X
 	select KARO_STM32
 	imply KARO_STM32_EMMC
 	help
@@ -81,8 +81,6 @@ config KARO_QSMP_1351
 	  an STM32MP135C SOC, 256MiB of SDRAM and 4GiB eMMC on a solder-in
 	  module.
 
-endif
-
 endchoice
 
 choice
