diff --git a/arch/arm/mach-stm32mp/Kconfig b/arch/arm/mach-stm32mp/Kconfig
index b28a0c6b625..91246e7f205 100644
--- a/arch/arm/mach-stm32mp/Kconfig
+++ b/arch/arm/mach-stm32mp/Kconfig
@@ -122,5 +122,5 @@ source "arch/arm/mach-stm32mp/Kconfig.13x"
 source "arch/arm/mach-stm32mp/Kconfig.15x"
 
 source "arch/arm/mach-stm32mp/cmd_stm32prog/Kconfig"
-source "board/karo/txmp/Kconfig"
+source "board/karo/stm32mp1/Kconfig"
 endif
diff --git a/board/karo/common/Kconfig b/board/karo/common/Kconfig
index 4a080329555..2cb2becd691 100644
--- a/board/karo/common/Kconfig
+++ b/board/karo/common/Kconfig
@@ -11,8 +11,7 @@ config SYS_FDT_ADDR
 	help
 	  The address the FDT file should be loaded to.
 
-config TARGET_KARO_TXMP
-	bool "Support Ka-Ro TXMP modules"
-	select STM32MP15x
+config TARGET_KARO_STM32MP1
+	bool "Support Ka-Ro STM32MP1 modules"
 	help
-		Ka-Ro electronics TXMP/QSMP board support
+	  Ka-Ro electronics STM32MP1 based TXMP/QSMP board support
diff --git a/board/karo/common/env.c b/board/karo/common/env.c
index 1980e3effa7..b6bf3cb0e7b 100644
--- a/board/karo/common/env.c
+++ b/board/karo/common/env.c
@@ -44,7 +44,7 @@ static const char * const cleanup_vars[] = {
 	"wdreset",
 };
 
-#if defined(CONFIG_TXMP_EMMC) && !defined(CONFIG_KARO_UBOOT_MFG)
+#if defined(CONFIG_KARO_STM32_EMMC) && !defined(CONFIG_KARO_UBOOT_MFG)
 static void karo_set_part_uuids(void)
 {
 	int ret;
diff --git a/configs/qsmp-1510_defconfig b/configs/qsmp-1510_defconfig
index 88d46290c43..e3cfc7f9a8a 100644
--- a/configs/qsmp-1510_defconfig
+++ b/configs/qsmp-1510_defconfig
@@ -7,7 +7,7 @@ CONFIG_SPL_TEXT_BASE=0x2FFC2500
 CONFIG_SPL=y
 CONFIG_TARGET_KARO_TXMP=y
 CONFIG_KARO_QSMP_1510=y
-CONFIG_TXMP_NAND=y
+CONFIG_KARO_STM32_NAND=y
 # CONFIG_SPL_LIBDISK_SUPPORT is not set
 CONFIG_SPL_SPI_SUPPORT=y
 # CONFIG_ARMV7_VIRT is not set
diff --git a/configs/qsmp-1510_mfg_defconfig b/configs/qsmp-1510_mfg_defconfig
index 1e1aac28e50..471d9396058 100644
--- a/configs/qsmp-1510_mfg_defconfig
+++ b/configs/qsmp-1510_mfg_defconfig
@@ -6,7 +6,7 @@ CONFIG_SYS_MALLOC_LEN=0x2000000
 CONFIG_TARGET_KARO_TXMP=y
 CONFIG_CMD_STM32PROG=y
 CONFIG_KARO_QSMP_1510=y
-CONFIG_TXMP_NAND=y
+CONFIG_KARO_STM32_NAND=y
 CONFIG_KARO_UBOOT_MFG=y
 CONFIG_DEFAULT_DEVICE_TREE="stm32mp151a-qsmp-1510"
 CONFIG_CHECK_REL_DYN=y
diff --git a/configs/qsmp-1510_noenv_defconfig b/configs/qsmp-1510_noenv_defconfig
index 6ec738b0bc3..35cdd4d7e5f 100644
--- a/configs/qsmp-1510_noenv_defconfig
+++ b/configs/qsmp-1510_noenv_defconfig
@@ -7,7 +7,7 @@ CONFIG_SPL_TEXT_BASE=0x2FFC2500
 CONFIG_SPL=y
 CONFIG_TARGET_KARO_TXMP=y
 CONFIG_KARO_QSMP_1510=y
-CONFIG_TXMP_NAND=y
+CONFIG_KARO_STM32_NAND=y
 CONFIG_KARO_UBOOT_NOENV=y
 CONFIG_SPL_SPI_SUPPORT=y
 # CONFIG_ARMV7_VIRT is not set
diff --git a/configs/qsmp-1510_trusted_defconfig b/configs/qsmp-1510_trusted_defconfig
index a077219f9a7..efddb964545 100644
--- a/configs/qsmp-1510_trusted_defconfig
+++ b/configs/qsmp-1510_trusted_defconfig
@@ -5,7 +5,7 @@ CONFIG_SYS_MALLOC_F_LEN=0x3000
 CONFIG_SYS_MALLOC_LEN=0x2000000
 CONFIG_TARGET_KARO_TXMP=y
 CONFIG_KARO_QSMP_1510=y
-CONFIG_TXMP_NAND=y
+CONFIG_KARO_STM32_NAND=y
 CONFIG_DEFAULT_DEVICE_TREE="stm32mp151a-qsmp-1510"
 CONFIG_CHECK_REL_DYN=y
 CONFIG_DISTRO_DEFAULTS=y
diff --git a/include/configs/txmp.h b/include/configs/txmp.h
index 7c1f416708f..cf9fcce1312 100644
--- a/include/configs/txmp.h
+++ b/include/configs/txmp.h
@@ -73,26 +73,28 @@
 #define TXMP_NOR_MTDPARTS		""
 #endif
 
-#ifdef CONFIG_TXMP_NAND
-#define TXMP_NAND_MTDPARTS \
+#ifdef CONFIG_KARO_STM32_NAND
+#define KARO_STM32_NAND_MTDPARTS \
 	"mtdparts_nand0=512k(fsbl),2m(ssbl),8m(linux),32m(rootfs),86528k(userfs),512k(dtb),512k@0x7f80000(bbt)ro\0"
 #else
-#define TXMP_NAND_MTDPARTS		""
+#define KARO_STM32_NAND_MTDPARTS		""
 #endif
 #define VIDEOMODE_DEFAULT		"videomode=ETV570\0"
 
 #define FDT_ADDR_STR			"c4000000"
 
 #if defined(CONFIG_KARO_TXMP_1570)
-#define TXMP_BASEBOARD			"mb7"
+#define KARO_STM32_BASEBOARD			"mb7"
 #elif defined(CONFIG_KARO_QSMP_1510)
-#define TXMP_BASEBOARD			"qsbase1"
+#define KARO_STM32_BASEBOARD			"qsbase1"
 #elif defined(CONFIG_KARO_QSMP_1530)
-#define TXMP_BASEBOARD			"qsbase1"
+#define KARO_STM32_BASEBOARD			"qsbase1"
 #elif defined(CONFIG_KARO_QSMP_1570)
-#define TXMP_BASEBOARD			"qsbase1"
+#define KARO_STM32_BASEBOARD			"qsbase1"
+#elif defined(CONFIG_KARO_QSMP_1351)
+#define KARO_STM32_BASEBOARD			"qsbase1"
 #else
-#define TXMP_BASEBOARD			""
+#define KARO_STM32_BASEBOARD			""
 #endif
 
 #endif /* CONFIG_SPL_BUILD */
