commit dfc11608f54c35bd4dd85209aa8d9a9665bdd117
Author: Markus Bauer <mb@karo-electronics.de>
Date:   Fri Jan 12 16:04:04 2024 +0100

    stm32mp13 part 3

diff --git a/board/karo/stm32mp1/Makefile b/board/karo/stm32mp1/Makefile
index f19e8ae3cd6..fc2548cab3f 100644
--- a/board/karo/stm32mp1/Makefile
+++ b/board/karo/stm32mp1/Makefile
@@ -4,7 +4,7 @@
 #
 
 ifeq ($(CONFIG_SPL_BUILD),)
-	obj-y += txmp.o
+	obj-y += stm32mp1.o
 endif
 obj-$(CONFIG_DEBUG_UART_BOARD_INIT) += dbg-uart.o
 obj-$(CONFIG_SYS_MTDPARTS_RUNTIME) += board_mtdparts.o
diff --git a/board/karo/stm32mp1/txmp.c b/board/karo/stm32mp1/stm32mp1.c
similarity index 100%
rename from board/karo/stm32mp1/txmp.c
rename to board/karo/stm32mp1/stm32mp1.c
