diff --git a/board/karo/txmp/txmp.c b/board/karo/txmp/txmp.c
index f97971cce48..da0a8c0b4d5 100644
--- a/board/karo/txmp/txmp.c
+++ b/board/karo/txmp/txmp.c
@@ -127,6 +127,8 @@ int checkboard(void)
 	printf("Board: TXMP-1570 in %s mode", mode);
 #elif defined(CONFIG_KARO_TXMP_1571)
 	printf("Board: TXMP-1571 in %s mode", mode);
+#elif defined(CONFIG_KARO_QSMP_1351)
+	printf("Board: QSMP-1351 in %s mode", mode);
 #elif defined(CONFIG_KARO_QSMP_1510)
 	printf("Board: QSMP-1510 in %s mode", mode);
 #elif defined(CONFIG_KARO_QSMP_1530)
