Index: u-boot/net/tftp.c
===================================================================
--- u-boot.orig/net/tftp.c
+++ u-boot/net/tftp.c
@@ -229,7 +229,6 @@ static int load_block(unsigned block, uc
 }
 #endif
 
-static void tftp_send(void);
 static void tftp_timeout_handler(void);
 
 /**********************************************************************/
@@ -746,7 +745,7 @@ void tftp_start(enum proto_t protocol)
 	tftp_remote_port = WELL_KNOWN_PORT;
 	timeout_count = 0;
 	/* Use a pseudo-random port unless a specific port is set */
-	tftp_our_port = 1024 + (get_timer(0) % 3072);
+	tftp_our_port = 1024 + (rand() % 3072);
 
 #ifdef CONFIG_TFTP_PORT
 	ep = env_get("tftpdstp");
