diff --git a/drivers/usb/gadget/f_mass_storage.c b/drivers/usb/gadget/f_mass_storage.c
index 439a31c255..92f85a5d99 100644
--- a/drivers/usb/gadget/f_mass_storage.c
+++ b/drivers/usb/gadget/f_mass_storage.c
@@ -245,6 +245,7 @@
 #include <common.h>
 #include <console.h>
 #include <g_dnl.h>
+#include <watchdog.h>
 #include <dm/devres.h>
 
 #include <linux/err.h>
@@ -670,6 +671,8 @@ static int sleep_thread(struct fsg_common *common)
 			if (ctrlc())
 				return -EPIPE;
 
+			WATCHDOG_RESET();
+
 			/* Check cable connection */
 			if (!g_dnl_board_usb_cable_connected())
 				return -EIO;
