Index: linux-dunfell-5.10.10/arch/arm/boot/dts/stm32mp151.dtsi
===================================================================
--- linux-dunfell-5.10.10.orig/arch/arm/boot/dts/stm32mp151.dtsi
+++ linux-dunfell-5.10.10/arch/arm/boot/dts/stm32mp151.dtsi
@@ -1621,12 +1621,6 @@
 			status = "disabled";
 		};
 
-		stmmac_axi_config_0: stmmac-axi-config {
-			snps,wr_osr_lmt = <0x7>;
-			snps,rd_osr_lmt = <0x7>;
-			snps,blen = <0 0 0 0 16 8 4>;
-		};
-
 		ethernet0: ethernet@5800a000 {
 			compatible = "st,stm32mp1-dwmac", "snps,dwmac-4.20a";
 			reg = <0x5800a000 0x2000>;
@@ -2031,4 +2025,10 @@
 			};
 		};
 	};
+
+	stmmac_axi_config_0: stmmac-axi-config {
+		snps,wr_osr_lmt = <0x7>;
+		snps,rd_osr_lmt = <0x7>;
+		snps,blen = <0 0 0 0 16 8 4>;
+	};
 };
