Index: linux-dunfell-5.10.10/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
===================================================================
--- linux-dunfell-5.10.10.orig/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ linux-dunfell-5.10.10/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -5152,9 +5152,14 @@ int stmmac_suspend(struct device *dev)
 	struct stmmac_priv *priv = netdev_priv(ndev);
 	u32 chan;
 
-	if (!ndev || !netif_running(ndev))
+	if (!ndev)
 		return 0;
 
+	if (!netif_running(ndev)) {
+		pinctrl_pm_select_sleep_state(priv->device);
+		return 0;
+	}
+
 	phylink_mac_change(priv->phylink, false);
 
 	mutex_lock(&priv->lock);
@@ -5242,8 +5247,10 @@ int stmmac_resume(struct device *dev)
 	struct stmmac_priv *priv = netdev_priv(ndev);
 	int ret;
 
-	if (!netif_running(ndev))
+	if (!netif_running(ndev)) {
+		pinctrl_pm_select_default_state(priv->device);
 		return 0;
+	}
 
 	/* Power Down bit, into the PM register, is cleared
 	 * automatically as soon as a magic packet or a Wake-up frame
