Index: linux-dunfell-5.10.10/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
===================================================================
--- linux-dunfell-5.10.10.orig/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
+++ linux-dunfell-5.10.10/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
@@ -307,6 +307,8 @@ int stmmac_mdio_reset(struct mii_bus *bu
 		reset_gpio = devm_gpiod_get_optional(priv->device,
 						     "snps,reset",
 						     GPIOD_OUT_LOW);
+		if (!reset_gpio)
+			goto out;
 		if (IS_ERR(reset_gpio))
 			return PTR_ERR(reset_gpio);
 
@@ -324,8 +326,11 @@ int stmmac_mdio_reset(struct mii_bus *bu
 		gpiod_set_value_cansleep(reset_gpio, 0);
 		if (delays[2])
 			msleep(DIV_ROUND_UP(delays[2], 1000));
+
+		gpiod_put(reset_gpio);
 	}
 #endif
+ out:
 
 	/* This is a workaround for problems with the STE101P PHY.
 	 * It doesn't complete its reset until at least one clock cycle
