Index: karo-tx-linux/arch/arm/boot/dts/stm32mp15xxab-pinctrl.dtsi
===================================================================
--- karo-tx-linux.orig/arch/arm/boot/dts/stm32mp15xxab-pinctrl.dtsi
+++ karo-tx-linux/arch/arm/boot/dts/stm32mp15xxab-pinctrl.dtsi
@@ -4,6 +4,12 @@
  * Author: Alexandre Torgue <alexandre.torgue@st.com> for STMicroelectronics.
  */
 
+/ {
+	soc {
+		/delete-node/ pin-controller-z@54004000;
+	};
+};
+
 &pinctrl {
 	st,package = <STM32MP_PKG_AB>;
 
Index: karo-tx-linux/arch/arm/boot/dts/stm32mp15xxad-pinctrl.dtsi
===================================================================
--- karo-tx-linux.orig/arch/arm/boot/dts/stm32mp15xxad-pinctrl.dtsi
+++ karo-tx-linux/arch/arm/boot/dts/stm32mp15xxad-pinctrl.dtsi
@@ -4,6 +4,12 @@
  * Author: Alexandre Torgue <alexandre.torgue@st.com> for STMicroelectronics.
  */
 
+/ {
+	soc {
+		/delete-node/ pin-controller-z@54004000;
+	};
+};
+
 &pinctrl {
 	st,package = <STM32MP_PKG_AD>;
 
