Index: karo-tx-linux/arch/arm/boot/dts/Makefile
===================================================================
--- karo-tx-linux.orig/arch/arm/boot/dts/Makefile
+++ karo-tx-linux/arch/arm/boot/dts/Makefile
@@ -1030,10 +1030,19 @@ dtb-$(CONFIG_ARCH_STM32) += \
 	stm32h743i-disco.dtb \
 	stm32mp157a-avenger96.dtb \
 	stm32mp157a-dk1.dtb \
+	stm32mp153a-qsmp-1530.dtb \
+	stm32mp153a-qsmp-1530-qsbase1.dtb \
+	stm32mp153a-txmp-1530.dtb \
+	stm32mp153a-txmp-1530-mb7.dtb \
 	stm32mp157c-dhcom-pdk2.dtb \
 	stm32mp157c-dk2.dtb \
 	stm32mp157c-ed1.dtb \
-	stm32mp157c-ev1.dtb
+	stm32mp157c-ev1.dtb \
+	stm32mp157c-qsmp-1570.dtb \
+	stm32mp157c-qsmp-1570-qsbase1.dtb \
+	stm32mp157c-txmp-1570.dtb \
+	stm32mp157c-txmp-1570-mb7.dtb \
+	stm32mp157c-txmp-1570-mipi-mb.dtb
 dtb-$(CONFIG_MACH_SUN4I) += \
 	sun4i-a10-a1000.dtb \
 	sun4i-a10-ba10-tvbox.dtb \
