From cfb176852494f6bad30f5860b0ccf99bcd5ef4c0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lothar=20Wa=C3=9Fmann?= <LW@KARO-electronics.de>
Date: Wed, 27 Mar 2019 14:55:34 +0100
Subject: [PATCH 14/35] common: autoboot: make 'Normal Boot' a debug message

---
 common/autoboot.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/common/autoboot.c b/common/autoboot.c
index 8ad3495..7d00c9e 100644
--- a/common/autoboot.c
+++ b/common/autoboot.c
@@ -321,7 +321,7 @@ const char *bootdelay_process(void)
 		printf("Boot from USB for uuu\n");
 		env_set("bootcmd", "fastboot 0");
 	} else {
-		printf("Normal Boot\n");
+		debug("Normal Boot\n");
 	}
 #endif
 
-- 
2.7.4

