From 0217cd3508a4f456876be562dea82805631eb67e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lothar=20Wa=C3=9Fmann?= <LW@KARO-electronics.de>
Date: Wed, 27 Mar 2019 14:58:33 +0100
Subject: [PATCH 17/35] common: add call to show_activity() in main cmd loop

---
 common/cli_readline.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/common/cli_readline.c b/common/cli_readline.c
index ecded11..a059ea1 100644
--- a/common/cli_readline.c
+++ b/common/cli_readline.c
@@ -272,6 +272,12 @@ static int cread_line(const char *const prompt, char *buf, unsigned int *len,
 			first = 0;
 		}
 
+#ifdef CONFIG_SHOW_ACTIVITY
+		while (!tstc()) {	/* while no incoming data */
+			show_activity(0);
+			WATCHDOG_RESET();
+		}
+#endif
 		ichar = getcmd_getch();
 
 		if ((ichar == '\n') || (ichar == '\r')) {
-- 
2.7.4

