Index: tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c
+++ tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c
@@ -29,6 +29,7 @@ int ddrphy_phyinit_sequence(struct stm32
 	struct pmu_smb_ddr_1d mb_ddr_1d; /* Firmware 1D Message Block structure */
 
 	/* Check user input pstate number consistency vs. SW capabilities */
+WARN("%s@%d: \n", __func__, __LINE__);
 	if (config->uib.numpstates > 1U) {
 		return -1;
 	}
@@ -40,6 +41,7 @@ int ddrphy_phyinit_sequence(struct stm32
 
 	/* Re-calculate Firmware Message Block input based on final user input */
 	ret = ddrphy_phyinit_calcmb(config, &mb_ddr_1d);
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 	if (ret != 0) {
 		return ret;
 	}
@@ -52,6 +54,7 @@ int ddrphy_phyinit_sequence(struct stm32
 
 	/* (C) Initialize PHY Configuration */
 	ret = ddrphy_phyinit_c_initphyconfig(config, &mb_ddr_1d, &ardptrinitval);
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 	if (ret != 0) {
 		return ret;
 	}
@@ -63,6 +66,7 @@ int ddrphy_phyinit_sequence(struct stm32
 
 	/* Stop retention register tracking for training firmware related registers */
 	ret = ddrphy_phyinit_reginterface(STOPTRACK, 0U, 0U);
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 	if (ret != 0) {
 		return ret;
 	}
@@ -79,12 +83,14 @@ int ddrphy_phyinit_sequence(struct stm32
 
 		/* (F) Write the Message Block parameters for the training firmware */
 		ret = ddrphy_phyinit_f_loaddmem(config, &mb_ddr_1d);
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 		if (ret != 0) {
 			return ret;
 		}
 
 		/* (G) Execute the Training Firmware */
 		ret = ddrphy_phyinit_g_execfw();
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 		if (ret != 0) {
 			return ret;
 		}
@@ -95,6 +101,7 @@ int ddrphy_phyinit_sequence(struct stm32
 
 	/* Start retention register tracking for training firmware related registers */
 	ret = ddrphy_phyinit_reginterface(STARTTRACK, 0U, 0U);
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 	if (ret != 0) {
 		return ret;
 	}
@@ -111,6 +118,7 @@ int ddrphy_phyinit_sequence(struct stm32
 	if (reten) {
 		/* Save value of tracked registers for retention restore sequence. */
 		ret = ddrphy_phyinit_usercustom_saveretregs(config);
+WARN("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 		if (ret != 0) {
 			return ret;
 		}
Index: tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
+++ tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
@@ -68,6 +68,7 @@ static int get_major_message(uint32_t *m
 	int ret;
 
 	ret = wait_uctwriteprotshadow(false);
+VERBOSE("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 	if (ret != 0) {
 		return ret;
 	}
@@ -77,6 +78,7 @@ static int get_major_message(uint32_t *m
 								   CSR_UCTWRITEONLYSHADOW_ADDR))));
 
 	ret = ack_message_receipt();
+VERBOSE("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 	if (ret != 0) {
 		return ret;
 	}
@@ -141,6 +143,7 @@ int ddrphy_phyinit_usercustom_g_waitfwdo
 
 	do {
 		ret = get_major_message(&fw_major_message);
+VERBOSE("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 		if (ret != 0) {
 			return ret;
 		}
@@ -153,6 +156,7 @@ int ddrphy_phyinit_usercustom_g_waitfwdo
 			uint32_t stream_len;
 
 			ret = get_streaming_message(&read_data);
+VERBOSE("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 			if (ret != 0) {
 				return ret;
 			}
@@ -161,6 +165,7 @@ int ddrphy_phyinit_usercustom_g_waitfwdo
 
 			for (i = 0U; i < stream_len; i++) {
 				ret = get_streaming_message(&read_data);
+VERBOSE("%s@%d: ret=%d\n", __func__, __LINE__, ret);
 				if (ret != 0) {
 					return ret;
 				}
Index: tf-a-txmp-2550/plat/st/stm32mp2/plat_ddr.c
===================================================================
--- tf-a-txmp-2550.orig/plat/st/stm32mp2/plat_ddr.c
+++ tf-a-txmp-2550/plat/st/stm32mp2/plat_ddr.c
@@ -46,21 +46,25 @@ static int ddr_power_init(void *fdt, int
 	 * enable VREF_DDR, VTT_DDR, VPP_DDR
 	 */
 	status = regulator_set_min_voltage(supply.vdd);
+WARN("%s@%d: status=%d\n", __func__, __LINE__, status);
 	if (status != 0) {
 		return status;
 	}
 
 	status = regulator_enable(supply.vdd);
+WARN("%s@%d: status=%d\n", __func__, __LINE__, status);
 	if (status != 0) {
 		return status;
 	}
 
 	status = regulator_enable(supply.vref);
+WARN("%s@%d: status=%d\n", __func__, __LINE__, status);
 	if (status != 0) {
 		return status;
 	}
 
 	status = regulator_enable(supply.vtt);
+WARN("%s@%d: status=%d\n", __func__, __LINE__, status);
 	return status;
 	}
 #endif /* STM32MP_DDR3_TYPE */
@@ -146,6 +150,8 @@ static int ddr_power_init(void *fdt, int
 	struct lpddr4_supply supply;
 
 	lpddr4_supply_read(fdt, node, &supply);
+WARN("%s@%d: vdd1=%p vdd2=%p vddq=%p\n", __func__, __LINE__,
+     supply.vdd1, supply.vdd2, supply.vddq);
 	if ((supply.vdd1 == NULL) || (supply.vdd2 == NULL) || (supply.vddq == NULL)) {
 		return -ENOENT;
 	}
@@ -203,11 +209,13 @@ int stm32mp_board_ddr_power_init(enum dd
 	panic();
 #endif
 
+WARN("%s@%d: \n", __func__, __LINE__);
 	if (fdt_get_address(&fdt) == 0) {
 		return -FDT_ERR_NOTFOUND;
 	}
 
 	node = fdt_node_offset_by_compatible(fdt, -1, DT_DDR_COMPAT);
+WARN("%s@%d: node=%d\n", __func__, __LINE__, node);
 	if (node < 0) {
 		ERROR("%s: Cannot read DDR node in DT\n", __func__);
 		return node;
Index: tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h
+++ tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h
@@ -35,4 +35,7 @@ int ddrphy_phyinit_softsetmb(struct pmu_
 			     uint32_t value);
 void ddrphy_phyinit_initstruct(struct stm32mp_ddr_config *config, struct pmu_smb_ddr_1d *mb_ddr_1d);
 
+#undef VERBOSE
+#define VERBOSE(...)	WARN(__VA_ARGS__)
+
 #endif /* DDRPHY_PHYINIT_H */
Index: tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_writeoutmem.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_writeoutmem.c
+++ tf-a-txmp-2550/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_writeoutmem.c
@@ -13,6 +13,17 @@
 
 #include <platform_def.h>
 
+static inline void __mmio_write_16(uintptr_t addr, uint16_t val,
+				   const char *fn, int ln)
+{
+#if 1
+	VERBOSE("%s@%d: Writing %04x to %08lx\n", __func__, __LINE__,
+		val, addr);
+#endif
+	mmio_write_16(addr, val);
+}
+#define _mmio_write_16(a, v)	__mmio_write_16(a, v, __func__, __LINE__)
+
 /*
  * Writes local memory content into the SRAM via APB interface.
  *
@@ -29,11 +40,16 @@ void ddrphy_phyinit_writeoutmem(uint32_t
 {
 	uint32_t index;
 
+VERBOSE("%s@%d: Writing %u(%x) bytes from %p..%p to %08x..%08x\n", __func__, __LINE__,
+	mem_size, mem_size, mem, &mem[mem_size / sizeof(uint32_t)],
+	DDRPHYC_BASE + mem_offset, DDRPHYC_BASE + mem_offset + mem_size);
+
+VERBOSE("%s@%d: %08x %08x %08x %08x \n", __func__, __LINE__, mem[0], mem[1], mem[2], mem[3]);
 	/*
 	 * 1. Enable access to the internal CSRs by setting the MicroContMuxSel CSR to 0.
 	 *    This allows the memory controller unrestricted access to the configuration CSRs.
 	 */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
+	_mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
 		      0x0U);
 
 	for (index = 0U; index < mem_size / sizeof(uint32_t); index++) {
@@ -49,7 +65,7 @@ void ddrphy_phyinit_writeoutmem(uint32_t
 	 * 2. Isolate the APB access from the internal CSRs by setting the MicroContMuxSel CSR to 1.
 	 *    This allows the firmware unrestricted access to the configuration CSRs.
 	 */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
+	_mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
 		      0x1U);
 }
 
@@ -58,11 +74,15 @@ void ddrphy_phyinit_writeoutmsgblk(uint1
 {
 	uint32_t index;
 
+VERBOSE("%s@%d: Writing %u(%x) bytes from %p..%p to %08x..%08x\n", __func__, __LINE__,
+	mem_size, mem_size, mem, &mem[mem_size / sizeof(uint16_t)],
+	DDRPHYC_BASE + mem_offset, DDRPHYC_BASE + mem_offset + mem_size);
+VERBOSE("%s@%d: %04x %04x %04x %04x \n", __func__, __LINE__, mem[0], mem[1], mem[2], mem[3]);
 	/*
 	 * 1. Enable access to the internal CSRs by setting the MicroContMuxSel CSR to 0.
 	 *    This allows the memory controller unrestricted access to the configuration CSRs.
 	 */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
+	_mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
 		      0x0U);
 
 
@@ -74,6 +94,6 @@ void ddrphy_phyinit_writeoutmsgblk(uint1
 	 * 2. Isolate the APB access from the internal CSRs by setting the MicroContMuxSel CSR to 1.
 	 *    This allows the firmware unrestricted access to the configuration CSRs.
 	 */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
+	_mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4U * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR))),
 		      0x1U);
 }
