Index: tf-a-qsmp-2550/plat/st/stm32mp2/stm32mp2_def.h
===================================================================
--- tf-a-qsmp-2550.orig/plat/st/stm32mp2/stm32mp2_def.h
+++ tf-a-qsmp-2550/plat/st/stm32mp2/stm32mp2_def.h
@@ -341,21 +341,80 @@ enum ddr_type {
 
 /* For UART crash console */
 #define STM32MP_DEBUG_USART_CLK_FRQ	64000000
-/* USART2 on HSI@64MHz, TX on GPIOA4 Alternate 6 */
+
+#ifdef STM32MP_DEBUG_UART
+#if STM32MP_DEBUG_UART == 2
 #define STM32MP_DEBUG_USART_BASE	USART2_BASE
+/* USART2 on HSI@64MHz, TX on GPIOA4 Alternate 6 */
 #define DEBUG_UART_TX_GPIO_BANK_ADDRESS	GPIOA_BASE
 #define DEBUG_UART_TX_GPIO_BANK_CLK_REG	RCC_GPIOACFGR
-#define DEBUG_UART_TX_GPIO_BANK_CLK_EN	RCC_GPIOxCFGR_GPIOxEN
 #define DEBUG_UART_TX_GPIO_PORT		4
 #define DEBUG_UART_TX_GPIO_ALTERNATE	6
 #define DEBUG_UART_TX_CLKSRC_REG	RCC_XBAR8CFGR
-#define DEBUG_UART_TX_CLKSRC		XBAR_SRC_HSI
 #define DEBUG_UART_TX_EN_REG		RCC_USART2CFGR
-#define DEBUG_UART_TX_EN		RCC_UARTxCFGR_UARTxEN
 #define DEBUG_UART_RST_REG		RCC_USART2CFGR
+#elif STM32MP_DEBUG_UART == 4
+/* UART4 on HSI@64MHz, TX on GPIOB7 Alternate 3 */
+#define STM32MP_DEBUG_USART_BASE	UART4_BASE
+#define DEBUG_UART_TX_GPIO_BANK_ADDRESS	GPIOB_BASE
+#define DEBUG_UART_TX_GPIO_BANK_CLK_REG	RCC_GPIOBCFGR
+#define DEBUG_UART_TX_GPIO_PORT		7
+#define DEBUG_UART_TX_GPIO_ALTERNATE	3
+#define DEBUG_UART_TX_CLKSRC_REG	RCC_XBAR8CFGR
+#define DEBUG_UART_TX_EN_REG		RCC_UART4CFGR
+#define DEBUG_UART_RST_REG		RCC_UART4CFGR
+#define DEBUG_UART_PWR_REG		PWR_CR1
+#define DEBUG_UART_PWR_EN		PWR_CR1_VDDIO4SV
+#elif STM32MP_DEBUG_UART == 5
+/* UART5 on HSI@64MHz, TX on GPIOG9 Alternate 5 */
+#define STM32MP_DEBUG_USART_BASE	UART5_BASE
+#define DEBUG_UART_TX_GPIO_BANK_ADDRESS	GPIOG_BASE
+#define DEBUG_UART_TX_GPIO_BANK_CLK_REG	RCC_GPIOGCFGR
+#define DEBUG_UART_TX_GPIO_PORT		9
+#define DEBUG_UART_TX_GPIO_ALTERNATE	5
+#define DEBUG_UART_TX_CLKSRC_REG	RCC_XBAR9CFGR
+#define DEBUG_UART_TX_EN_REG		RCC_UART5CFGR
+#define DEBUG_UART_RST_REG		RCC_UART5CFGR
+#elif STM32MP_DEBUG_UART == 3
+/* USART3 on HSI@64MHz, TX on GPIOI6 Alternate 6 */
+#define STM32MP_DEBUG_USART_BASE	USART3_BASE
+#define DEBUG_UART_TX_GPIO_BANK_ADDRESS	GPIOI_BASE
+#define DEBUG_UART_TX_GPIO_BANK_CLK_REG	RCC_GPIOICFGR
+#define DEBUG_UART_TX_GPIO_PORT		6
+#define DEBUG_UART_TX_GPIO_ALTERNATE	6
+#define DEBUG_UART_TX_CLKSRC_REG	RCC_XBAR9CFGR
+#define DEBUG_UART_TX_EN_REG		RCC_USART3CFGR
+#define DEBUG_UART_RST_REG		RCC_USART3CFGR
+#elif STM32MP_DEBUG_UART == 7
+/* UART7 on HSI@64MHz, TX on GPIOD3 Alternate 6 */
+#define STM32MP_DEBUG_USART_BASE	UART7_BASE
+#define DEBUG_UART_TX_GPIO_BANK_ADDRESS	GPIOD_BASE
+#define DEBUG_UART_TX_GPIO_BANK_CLK_REG	RCC_GPIODCFGR
+#define DEBUG_UART_TX_GPIO_PORT		3
+#define DEBUG_UART_TX_GPIO_ALTERNATE	6
+#define DEBUG_UART_TX_CLKSRC_REG	RCC_XBAR21CFGR
+#define DEBUG_UART_TX_EN_REG		RCC_UART7CFGR
+#define DEBUG_UART_RST_REG		RCC_UART7CFGR
+#elif STM32MP_DEBUG_UART == 8
+/* UART8 on HSI@64MHz, TX on GPIOG3 Alternate 3 */
+#define STM32MP_DEBUG_USART_BASE	UART8_BASE
+#define DEBUG_UART_TX_GPIO_BANK_ADDRESS	GPIOG_BASE
+#define DEBUG_UART_TX_GPIO_BANK_CLK_REG	RCC_GPIOGCFGR
+#define DEBUG_UART_TX_GPIO_PORT		3
+#define DEBUG_UART_TX_GPIO_ALTERNATE	3
+#define DEBUG_UART_TX_CLKSRC_REG	RCC_XBAR21CFGR
+#define DEBUG_UART_TX_EN_REG		RCC_UART8CFGR
+#define DEBUG_UART_RST_REG		RCC_UART8CFGR
+#else
+#error Unsupported DEBUG UART
+#endif
+#define DEBUG_UART_TX_GPIO_BANK_CLK_EN	RCC_GPIOxCFGR_GPIOxEN
+#define DEBUG_UART_TX_CLKSRC		XBAR_SRC_HSI
 #define DEBUG_UART_RST_BIT		RCC_UARTxCFGR_UARTxRST
+#define DEBUG_UART_TX_EN		RCC_UARTxCFGR_UARTxEN
 #define DEBUG_UART_PREDIV_CFGR		RCC_PREDIV8CFGR
 #define DEBUG_UART_FINDIV_CFGR		RCC_FINDIV8CFGR
+#endif
 
 /*******************************************************************************
  * STM32MP2 SDMMC
Index: tf-a-qsmp-2550/plat/st/stm32mp2/aarch64/stm32mp2_helper.S
===================================================================
--- tf-a-qsmp-2550.orig/plat/st/stm32mp2/aarch64/stm32mp2_helper.S
+++ tf-a-qsmp-2550/plat/st/stm32mp2/aarch64/stm32mp2_helper.S
@@ -140,6 +140,12 @@ func plat_crash_console_init
 	orr	w2, w2, w0, lsl (DEBUG_UART_TX_GPIO_PORT << 2)
 	str	w2, [x1, #GPIO_AFRL_OFFSET]
 #endif
+#ifdef DEBUG_UART_PWR_REG
+	mov_imm	x1, PWR_BASE
+	ldr	w2, [x1, #DEBUG_UART_PWR_REG]
+	orr	w2, w2, #DEBUG_UART_PWR_EN
+	str	w2, [x1, #DEBUG_UART_PWR_REG]
+#endif
 	/* Clear UART clock flexgen divisors, keep enable bit */
 	mov_imm	x1, (RCC_BASE + DEBUG_UART_PREDIV_CFGR)
 	mov	x2, #0
Index: tf-a-qsmp-2550/plat/st/stm32mp2/platform.mk
===================================================================
--- tf-a-qsmp-2550.orig/plat/st/stm32mp2/platform.mk
+++ tf-a-qsmp-2550/plat/st/stm32mp2/platform.mk
@@ -181,6 +181,9 @@ PLAT_INCLUDES		+=	-Iplat/st/stm32mp2/inc
 PLAT_INCLUDES		+=	-Idrivers/st/ddr/phy/phyinit/include/
 PLAT_INCLUDES		+=	-Idrivers/st/ddr/phy/firmware/include/
 
+ifneq ($(STM32MP_DEBUG_UART),)
+STM32MP_CPPFLAGS	+=	-DSTM32MP_DEBUG_UART=$(STM32MP_DEBUG_UART)
+endif
 ifneq ($(STM32MP_BL31_SIZE),)
 STM32MP_CPPFLAGS	+=	-DSTM32MP_BL31_SIZE=$(STM32MP_BL31_SIZE)
 endif
