Index: tf-a-txmp-2550/fdts/stm32mp25-bl2.dtsi
===================================================================
--- tf-a-txmp-2550.orig/fdts/stm32mp25-bl2.dtsi
+++ tf-a-txmp-2550/fdts/stm32mp25-bl2.dtsi
@@ -5,40 +5,25 @@
 
 #include <platform_def.h>
 
-/omit-if-no-ref/ &i2c7_pins_a;
-/omit-if-no-ref/ &ospi_port1_clk_pins_a;
-/omit-if-no-ref/ &ospi_port1_cs0_pins_a;
-/omit-if-no-ref/ &ospi_port1_io03_pins_a;
-/omit-if-no-ref/ &sdmmc1_b4_pins_a;
-/omit-if-no-ref/ &sdmmc2_b4_pins_a;
-/omit-if-no-ref/ &sdmmc2_d47_pins_a;
-/omit-if-no-ref/ &usart2_pins_a;
-
-/ {
-	soc@0 {
 #if !STM32MP_SPI_NAND && !STM32MP_SPI_NOR && !STM32MP_HYPERFLASH
-		/delete-node/ ommanager@40500000;
+/delete-node/ &ommanager;
 #else
-		ommanager@40500000 {
 #if STM32MP_HYPERFLASH
-			/delete-node/ spi@40430000;
+/delete-node/ &ospi1;
 #endif
-			/delete-node/ spi@40440000;
-		};
+/delete-node/ &ospi2;
 #endif
 #if !STM32MP_RAW_NAND
-		/delete-node/ memory-controller@48200000;
+/delete-node/ &fmc;
 #endif
 #if !STM32MP_EMMC && !STM32MP_SDMMC
-		rifsc@42080000 {
-			/delete-node/ mmc@48220000;
-			/delete-node/ mmc@48230000;
-		};
+/delete-node/ &sdmmc1;
+/delete-node/ &sdmmc2;
 #endif
-	};
 
+/ {
 	/*
-	 * UUID's here are UUID RFC 4122 compliant meaning fieds are stored in
+	 * UUID's here are UUID RFC 4122 compliant meaning fields are stored in
 	 * network order (big endian)
 	 */
 
@@ -84,5 +69,4 @@
 
 #include "stm32mp2-cot-descriptors.dtsi"
 #endif
-
 };
Index: tf-a-txmp-2550/fdts/stm32mp25-bl31.dtsi
===================================================================
--- tf-a-txmp-2550.orig/fdts/stm32mp25-bl31.dtsi
+++ tf-a-txmp-2550/fdts/stm32mp25-bl31.dtsi
@@ -3,22 +3,7 @@
  * Copyright (C) STMicroelectronics 2023 - All Rights Reserved
  */
 
-/omit-if-no-ref/ &i2c7_pins_a;
-/omit-if-no-ref/ &ospi_port1_clk_pins_a;
-/omit-if-no-ref/ &ospi_port1_cs0_pins_a;
-/omit-if-no-ref/ &ospi_port1_io03_pins_a;
-/omit-if-no-ref/ &sdmmc1_b4_pins_a;
-/omit-if-no-ref/ &sdmmc2_b4_pins_a;
-/omit-if-no-ref/ &sdmmc2_d47_pins_a;
-/omit-if-no-ref/ &usart2_pins_a;
-
-/ {
-	soc@0 {
-		/delete-node/ ommanager@40500000;
-		/delete-node/ memory-controller@48200000;
-		rifsc@42080000 {
-			/delete-node/ mmc@48220000;
-			/delete-node/ mmc@48230000;
-		};
-	};
-};
+/delete-node/ &ommanager;
+/delete-node/ &fmc;
+/delete-node/ &sdmmc1;
+/delete-node/ &sdmmc2;
