Index: tf-a-txmp-2550/fdts/stm32mp251.dtsi
===================================================================
--- tf-a-txmp-2550.orig/fdts/stm32mp251.dtsi
+++ tf-a-txmp-2550/fdts/stm32mp251.dtsi
@@ -139,6 +139,8 @@
 				reg = <0 0 0x400>;
 				clocks = <&rcc CK_KER_OSPI1>;
 				resets = <&rcc OSPI1_R>, <&rcc OSPI1DLL_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -147,6 +149,8 @@
 				reg = <1 0 0x400>;
 				clocks = <&rcc CK_KER_OSPI2>;
 				resets = <&rcc OSPI2_R>, <&rcc OSPI2DLL_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 		};
@@ -194,6 +198,8 @@
 				reg = <0x40120000 0x400>;
 				clocks = <&rcc CK_KER_I2C1>;
 				resets = <&rcc I2C1_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -202,6 +208,8 @@
 				reg = <0x40130000 0x400>;
 				clocks = <&rcc CK_KER_I2C2>;
 				resets = <&rcc I2C2_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -210,6 +218,8 @@
 				reg = <0x40140000 0x400>;
 				clocks = <&rcc CK_KER_I2C3>;
 				resets = <&rcc I2C3_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -218,6 +228,8 @@
 				reg = <0x40150000 0x400>;
 				clocks = <&rcc CK_KER_I2C4>;
 				resets = <&rcc I2C4_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -226,6 +238,8 @@
 				reg = <0x40160000 0x400>;
 				clocks = <&rcc CK_KER_I2C5>;
 				resets = <&rcc I2C5_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -234,6 +248,8 @@
 				reg = <0x40170000 0x400>;
 				clocks = <&rcc CK_KER_I2C6>;
 				resets = <&rcc I2C6_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -242,6 +258,8 @@
 				reg = <0x40180000 0x400>;
 				clocks = <&rcc CK_KER_I2C7>;
 				resets = <&rcc I2C7_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -314,6 +332,8 @@
 				reg = <0x46040000 0x400>;
 				clocks = <&rcc CK_KER_I2C8>;
 				resets = <&rcc I2C8_R>;
+				#address-cells = <1>;
+				#size-cells = <0>;
 				status = "disabled";
 			};
 
@@ -364,33 +384,42 @@
 			#address-cells = <1>;
 			#size-cells = <1>;
 
-			uid_otp@14 {
+			uid-otp@14 {
 				reg = <0x14 0xc>;
 			};
-			part_number_otp@24 {
+
+			part-number-otp@24 {
 				reg = <0x24 0x4>;
 			};
+
 			nand_otp: otp16@40 {
 				reg = <0x40 0x4>;
 			};
+
 			lifecycle2_otp: otp18@48 {
 				reg = <0x48 0x4>;
 			};
+
 			nand2_otp: otp20@50 {
 				reg = <0x50 0x4>;
 			};
-			rev_otp@198 {
+
+			rev-otp@198 {
 				reg = <0x198 0x4>;
 			};
-			package_otp@1e8 {
+
+			package-otp@1e8 {
 				reg = <0x1e8 0x1>;
 			};
+
 			hconf1_otp: otp124@1f0 {
 				reg = <0x1f0 0x4>;
 			};
+
 			pkh_otp: otp144@240 {
 				reg = <0x240 0x20>;
 			};
+
 			oem_fip_enc_key: otp260@410 {
 				reg = <0x410 0x20>;
 			};
Index: tf-a-txmp-2550/lib/fconf/fconf_dyn_cfg_getter.c
===================================================================
--- tf-a-txmp-2550.orig/lib/fconf/fconf_dyn_cfg_getter.c
+++ tf-a-txmp-2550/lib/fconf/fconf_dyn_cfg_getter.c
@@ -112,20 +112,23 @@ int fconf_populate_dtb_registry(uintptr_
 		/* Read configuration dtb information */
 		rc = fdt_read_uint64(dtb, child, "load-address", &val64);
 		if (rc < 0) {
-			ERROR("FCONF: Incomplete configuration property in dtb-registry.\n");
+			ERROR("FCONF: 'load-address' property is missing in %s.\n",
+			      fdt_get_name(dtb, child, NULL));
 			return rc;
 		}
 		config_addr = (uintptr_t)val64;
 
 		rc = fdt_read_uint32(dtb, child, "max-size", &config_max_size);
 		if (rc < 0) {
-			ERROR("FCONF: Incomplete configuration property in dtb-registry.\n");
+			ERROR("FCONF: 'max-size' property is missing in %s.\n",
+			      fdt_get_name(dtb, child, NULL));
 			return rc;
 		}
 
 		rc = fdt_read_uint32(dtb, child, "id", &config_id);
 		if (rc < 0) {
-			ERROR("FCONF: Incomplete configuration property in dtb-registry.\n");
+			ERROR("FCONF: 'id' property is missing in %s.\n",
+			      fdt_get_name(dtb, child, NULL));
 			return rc;
 		}
 
Index: tf-a-txmp-2550/plat/st/stm32mp2/stm32mp2_def.h
===================================================================
--- tf-a-txmp-2550.orig/plat/st/stm32mp2/stm32mp2_def.h
+++ tf-a-txmp-2550/plat/st/stm32mp2/stm32mp2_def.h
@@ -157,7 +157,7 @@
 #define RETRAM_SIZE			U(0x00020000)
 #define STM32MP_BACKUP_RAM_BASE		U(0x42000000)
 
-/* the first 4KB of SRAM1 are reserved are for BSEC shadow */
+/* the first 4KiB of SRAM1 are reserved are for BSEC shadow */
 #define STM32MP_SEC_SRAM1_SIZE		U(0x1000)
 
 #define STM32MP_SEC_SYSRAM_BASE		STM32MP_SYSRAM_BASE
@@ -193,7 +193,7 @@ enum ddr_type {
 #endif
 
 /* Section used inside TF binaries */
-#define STM32MP_PARAM_LOAD_SIZE			U(0x00002400) /* 9 KB for param */
+#define STM32MP_PARAM_LOAD_SIZE			U(0x00002400)	/* 9 KiB for param */
 #if STM32MP21
 /* 576 Bytes reserved for header */
 #define STM32MP_HEADER_SIZE			U(0x00000240)
@@ -217,25 +217,25 @@ enum ddr_type {
 
 #if TRUSTED_BOARD_BOOT
 #if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-#define STM32MP_BL2_RO_SIZE		U(0x00021000)	/* 132 KB */
-#define STM32MP_BL2_SIZE		U(0x0002C000)	/* 176 KB for BL2 */
+#define STM32MP_BL2_RO_SIZE			U(0x00021000)	/* 132 KiB */
+#define STM32MP_BL2_SIZE			U(0x0002C000)	/* 176 KiB for BL2 */
 #else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
-#define STM32MP_BL2_RO_SIZE		U(0x00021000)	/* 132 KB */
-#define STM32MP_BL2_SIZE		U(0x0002C000)	/* 176 KB for BL2 */
+#define STM32MP_BL2_RO_SIZE			U(0x00021000)	/* 132 KiB */
+#define STM32MP_BL2_SIZE			U(0x0002C000)	/* 176 KiB for BL2 */
 #endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 #else
 #if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-#define STM32MP_BL2_RO_SIZE		U(0x0001F000)	/* 124 KB */
-#define STM32MP_BL2_SIZE		U(0x00028000)	/* 160 KB for BL2 */
+#define STM32MP_BL2_RO_SIZE			U(0x0001F000)	/* 124 KiB */
+#define STM32MP_BL2_SIZE			U(0x00028000)	/* 160 KiB for BL2 */
 #else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
-#define STM32MP_BL2_RO_SIZE		U(0x00020000)	/* 128 KB */
-#define STM32MP_BL2_SIZE		U(0x00029000)	/* 164 KB for BL2 */
+#define STM32MP_BL2_RO_SIZE			U(0x00020000)	/* 128 KiB */
+#define STM32MP_BL2_SIZE			U(0x00029000)	/* 164 KiB for BL2 */
 #endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
-#endif
+#endif /* TRUSTED_BOARD_BOOT */
 
 /* Allocate remaining sysram to BL31 Binary only */
-#define STM32MP_BL31_SIZE		(STM32MP_SEC_SYSRAM_SIZE - \
-					 STM32MP_BL2_SIZE)
+#define STM32MP_BL31_SIZE			(STM32MP_SEC_SYSRAM_SIZE - \
+						 STM32MP_BL2_SIZE)
 
 #define BL31_PROGBITS_LIMIT		STM32MP_BL31_SIZE
 
@@ -257,7 +257,7 @@ enum ddr_type {
 					 STM32MP_MBEDTLS_HEAP_SIZE)
 
 /* BL2 and BL32/sp_min require 4 tables */
-#define MAX_XLAT_TABLES				U(4)	/* 16 KB for mapping */
+#define MAX_XLAT_TABLES				U(4)	/* 16 KiB for mapping */
 
 /*
  * MAX_MMAP_REGIONS is usually:
@@ -270,7 +270,7 @@ enum ddr_type {
 #endif
 
 /* DTB initialization value */
-#define STM32MP_BL2_DTB_SIZE			U(0x00006000)	/* 24 KB for DTB */
+#define STM32MP_BL2_DTB_SIZE			U(0x00006000)	/* 24 KiB for DTB */
 
 #define STM32MP_BL2_DTB_BASE			(STM32MP_BL2_BASE - \
 						 STM32MP_BL2_DTB_SIZE)
@@ -463,14 +463,14 @@ enum ddr_type {
 #define STM32MP2_UPPER_OTP_START		0x100U
 
 /* OTP labels */
-#define PART_NUMBER_OTP			"part_number_otp"
-#define REVISION_OTP			"rev_otp"
-#define PACKAGE_OTP			"package_otp"
+#define PART_NUMBER_OTP			"part-number-otp"
+#define REVISION_OTP			"rev-otp"
+#define PACKAGE_OTP			"package-otp"
 #define HCONF1_OTP			"otp124"
 #define NAND_OTP			"otp16"
 #define NAND2_OTP			"otp20"
-#define BOARD_ID_OTP			"board_id"
-#define UID_OTP				"uid_otp"
+#define BOARD_ID_OTP			"board-id"
+#define UID_OTP				"uid-otp"
 #define LIFECYCLE2_OTP			"otp18"
 #if STM32MP21
 #define PKH_OTP				"otp152"
@@ -478,6 +478,12 @@ enum ddr_type {
 #define PKH_OTP				"otp144"
 #endif
 #define ENCKEY_OTP			"otp260"
+/* legacy definitions */
+#define PART_NUMBER_OTP_LEGACY		"part_number_otp"
+#define REVISION_OTP_LEGACY		"rev_otp"
+#define PACKAGE_OTP_LEGACY		"package_otp"
+#define BOARD_ID_OTP_LEGACY		"board_id"
+#define UID_OTP_LEGACY			"uid_otp"
 
 /* OTP mask */
 /* PACKAGE */
Index: tf-a-txmp-2550/plat/st/stm32mp2/platform.mk
===================================================================
--- tf-a-txmp-2550.orig/plat/st/stm32mp2/platform.mk
+++ tf-a-txmp-2550/plat/st/stm32mp2/platform.mk
@@ -265,9 +265,7 @@ PLAT_BL_COMMON_SOURCES		+=	drivers/st/cl
 
 PLAT_BL_COMMON_SOURCES  	+=	drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c
 
-ifeq ($(filter 1,${STM32MP_UART_PROGRAMMER} ${STM32MP_USB_PROGRAMMER}),)
 PLAT_BL_COMMON_SOURCES  	+=	plat/st/stm32mp2/stm32mp2_context.c
-endif
 
 BL2_SOURCES			+=	plat/st/stm32mp2/plat_bl2_mem_params_desc.c
 
@@ -352,9 +350,7 @@ BL31_SOURCES			+=	${FDT_WRAPPERS_SOURCES
 BL31_SOURCES			+=	plat/st/stm32mp2/bl31_plat_setup.c			\
 					plat/st/stm32mp2/stm32mp2_topology.c
 
-ifeq ($(filter 1,${STM32MP_UART_PROGRAMMER} ${STM32MP_USB_PROGRAMMER}),)
 BL31_SOURCES			+=	plat/st/stm32mp2/stm32mp2_pm.c
-endif
 
 BL31_SOURCES			+=	drivers/st/ddr/stm32mp2_ddr_helpers.c
 
Index: tf-a-txmp-2550/plat/st/stm32mp2/bl31_plat_setup.c
===================================================================
--- tf-a-txmp-2550.orig/plat/st/stm32mp2/bl31_plat_setup.c
+++ tf-a-txmp-2550/plat/st/stm32mp2/bl31_plat_setup.c
@@ -185,22 +185,3 @@ entry_point_info_t *bl31_plat_get_next_i
 
 	return NULL;
 }
-
-#if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-static const plat_psci_ops_t null_psci_ops;
-
-/* Stub PSCI platform functions */
-plat_local_state_t plat_get_target_pwr_state(unsigned int lvl,
-					     const plat_local_state_t *states,
-					     unsigned int ncpu)
-{
-	return 0U;
-}
-
-int plat_setup_psci_ops(uintptr_t sec_entrypoint,
-			const plat_psci_ops_t **psci_ops)
-{
-	*psci_ops = &null_psci_ops;
-	return 0;
-}
-#endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
Index: tf-a-txmp-2550/drivers/st/nvmem/nvmem.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/nvmem/nvmem.c
+++ tf-a-txmp-2550/drivers/st/nvmem/nvmem.c
@@ -25,13 +25,11 @@ static struct nvmem_dev nvmem_dev_array[
 int nvmem_driver_register(int nodeoffset, const struct nvmem_ops *nvmem_ops,
 			  void *drv_data)
 {
-	struct nvmem_dev *nvmem_dev = NULL;
+	struct nvmem_dev *nvmem_dev;
 
-	for_each_nvmem_dev(nvmem_dev) {
-		if (nvmem_dev->drv_data == NULL) {
+	for_each_nvmem_dev(nvmem_dev)
+		if (!nvmem_dev->drv_data)
 			break;
-		}
-	}
 
 	if (nvmem_dev > &nvmem_dev_array[PLAT_NB_NVMEM_DEVS - 1U]) {
 		WARN("Not enough place for NVMEM dev, PLAT_NB_NVMEM_DEVS should be increased.\n");
@@ -47,13 +45,11 @@ int nvmem_driver_register(int nodeoffset
 
 static struct nvmem_dev *nvmem_get_by_node(int node_offset)
 {
-	struct nvmem_dev *nvmem_dev = NULL;
+	struct nvmem_dev *nvmem_dev;
 
-	for_each_nvmem_dev(nvmem_dev) {
-		if (nvmem_dev->offset == node_offset) {
+	for_each_nvmem_dev(nvmem_dev)
+		if (nvmem_dev->offset == node_offset)
 			return nvmem_dev;
-		}
-	}
 
 	WARN("%s: node %d not found\n", __func__, node_offset);
 	return NULL;
@@ -62,12 +58,12 @@ static struct nvmem_dev *nvmem_get_by_no
 int nvmem_get_cell_by_index(const void *fdt, int nodeoffset, unsigned int index,
 			    struct nvmem_cell *cell)
 {
-	int len = 0;
-	const fdt32_t *nvmem_cells_prop = NULL, *nvmem_cell_reg_prop = NULL;
+	int len;
+	const fdt32_t *nvmem_cells_prop, *nvmem_cell_reg_prop;
 	uint32_t phandle;
 	int nvmem_cell_node;
 	int parent_node;
-	struct nvmem_dev *dev = NULL;
+	struct nvmem_dev *dev;
 
 	nvmem_cells_prop = fdt_getprop(fdt, nodeoffset, "nvmem-cells", &len);
 	if (nvmem_cells_prop == NULL) {
@@ -80,10 +76,10 @@ int nvmem_get_cell_by_index(const void *
 
 	nvmem_cell_reg_prop = fdt_getprop(fdt, nvmem_cell_node, "reg", &len);
 
-	if ((nvmem_cell_reg_prop == NULL) ||
-	    ((unsigned int)len != (sizeof(uint32_t) * U(2)))) {
+	if (!nvmem_cell_reg_prop)
+		return len;
+	if ((unsigned int)len != (sizeof(uint32_t) * U(2)))
 		return -FDT_ERR_NOTFOUND;
-	}
 
 	cell->offset = fdt32_to_cpu(nvmem_cell_reg_prop[0]);
 	cell->len = fdt32_to_cpu(nvmem_cell_reg_prop[1]);
@@ -103,9 +99,8 @@ int nvmem_get_cell_by_name(const void *f
 {
 	int index = fdt_stringlist_search(fdt, nodeoffset, "nvmem-cell-names", name);
 
-	if (index < 0) {
-		return -FDT_ERR_NOTFOUND;
-	}
+	if (index < 0)
+		return index;
 
 	return nvmem_get_cell_by_index(fdt, nodeoffset, (unsigned int)index,
 				       cell);
@@ -114,9 +109,8 @@ int nvmem_get_cell_by_name(const void *f
 int nvmem_cell_read(struct nvmem_cell *cell, uint8_t *output_buf,
 		    size_t buf_len, size_t *read_len)
 {
-	if (cell->dev->nvmem_ops->read_cell == NULL) {
+	if (!cell->dev->nvmem_ops->read_cell)
 		return -ENOTSUP;
-	}
 
 	return cell->dev->nvmem_ops->read_cell(cell, output_buf, buf_len,
 					       read_len);
@@ -125,9 +119,8 @@ int nvmem_cell_read(struct nvmem_cell *c
 int nvmem_cell_write(struct nvmem_cell *cell, uint8_t *input_buf,
 		     size_t buf_len)
 {
-	if (cell->dev->nvmem_ops->write_cell == NULL) {
+	if (!cell->dev->nvmem_ops->write_cell)
 		return -ENOTSUP;
-	}
 
 	return cell->dev->nvmem_ops->write_cell(cell, input_buf, buf_len);
 }
Index: tf-a-txmp-2550/plat/st/common/stm32mp_common.c
===================================================================
--- tf-a-txmp-2550.orig/plat/st/common/stm32mp_common.c
+++ tf-a-txmp-2550/plat/st/common/stm32mp_common.c
@@ -180,32 +180,33 @@ int stm32_get_otp_index(const char *otp_
 
 int stm32_get_otp_value(const char *otp_name, uint32_t *otp_val)
 {
+	int ret;
 	uint32_t otp_idx;
 
 	assert(otp_name != NULL);
 	assert(otp_val != NULL);
 
-	if (stm32_get_otp_index(otp_name, &otp_idx, NULL) != 0) {
-		return -1;
+	ret = stm32_get_otp_index(otp_name, &otp_idx, NULL);
+	if (ret) {
+		return ret;
 	}
 
-	if (stm32_get_otp_value_from_idx(otp_idx, otp_val) != 0) {
+	ret = stm32_get_otp_value_from_idx(otp_idx, otp_val);
+	if (ret != BSEC_OK)
 		ERROR("BSEC: %s Read Error\n", otp_name);
-		return -1;
-	}
 
-	return 0;
+	return ret;
 }
 
 int stm32_get_otp_value_from_idx(const uint32_t otp_idx, uint32_t *otp_val)
 {
-	uint32_t ret = BSEC_NOT_SUPPORTED;
+	uint32_t ret;
 
 	assert(otp_val != NULL);
 
-        if (otp_mirror_read(otp_val, otp_idx) == BSEC_OK) {
-                return 0;
-        }
+	if (otp_mirror_read(otp_val, otp_idx) == BSEC_OK) {
+		return 0;
+	}
 
 #if defined(IMAGE_BL2)
 	ret = stm32_otp_shadow_read(otp_val, otp_idx);
@@ -216,7 +217,7 @@ int stm32_get_otp_value_from_idx(const u
 #endif
 	if (ret != BSEC_OK) {
 		ERROR("BSEC: idx=%u Read Error\n", otp_idx);
-		return -1;
+		return BSEC_ERROR;
 	}
 
 	return 0;
@@ -344,16 +345,14 @@ int stm32mp_uart_console_setup(void)
 
 	result = dt_get_stdout_uart_info(&dt_uart_info);
 
-	if ((result <= 0) ||
-	    (dt_uart_info.status == DT_DISABLED)) {
+	if (result <= 0 ||
+	    dt_uart_info.status == DT_DISABLED)
 		return -ENODEV;
-	}
 
 #if defined(IMAGE_BL2)
-	if ((dt_uart_info.clock < 0) ||
-	    (dt_uart_info.reset < 0)) {
+	if (dt_uart_info.clock < 0 ||
+	    dt_uart_info.reset < 0)
 		return -ENODEV;
-	}
 #endif
 
 #if (STM32MP_UART_PROGRAMMER || !defined(IMAGE_BL2)) && !STM32MP_SSP
@@ -366,9 +365,8 @@ int stm32mp_uart_console_setup(void)
 #endif
 
 #if defined(IMAGE_BL2)
-	if (dt_set_stdout_pinctrl() != 0) {
+	if (dt_set_stdout_pinctrl())
 		return -ENODEV;
-	}
 
 	clk_enable((unsigned long)dt_uart_info.clock);
 
Index: tf-a-txmp-2550/plat/st/common/stm32mp_dt.c
===================================================================
--- tf-a-txmp-2550.orig/plat/st/common/stm32mp_dt.c
+++ tf-a-txmp-2550/plat/st/common/stm32mp_dt.c
@@ -6,15 +6,14 @@
 
 #include <assert.h>
 #include <errno.h>
-
+#include <libfdt.h>
+#include <platform_def.h>
+#include <stm32mp_dt.h>
+#include <string.h>
 #include <common/debug.h>
 #include <common/fdt_wrappers.h>
 #include <drivers/st/regulator.h>
 #include <drivers/st/stm32_gpio.h>
-#include <libfdt.h>
-
-#include <platform_def.h>
-#include <stm32mp_dt.h>
 
 static void *fdt;
 
@@ -27,9 +26,8 @@ int dt_open_and_check(uintptr_t dt_addr)
 	int ret;
 
 	ret = fdt_check_header((void *)dt_addr);
-	if (ret == 0) {
+	if (ret == 0)
 		fdt = (void *)dt_addr;
-	}
 
 	return ret;
 }
@@ -41,9 +39,8 @@ int dt_open_and_check(uintptr_t dt_addr)
  ******************************************************************************/
 int fdt_get_address(void **fdt_addr)
 {
-	if (fdt == NULL) {
+	if (fdt == NULL)
 		return 0;
-	}
 
 	*fdt_addr = fdt;
 
@@ -343,6 +340,26 @@ const char *dt_get_board_model(void)
 	return (const char *)fdt_getprop(fdt, node, "model", NULL);
 }
 
+static int find_legacy_node_name(const char *name, int node)
+{
+	int ret;
+	char legacy_name[42];
+	char *l;
+
+	if (strlen(name) >= sizeof(legacy_name))
+		return -FDT_ERR_NOTFOUND;
+
+	strlcpy(legacy_name, name, sizeof(legacy_name));
+	for (l = legacy_name; *l != '\0'; l++)
+		if (*l == '-')
+			*l = '_';
+	ret = fdt_subnode_offset(fdt, node, legacy_name);
+	if (ret == 0)
+		WARN("Found legacy property '%s' in DTB; please convert to: '%s'\n",
+		     legacy_name, name);
+	return ret;
+}
+
 /*******************************************************************************
  * dt_find_otp_name: get OTP ID and length in DT.
  * name: sub-node name to look up.
@@ -352,24 +369,27 @@ const char *dt_get_board_model(void)
  ******************************************************************************/
 int dt_find_otp_name(const char *name, uint32_t *otp, uint32_t *otp_len)
 {
+	int ret;
 	int node;
 	int len;
 	const fdt32_t *cuint;
 
-	if ((name == NULL) || (otp == NULL)) {
+	if ((name == NULL) || (otp == NULL))
 		return -FDT_ERR_BADVALUE;
-	}
 
 	node = fdt_node_offset_by_compatible(fdt, -1, DT_BSEC_COMPAT);
-	if (node < 0) {
+	if (node < 0)
 		return node;
-	}
 
-	node = fdt_subnode_offset(fdt, node, name);
-	if (node < 0) {
-		ERROR("nvmem node %s not found\n", name);
-		return node;
+	ret = fdt_subnode_offset(fdt, node, name);
+	if (ret < 0)
+		ret = find_legacy_node_name(name, node);
+
+	if (ret < 0) {
+		INFO("nvmem node %s not found\n", name);
+		return ret;
 	}
+	node = ret;
 
 	cuint = fdt_getprop(fdt, node, "reg", &len);
 	if ((cuint == NULL) || (len != (2 * (int)sizeof(uint32_t)))) {
Index: tf-a-txmp-2550/plat/st/stm32mp2/stm32mp2_private.c
===================================================================
--- tf-a-txmp-2550.orig/plat/st/stm32mp2/stm32mp2_private.c
+++ tf-a-txmp-2550/plat/st/stm32mp2/stm32mp2_private.c
@@ -170,9 +170,9 @@ static const uintptr_t stm32mp2_uart_add
 
 uintptr_t get_uart_address(uint32_t instance_nb)
 {
-	if ((instance_nb == 0U) ||
-	    (instance_nb > STM32MP_NB_OF_UART)) {
-		return 0U;
+	if (instance_nb == 0 ||
+	    instance_nb > STM32MP_NB_OF_UART) {
+		return 0;
 	}
 
 	return stm32mp2_uart_addresses[instance_nb - 1U];
@@ -199,13 +199,11 @@ uint32_t stm32mp_get_chip_version(void)
 {
 	static uint32_t rev;
 
-	if (rev != 0U) {
+	if (rev)
 		return rev;
-	}
 
-	if (stm32_get_otp_value(REVISION_OTP, &rev) != 0) {
+	if (stm32_get_otp_value(REVISION_OTP, &rev))
 		panic();
-	}
 
 	return rev;
 }
@@ -219,13 +217,11 @@ static uint32_t get_part_number(void)
 {
 	static uint32_t part_number;
 
-	if (part_number != 0U) {
+	if (part_number)
 		return part_number;
-	}
 
-	if (stm32_get_otp_value(PART_NUMBER_OTP, &part_number) != 0) {
+	if (stm32_get_otp_value(PART_NUMBER_OTP, &part_number))
 		panic();
-	}
 
 	return part_number;
 }
@@ -235,7 +231,7 @@ static uint32_t get_cpu_package(void)
 	static uint32_t package = UINT32_MAX;
 
 	if (package == UINT32_MAX) {
-		if (stm32_get_otp_value(PACKAGE_OTP, &package) != 0) {
+		if (stm32_get_otp_value(PACKAGE_OTP, &package)) {
 			panic();
 		}
 	}
@@ -468,15 +464,13 @@ void stm32mp_print_cpuinfo(void)
 
 void stm32mp_print_boardinfo(void)
 {
-	uint32_t board_id = 0U;
+	uint32_t board_id;
 
-	if (stm32_get_otp_value(BOARD_ID_OTP, &board_id) != 0) {
+	if (stm32_get_otp_value(BOARD_ID_OTP, &board_id))
 		return;
-	}
 
-	if (board_id != 0U) {
+	if (board_id)
 		stm32_display_board_info(board_id);
-	}
 }
 
 /* Return true when SoC provides a single Cortex-A35 core, and false otherwise */
@@ -516,35 +510,29 @@ uint32_t stm32mp_check_closed_device(voi
 	static uint32_t otp_idx;
 	static uint32_t otp_len;
 	uint32_t otp_val;
-	uint32_t status = STM32MP_CHIP_SEC_CLOSED;
 
-	if (otp_len == 0U) {
-		if (stm32_get_otp_index(LIFECYCLE2_OTP, &otp_idx, &otp_len) != 0) {
+	if (!otp_len) {
+		if (stm32_get_otp_index(LIFECYCLE2_OTP, &otp_idx, &otp_len))
 			panic();
-		}
 
 		if (otp_len != __WORD_BIT) {
 			panic();
 		}
 	}
 
-	if (stm32_get_otp_value_from_idx(otp_idx, &otp_val) != 0) {
+	if (stm32_get_otp_value_from_idx(otp_idx, &otp_val))
 		panic();
-	}
 
 	if (!stm32_otp_is_closed_device() ||
-	    ((otp_val & SECURE_BOOT_CLOSED_SECURE) == 0U)) {
-		status = STM32MP_CHIP_SEC_OPEN;
-	}
+	    ((otp_val & SECURE_BOOT_CLOSED_SECURE) == 0))
+		return STM32MP_CHIP_SEC_OPEN;
 
-	return status;
+	return STM32MP_CHIP_SEC_CLOSED;
 }
 
 /* Return true when device supports secure boot */
 bool stm32mp_is_auth_supported(void)
 {
-	bool supported = false;
-
 	switch (get_part_number()) {
 #if STM32MP21
 	case STM32MP211C_PART_NB:
@@ -572,18 +560,15 @@ bool stm32mp_is_auth_supported(void)
 	case STM32MP257C_PART_NB:
 	case STM32MP257F_PART_NB:
 #endif /* STM32MP25 */
-		supported = true;
-		break;
+		return true;
 	default:
-		break;
+		return false;
 	}
-
-	return supported;
 }
 
 uint32_t stm32_iwdg_get_instance(uintptr_t base)
 {
-	uint32_t instance = UINT32_MAX;
+	uint32_t instance;
 
 	switch (base) {
 	case IWDG1_BASE:
@@ -593,10 +578,6 @@ uint32_t stm32_iwdg_get_instance(uintptr
 		instance = IWDG2_INST;
 		break;
 	default:
-		break;
-	}
-
-	if (instance == UINT32_MAX) {
 		panic();
 	}
 
@@ -605,14 +586,14 @@ uint32_t stm32_iwdg_get_instance(uintptr
 
 uint32_t stm32_iwdg_get_otp_config(uint32_t iwdg_inst)
 {
-	uint32_t iwdg_cfg = 0U;
+	uint32_t iwdg_cfg = 0;
 	uint32_t otp_value;
 
-	if (stm32_get_otp_value(HCONF1_OTP, &otp_value) != 0U) {
+	if (stm32_get_otp_value(HCONF1_OTP, &otp_value)) {
 		panic();
 	}
 
-	if ((otp_value & HCONF1_OTP_IWDG_HW_MASK(iwdg_inst)) != 0U) {
+	if ((otp_value & HCONF1_OTP_IWDG_HW_MASK(iwdg_inst))) {
 		iwdg_cfg |= IWDG_HW_ENABLED;
 	}
 
@@ -635,14 +616,13 @@ bool stm32mp_is_wakeup_from_standby(void
 	if (standby == -1) {
 		rstsr = mmio_read_32(stm32mp_rcc_base() + RCC_C1BOOTRSTSCLRR);
 
-		if ((rstsr & RCC_C1BOOTRSTSCLRR_PADRSTF) != 0U) {
+		if ((rstsr & RCC_C1BOOTRSTSCLRR_PADRSTF))
 			return false;
-		}
-		if (stm32_pm_context_is_valid()) {
+
+		if (stm32_pm_context_is_valid())
 			standby = 1;
-		} else {
+		else
 			standby = 0;
-		}
 	}
 
 	return standby == 1;
@@ -669,7 +649,7 @@ uintptr_t stm32_risaf_get_base(int insta
 	case RISAF4_INST:
 		return (uintptr_t)RISAF4_BASE;
 	default:
-		return 0U;
+		return 0;
 	}
 }
 
@@ -693,7 +673,7 @@ uintptr_t stm32_risaf_get_memory_base(in
 	case RISAF4_INST:
 		return (uintptr_t)STM32MP_DDR_BASE;
 	default:
-		return 0U;
+		return 0;
 	}
 }
 
@@ -705,7 +685,7 @@ size_t stm32_risaf_get_memory_size(int i
 	case RISAF4_INST:
 		return dt_get_ddr_size();
 	default:
-		return 0U;
+		return 0;
 	}
 }
 
Index: tf-a-txmp-2550/plat/st/stm32mp2/aarch64/stm32mp2_helper.S
===================================================================
--- tf-a-txmp-2550.orig/plat/st/stm32mp2/aarch64/stm32mp2_helper.S
+++ tf-a-txmp-2550/plat/st/stm32mp2/aarch64/stm32mp2_helper.S
@@ -126,17 +126,17 @@ func plat_crash_console_init
 	bic	w2, w2, #(GPIO_PULL_MASK << GPIO_TX_SHIFT)
 	str	w2, [x1, #GPIO_PUPD_OFFSET]
 	/* Set alternate */
+	mov	w0, #DEBUG_UART_TX_GPIO_ALTERNATE
 #if DEBUG_UART_TX_GPIO_PORT >= GPIO_ALT_LOWER_LIMIT
 	ldr	w2, [x1, #GPIO_AFRH_OFFSET]
 	bic	w2, w2, #(GPIO_ALTERNATE_MASK << \
 				((DEBUG_UART_TX_GPIO_PORT - GPIO_ALT_LOWER_LIMIT) << 2))
-	orr	w2, w2, #(DEBUG_UART_TX_GPIO_ALTERNATE << \
-				((DEBUG_UART_TX_GPIO_PORT - GPIO_ALT_LOWER_LIMIT) << 2))
+	orr	w2, w2, w0, lsl ((DEBUG_UART_TX_GPIO_PORT - GPIO_ALT_LOWER_LIMIT) << 2)
 	str	w2, [x1, #GPIO_AFRH_OFFSET]
 #else
 	ldr	w2, [x1, #GPIO_AFRL_OFFSET]
 	bic	w2, w2, #(GPIO_ALTERNATE_MASK << (DEBUG_UART_TX_GPIO_PORT << 2))
-	orr	w2, w2, #(DEBUG_UART_TX_GPIO_ALTERNATE << (DEBUG_UART_TX_GPIO_PORT << 2))
+	orr	w2, w2, w0, lsl (DEBUG_UART_TX_GPIO_PORT << 2)
 	str	w2, [x1, #GPIO_AFRL_OFFSET]
 #endif
 	/* Clear UART clock flexgen divisors, keep enable bit */
Index: tf-a-txmp-2550/drivers/st/usb_dwc3/usb_dwc3.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/usb_dwc3/usb_dwc3.c
+++ tf-a-txmp-2550/drivers/st/usb_dwc3/usb_dwc3.c
@@ -668,8 +668,8 @@ static enum usb_status dwc3_execute_dep_
 	uint64_t timeout;
 
 	VERBOSE("PHYEP%d: cmd '%s' %08x params %08x %08x %08x\n", phy_epnum,
-		   dwc3_get_ep_cmd_str(cmd), (uint32_t)cmd, (uint32_t)params->param0,
-		   (uint32_t)params->param1, (uint32_t)params->param2);
+		dwc3_get_ep_cmd_str(cmd), (uint32_t)cmd, (uint32_t)params->param0,
+		(uint32_t)params->param1, (uint32_t)params->param2);
 
 	DWC3_regwrite(dwc3_handle->usb_device, DWC3_DEPCMDPAR0(phy_epnum), params->param0);
 	DWC3_regwrite(dwc3_handle->usb_device, DWC3_DEPCMDPAR1(phy_epnum), params->param1);
@@ -683,7 +683,7 @@ static enum usb_status dwc3_execute_dep_
 		/* "Command Complete" */
 		if ((reg & USB3_DEPCMD_CMDACT) == 0U) {
 			VERBOSE("Command Complete --> %u PHYEP%u %x\n", DWC3_DEPCMD_STATUS(reg),
-				   phy_epnum, reg);
+				phy_epnum, reg);
 			if (DWC3_DEPCMD_STATUS(reg) != 0U) {
 				return USBD_FAIL;
 			}
@@ -773,10 +773,10 @@ static enum usb_status dwc3_ep_start_xfe
 	}
 
 	VERBOSE("EP%d%s: trb %p:%08x:%08x:%08x:%08x length %u %d\n",
-		   ep->num, ep->is_in ? "IN" : "OUT", dwc3_ep->trb_addr,
-		   (uint32_t)dwc3_ep->trb_addr->bph, (uint32_t)dwc3_ep->trb_addr->bpl,
-		   (uint32_t)dwc3_ep->trb_addr->size, (uint32_t)dwc3_ep->trb_addr->ctrl,
-		   ep->xfer_len, dwc3_ep->flags);
+		ep->num, ep->is_in ? "IN" : "OUT", dwc3_ep->trb_addr,
+		(uint32_t)dwc3_ep->trb_addr->bph, (uint32_t)dwc3_ep->trb_addr->bpl,
+		(uint32_t)dwc3_ep->trb_addr->size, (uint32_t)dwc3_ep->trb_addr->ctrl,
+		ep->xfer_len, dwc3_ep->flags);
 
 	ret = dwc3_execute_dep_cmd(dwc3_handle, dwc3_ep->phy_epnum, cmd, &params);
 	if (ret != USBD_OK) {
@@ -933,7 +933,7 @@ static enum usb_status dwc3_ep_stop_xfer
 
 	dwc3_ep->flags &= ~USB_DWC3_EP_REQ_QUEUED;
 
-        if ((!ep->is_in) && (ep->num == 0U)) {
+	if ((!ep->is_in) && (ep->num == 0U)) {
 		dwc3_handle->EP0_State = HAL_PCD_EP0_SETUP_COMPLETED;
 	}
 
@@ -1263,11 +1263,11 @@ static enum usb_action dwc3_handle_ep_ev
 	switch (type) {
 	case USB_DWC3_DEPEVT_XFERCOMPLETE:
 		VERBOSE("EP%d%s: Transfer Complete Event=%x\n", ep_num,
-			   ep->is_in ? "IN" : "OUT", event);
+			ep->is_in ? "IN" : "OUT", event);
 
 		if (ep->type == EP_TYPE_ISOC) {
 			VERBOSE("EP%d%s is an Isochronous endpoint\n", ep_num,
-				   ep->is_in ? "IN" : "OUT");
+				ep->is_in ? "IN" : "OUT");
 			break;
 		}
 
@@ -1277,14 +1277,14 @@ static enum usb_action dwc3_handle_ep_ev
 		if ((ep_num == 0U) &&
 		    (dwc3_get_ep_trbstatus(dwc3_ep) == USB_DWC3_TRBSTS_SETUP_PENDING)) {
 			VERBOSE("EP%d%s Setup Pending received\n", ep_num,
-				   ep->is_in ? "IN" : "OUT");
+				ep->is_in ? "IN" : "OUT");
 		}
 
 		VERBOSE("EP%d%s: Transfer Complete trb %p:%08x:%08x:%08x:%08x TRB-Status %u\n",
-			   ep_num, ep->is_in ? "IN" : "OUT", dwc3_ep->trb_addr,
-			   (uint32_t)dwc3_ep->trb_addr->bph, (uint32_t)dwc3_ep->trb_addr->bpl,
-			   (uint32_t)dwc3_ep->trb_addr->size, (uint32_t)dwc3_ep->trb_addr->ctrl,
-			   dwc3_get_ep_trbstatus(dwc3_ep));
+			ep_num, ep->is_in ? "IN" : "OUT", dwc3_ep->trb_addr,
+			(uint32_t)dwc3_ep->trb_addr->bph, (uint32_t)dwc3_ep->trb_addr->bpl,
+			(uint32_t)dwc3_ep->trb_addr->size, (uint32_t)dwc3_ep->trb_addr->ctrl,
+			dwc3_get_ep_trbstatus(dwc3_ep));
 
 		if ((ep_addr & EP_DIR_MASK) == EP_DIR_IN) {  /* IN EP */
 			ep->xfer_count = ep->xfer_len - dwc3_get_ep_trblen(dwc3_ep);
@@ -1341,8 +1341,8 @@ static enum usb_action dwc3_handle_ep_ev
 					   ep_num, ep->is_in ? "IN" : "OUT",
 					   dwc3_handle->setup_addr[0], dwc3_handle->setup_addr[1],
 					   dwc3_handle->setup_addr[2], dwc3_handle->setup_addr[3],
-					   dwc3_handle->setup_addr[4], dwc3_handle->setup_addr[5],
-					   dwc3_handle->setup_addr[6], dwc3_handle->setup_addr[7]);
+					dwc3_handle->setup_addr[4], dwc3_handle->setup_addr[5],
+					dwc3_handle->setup_addr[6], dwc3_handle->setup_addr[7]);
 
 				// if 2-stage then wait for the XFERNOTREADY(control-status)
 				if (__HAL_PCD_SETUP_REQ_LEN(dwc3_handle->setup_addr) != 0U) {
@@ -1391,8 +1391,8 @@ static enum usb_action dwc3_handle_ep_ev
 
 	case USB_DWC3_DEPEVT_XFERINPROGRESS:
 		VERBOSE("EP%d%s: Transfer In-Progress %u Event=%x\n", ep_num,
-			   ep->is_in ? "IN" : "OUT", dwc3_handle->intbuffers.evtbufferpos[0],
-			   event);
+			ep->is_in ? "IN" : "OUT", dwc3_handle->intbuffers.evtbufferpos[0],
+			event);
 
 		if (ep_num == 0U) { // ***ToCheck
 			break;
@@ -1415,7 +1415,7 @@ static enum usb_action dwc3_handle_ep_ev
 
 	case USB_DWC3_DEPEVT_XFERNOTREADY:
 		VERBOSE("EP%d%s: Transfer Not Ready Event=%x\n", ep_num,
-			   ep->is_in ? "IN" : "OUT", event);
+			ep->is_in ? "IN" : "OUT", event);
 		/* For EP0IN or EP0OUT, currently no plan to use xfernotready interrupt */
 
 		//active = PCD_READ_DEPEVT_STATUS(event);
@@ -1456,7 +1456,7 @@ static enum usb_action dwc3_handle_ep_ev
 		break;
 	case USB_DWC3_DEPEVT_EPCMDCMPLT: /* Endpoint Command Complete */
 		VERBOSE("EP%d%s: Endpoint Command Complete Event=%x\n", ep_num,
-			   ep->is_in ? "IN" : "OUT", event);
+			ep->is_in ? "IN" : "OUT", event);
 		break;
 	default:
 		VERBOSE("Invalid type %d: %s: %d\n", type, __func__, __LINE__);
@@ -1532,27 +1532,27 @@ static uint8_t dwc3_get_dev_speed(dwc3_h
 	switch (reg) {
 	case DWC3_DSTS_SUPERSPEED:
 		ret = USB_DWC3_SPEED_SUPER;
-		INFO("%s = SuperSpeed\n", __func__);
+		VERBOSE("%s = SuperSpeed\n", __func__);
 		break;
 	case DWC3_DSTS_HIGHSPEED:
 		ret = USB_DWC3_SPEED_HIGH;
-		INFO("%s = HighSpeed\n", __func__);
+		VERBOSE("%s = HighSpeed\n", __func__);
 		break;
 	case DWC3_DSTS_FULLSPEED1:
 		ret = USB_DWC3_SPEED_FULL_48;
-		INFO("%s = FullSpeed_48M\n", __func__);
+		VERBOSE("%s = FullSpeed_48M\n", __func__);
 		break;
 	case DWC3_DSTS_FULLSPEED2:
 		ret = USB_DWC3_SPEED_FULL;
-		INFO("%s = FullSpeed\n", __func__);
+		VERBOSE("%s = FullSpeed\n", __func__);
 		break;
 	case DWC3_DSTS_LOWSPEED:
 		ret = USB_DWC3_SPEED_LOW;
-		INFO("%s = LowSpeed\n", __func__);
+		VERBOSE("%s = LowSpeed\n", __func__);
 		break;
 	default: /* Invalid */
 		ret = USB_DWC3_SPEED_INVALID;
-		INFO("%s = Invalid\n", __func__);
+		VERBOSE("%s = Invalid\n", __func__);
 		break;
 	}
 
@@ -1667,14 +1667,14 @@ static enum usb_action dwc3_handle_dev_e
 
 	switch (type) {
 	case USB_DWC3_DEVICE_EVENT_DISCONNECT:
-		INFO("Event: Device Disconnect\n");
+		VERBOSE("Event: Device Disconnect\n");
 		/* For SuperSpeed, set DCTL[8:5] to 5*/
 
 		action = USB_DISCONNECT;
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_RESET:
-		INFO("Event: Device Reset\n");
+		VERBOSE("Event: Device Reset\n");
 		/*
 		 * Spec says, Wait till EP0 finishes current transfer and reaches Setup-Stage.
 		 * But there is no wait in Linux driver which means that Host will take care while
@@ -1684,13 +1684,11 @@ static enum usb_action dwc3_handle_dev_e
 
 		if (dwc3_handle->EP0_State != HAL_PCD_EP0_SETUP_QUEUED) {
 			ret = dwc3_ep_stop_xfer(dwc3_handle, &dwc3_handle->pcd_handle->out_ep[0]);
-			if (ret != USBD_OK) {
-				ERROR("%s: %d\n", __func__, __LINE__);
-			}
+			if (ret != USBD_OK)
+				INFO("%s: EP0 STOP XFER failed: %u\n", __func__, ret);
 			ret = dwc3_epaddr_set_stall(dwc3_handle, EP0_OUT); // OUT EP0
-			if (ret != USBD_OK) {
-				ERROR("%s: %d\n", __func__, __LINE__);
-			}
+			if (ret != USBD_OK)
+				INFO("%s: EP SET STALL failed: %u\n", __func__, ret);
 		}
 
 		/*
@@ -1756,7 +1754,7 @@ static enum usb_action dwc3_handle_dev_e
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_CONNECT_DONE:
-		INFO("Event: Device Connection Done\n");
+		VERBOSE("Event: Device Connection Done\n");
 
 		/*
 		 * hpcd->Init.ep0_mps = MaxPacketSize to 512 (SuperSpeed), 64 (High-Speed),
@@ -1815,52 +1813,52 @@ static enum usb_action dwc3_handle_dev_e
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_LINK_STATUS_CHANGE:
-		INFO("Event: Link Status Change : %s(%d)\n",
-			   dwc3_get_linkstate_str(dwc3_read_dev_evt_linkstate(event)),
-			   dwc3_read_dev_evt_linkstate(event));
+		VERBOSE("Event: Link Status Change : %s(%d)\n",
+			dwc3_get_linkstate_str(dwc3_read_dev_evt_linkstate(event)),
+			dwc3_read_dev_evt_linkstate(event));
 		/* Handle link state changes, U0,U1,U2,U3 with suspend/resume functions */
 
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_WAKEUP:
-		INFO("Event: Wake-UP\n");
+		VERBOSE("Event: Wake-UP\n");
 		// resume function
 		action = USB_RESUME;
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_HIBER_REQ:
-		INFO("Event: Hibernation Request\n");
+		VERBOSE("Event: Hibernation Request\n");
 		/* hibernation function */
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_EOPF: /* End of Periodic Frame */
-		INFO("Event: Suspend\n");
+		VERBOSE("Event: Suspend\n");
 		action = USB_SUSPEND;
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_SOF: /* Start of Periodic Frame */
-		INFO("Event: Start of Periodic Frame\n");
+		VERBOSE("Event: Start of Periodic Frame\n");
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_L1SUSPEND:
-		INFO("Event: L1 Suspend\n");
+		VERBOSE("Event: L1 Suspend\n");
 		action = USB_LPM;
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_ERRATIC_ERROR: /* Erratic Error */
-		INFO("Event: Erratic Error\n");
+		VERBOSE("Event: Erratic Error\n");
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_CMD_CMPL: /* Command Complete */
-		INFO("Event: Command Complete\n");
+		VERBOSE("Event: Command Complete\n");
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_OVERFLOW: /* Overflow */
-		INFO("Event: Overflow\n");
+		VERBOSE("Event: Overflow\n");
 		break;
 
 	case USB_DWC3_DEVICE_EVENT_L1WAKEUP:
-		INFO("Event: L1 Wake-UP\n");
+		VERBOSE("Event: L1 Wake-UP\n");
 		// resume function
 		action = USB_RESUME;
 		break;
@@ -1989,7 +1987,7 @@ static enum usb_status dwc3_core_init(dw
 {
 	enum usb_status ret;
 
-	VERBOSE("Core ID %08x\n", (uint32_t)DWC3_regread(dwc3_handle->usb_global, DWC3_GSNPSID));
+	VERBOSE("Core ID %08x\n", DWC3_regread(dwc3_handle->usb_global, DWC3_GSNPSID));
 
 	/* Set GUSB3PIPECTL for all USB3 ports (1-n), currently doing only for 0 */
 	if ((phy_itface & USB3PHY_MASK) != 0U) {
@@ -2121,23 +2119,23 @@ static enum usb_status dwc3_set_dev_spee
 	switch (speed) {
 	case USB_DWC3_SPEED_SUPER:
 		DWC3_regwrite(dwc3_handle->usb_device, DWC3_DCFG, reg | DWC3_DSTS_SUPERSPEED);
-		INFO("%s = SuperSpeed\n", __func__);
+		VERBOSE("%s = SuperSpeed\n", __func__);
 		break;
 	case USB_DWC3_SPEED_HIGH:
 		DWC3_regwrite(dwc3_handle->usb_device, DWC3_DCFG, reg |  DWC3_DSTS_HIGHSPEED);
-		INFO("%s = HighSpeed\n", __func__);
+		VERBOSE("%s = HighSpeed\n", __func__);
 		break;
 	case USB_DWC3_SPEED_FULL_48:
 		DWC3_regwrite(dwc3_handle->usb_device, DWC3_DCFG, reg |  DWC3_DSTS_FULLSPEED1);
-		INFO("%s = FullSpeed_48M\n", __func__);
+		VERBOSE("%s = FullSpeed_48M\n", __func__);
 		break;
 	case USB_DWC3_SPEED_FULL:
 		DWC3_regwrite(dwc3_handle->usb_device, DWC3_DCFG, reg |  DWC3_DSTS_FULLSPEED2);
-		INFO("%s = FullSpeed\n", __func__);
+		VERBOSE("%s = FullSpeed\n", __func__);
 		break;
 	case USB_DWC3_SPEED_LOW:
 		DWC3_regwrite(dwc3_handle->usb_device, DWC3_DCFG, reg |  DWC3_DSTS_LOWSPEED);
-		INFO("%s = LowSpeed\n", __func__);
+		VERBOSE("%s = LowSpeed\n", __func__);
 		break;
 	default:
 		ret = USBD_FAIL;
@@ -2379,10 +2377,10 @@ void usb_dwc3_init_driver(struct usb_han
 
 		dwc3_handle->intbuffers.evtbufferpos[i] = 0;
 
-		INFO("EventBuffer%u: BuffArea=%lx DmaAddr=%08x CoherentMapAddr=%p\n", i,
-			   (PHYS_AREA + EVTBUF_AREA_OFFSET),
-			   (uint32_t)dwc3_handle->intbuffers.evtbuffer_dma_addr[i],
-			   dwc3_handle->intbuffers.evtbuffer_addr[i]);
+		VERBOSE("EventBuffer%u: BuffArea=%lx DmaAddr=%08x CoherentMapAddr=%p\n", i,
+			(PHYS_AREA + EVTBUF_AREA_OFFSET),
+			(uint32_t)dwc3_handle->intbuffers.evtbuffer_dma_addr[i],
+			dwc3_handle->intbuffers.evtbuffer_addr[i]);
 	}
 
 	/* MAP TRB Coherent and DMA address for EP0IN and EP0OUT */
Index: tf-a-txmp-2550/common/bl_common.c
===================================================================
--- tf-a-txmp-2550.orig/common/bl_common.c
+++ tf-a-txmp-2550/common/bl_common.c
@@ -47,18 +47,6 @@ static int dyn_is_auth_disabled(void)
 }
 #endif /* TRUSTED_BOARD_BOOT */
 
-uintptr_t page_align(uintptr_t value, unsigned dir)
-{
-	/* Round up the limit to the next page boundary */
-	if ((value & PAGE_SIZE_MASK) != 0U) {
-		value &= ~PAGE_SIZE_MASK;
-		if (dir == UP)
-			value += PAGE_SIZE;
-	}
-
-	return value;
-}
-
 /*******************************************************************************
  * Internal function to load an image at a specific address given
  * an image ID and extents of free memory.
@@ -110,7 +98,9 @@ static int load_image(unsigned int image
 
 	/* Check that the image size to load is within limit */
 	if (image_size > image_data->image_max_size) {
-		WARN("Image id=%u size out of bounds\n", image_id);
+		WARN("Image id=%u max image size %u exceeded by %zu bytes\n",
+		     image_id, image_data->image_max_size,
+		     image_size - image_data->image_max_size);
 		io_result = -EFBIG;
 		goto exit;
 	}
Index: tf-a-txmp-2550/include/common/bl_common.h
===================================================================
--- tf-a-txmp-2550.orig/include/common/bl_common.h
+++ tf-a-txmp-2550/include/common/bl_common.h
@@ -18,6 +18,7 @@
 #endif /* __ASSEMBLER__ */
 
 #include <export/common/bl_common_exp.h>
+#include <lib/xlat_tables/xlat_tables_defs.h>
 
 #define UP	U(1)
 #define DOWN	U(0)
@@ -185,7 +186,11 @@ extern const char version_string[];
 const char *get_version(void);
 
 void print_entry_point_info(const entry_point_info_t *ep_info);
-uintptr_t page_align(uintptr_t value, unsigned dir);
+
+static inline uintptr_t page_align(uintptr_t value, unsigned dir)
+{
+	return (value + dir == UP ? PAGE_SIZE - 1 : 0) & ~PAGE_SIZE_MASK;
+}
 
 struct mmap_region;
 
Index: tf-a-txmp-2550/drivers/st/ddr/stm32mp1_ram.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/ddr/stm32mp1_ram.c
+++ tf-a-txmp-2550/drivers/st/ddr/stm32mp1_ram.c
@@ -148,7 +148,7 @@ static int stm32mp1_ddr_setup(void)
 			panic();
 		}
 
-		INFO("Memory size = 0x%zx (%zu MB)\n", retsize, retsize / (1024U * 1024U));
+		INFO("Memory size = 0x%zx (%zu MiB)\n", retsize, retsize / (1024U * 1024U));
 	}
 
 	/*
Index: tf-a-txmp-2550/drivers/st/ddr/stm32mp2_ram.c
===================================================================
--- tf-a-txmp-2550.orig/drivers/st/ddr/stm32mp2_ram.c
+++ tf-a-txmp-2550/drivers/st/ddr/stm32mp2_ram.c
@@ -166,7 +166,7 @@ static int stm32mp2_ddr_setup(void)
 			panic();
 		}
 
-		INFO("Memory size = 0x%zx (%zu MB)\n", retsize, retsize / (1024U * 1024U));
+		INFO("Memory size = 0x%zx (%zu MiB)\n", retsize, retsize / (1024U * 1024U));
 	}
 
 	/*
