diff -Naur a/drivers/i2c/busses/i2c-stm32f7.c b/drivers/i2c/busses/i2c-stm32f7.c
--- a/drivers/i2c/busses/i2c-stm32f7.c	2022-04-06 14:38:49.411401174 -0400
+++ b/drivers/i2c/busses/i2c-stm32f7.c	2022-04-07 03:05:06.674040021 -0400
@@ -2427,12 +2427,12 @@
 {
 	struct stm32f7_i2c_dev *i2c_dev = dev_get_drvdata(dev);
 
-	i2c_mark_adapter_suspended(&i2c_dev->adap);
-
 	if (!device_may_wakeup(dev) && !device_wakeup_path(dev))
-		pinctrl_pm_select_sleep_state(dev);
-
-	pm_runtime_force_suspend(dev);
+	{
+	  i2c_mark_adapter_suspended(&i2c_dev->adap);
+	  pinctrl_pm_select_sleep_state(dev);
+	  pm_runtime_force_suspend(dev);
+	}
 
 	return 0;
 }
diff -Naur a/drivers/input/touchscreen/edt-ft5x06.c b/drivers/input/touchscreen/edt-ft5x06.c
--- a/drivers/input/touchscreen/edt-ft5x06.c	2022-04-01 11:03:56.404716645 -0400
+++ b/drivers/input/touchscreen/edt-ft5x06.c	2022-04-07 03:05:19.370040140 -0400
@@ -1082,6 +1082,7 @@
 	unsigned long irq_flags;
 	int error;
 	char fw_version[EDT_NAME_LEN];
+	bool wakeup_src;
 
 	dev_dbg(&client->dev, "probing for EDT FT5x06 I2C\n");
 
@@ -1238,6 +1239,18 @@
 		return error;
 	}
 
+	wakeup_src = device_property_read_bool(input->dev.parent,
+						    "wakeup-source");
+	if (wakeup_src == true)
+	{
+	  error = enable_irq_wake(client->irq);
+	  printk("enable_irq_wake in probe\n");
+	  if (error)
+	    {
+	      printk("%s error to call irq_wake\n", __func__);
+	    }	    
+	}
+	
 	error = devm_device_add_group(&client->dev, &edt_ft5x06_attr_group);
 	if (error)
 		return error;
@@ -1266,6 +1279,10 @@
 		tsdata->wake_gpio ? desc_to_gpio(tsdata->wake_gpio) : -1,
 		tsdata->reset_gpio ? desc_to_gpio(tsdata->reset_gpio) : -1);
 
+	if (wakeup_src == true)
+	{
+	  device_init_wakeup(&(client->dev), 1);	    
+	}
 	return 0;
 }
 
