diff --git a/drivers/usb/dwc2/platform.c b/drivers/usb/dwc2/platform.c
index 00c2e905b421..f0233f0add4c 100644
--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -553,7 +553,8 @@ static int dwc2_driver_probe(struct platform_device *dev)
 		}
 
 		ggpio = dwc2_readl(hsotg, GGPIO);
-		ggpio |= GGPIO_STM32_OTG_GCCFG_IDEN;
+		if (hsotg->dr_mode == USB_DR_MODE_OTG)
+			ggpio |= GGPIO_STM32_OTG_GCCFG_IDEN;
 		ggpio |= GGPIO_STM32_OTG_GCCFG_VBDEN;
 		dwc2_writel(hsotg, ggpio, GGPIO);
 
@@ -740,7 +741,8 @@ static int __maybe_unused dwc2_resume(struct device *dev)
 			return ret;
 
 		ggpio = dwc2_readl(dwc2, GGPIO);
-		ggpio |= GGPIO_STM32_OTG_GCCFG_IDEN;
+		if (dwc2->dr_mode == USB_DR_MODE_OTG)
+			ggpio |= GGPIO_STM32_OTG_GCCFG_IDEN;
 		ggpio |= GGPIO_STM32_OTG_GCCFG_VBDEN;
 		dwc2_writel(dwc2, ggpio, GGPIO);
 
