Index: git/main.c
===================================================================
--- git.orig/main.c
+++ git/main.c
@@ -25,9 +25,9 @@
 #include "ddr.h"
 #include "sysc.h"
 
-#define	RZG2L_DEVID		(0x841C447)
-#define	RZV2L_DEVID		(0x8447447)
-#define	RZG2UL_DEVID	(0x8450447)
+#define	RZG2L_DEVID	0x841C447
+#define	RZV2L_DEVID	0x8447447
+#define	RZG2UL_DEVID	0x8450447
 
 extern const char *const StartMessMonitorG2[START_MESS_MON_LINE];
 extern const char *const StartMessMonitorV2[START_MESS_MON_LINE];
@@ -42,7 +42,7 @@ extern uintptr_t gUMem[3];
 extern uint32_t gDumpMode;
 uint32_t gDumpStatus;
 
-#define RZG2L_SYC_INCK_HZ           (24000000)
+#define RZG2L_SYC_INCK_HZ	24000000
 
 void Main(void)
 {
@@ -93,22 +93,22 @@ void InitMain(void)
 
 void StartMess( void )
 {
-	unsigned int devceid;
+	unsigned int deviceid;
 
-	devceid = sysc_get_device_id();
+	deviceid = sysc_get_device_id();
 
-	PutStr("  ",1);
+	PutStr("  ", 1);
 
-	switch(devceid)
+	switch(deviceid)
 	{
 		case RZG2L_DEVID:
 			PutMess(StartMessMonitorG2);
 			PutStr(" Product Code : ", 0);
 #if (RZG2L == 1)
-			PutStr("RZ/G2L" ,1);
+			PutStr("RZ/G2L", 1);
 #endif
 #if (RZG2LC == 1)
-			PutStr("RZ/G2LC" ,1);
+			PutStr("RZ/G2LC", 1);
 #endif
 #if (TXRZ_G2L0 == 1)
 			PutStr(" Board : TXRZ-G2L0", 1);
@@ -116,25 +116,28 @@ void StartMess( void )
 #if (TXRZ_G2L1 == 1)
 			PutStr(" Board : TXRZ-G2L1", 1);
 #endif
+#if (TXRZ_G2L2 == 1)
+			PutStr(" Board : TXRZ-G2L2", 1);
+#endif
 #if (QSRZ_G2L0 == 1)
 			PutStr(" Board : QSRZ-G2L0", 1);
 #endif
-#if (QSRZ_G2L0 == 1)
+#if (QSRZ_G2L1 == 1)
 			PutStr(" Board : QSRZ-G2L1", 1);
 #endif
 		break;
 		case RZV2L_DEVID:
 			PutMess(StartMessMonitorV2);
 			PutStr(" Product Code : ", 0);
-			PutStr("RZ/V2L" ,1);
+			PutStr("RZ/V2L", 1);
 		break;
 		case RZG2UL_DEVID:
 			PutMess(StartMessMonitorG2);
 			PutStr(" Product Code : ", 0);
 #if (DEVICE_TYPE == 1)
-			PutStr("RZ/G2UL Type1" ,1);
+			PutStr("RZ/G2UL Type1", 1);
 #else
-			PutStr("RZ/G2UL Type2" ,1);
+			PutStr("RZ/G2UL Type2", 1);
 #endif
 		break;
 		default:
@@ -148,6 +151,7 @@ void DecCom(void)
 	char tmp[64], chCnt, chPtr;
 	uint32_t rtn = 0;
 	uint32_t res;
+
 	chCnt = 1;
 
 	while (rtn == 0)
@@ -172,11 +176,11 @@ void DecCom(void)
 			}
 		}
 	}
-	while(1)
+	while (1)
 	{
-		GetStr(gKeyBuf+chCnt,&chCnt);
+		GetStr(gKeyBuf + chCnt, &chCnt);
 		chPtr = 0;
-		GetStrBlk(gKeyBuf,tmp,&chPtr,0);
+		GetStrBlk(gKeyBuf, tmp, &chPtr, 0);
 		if (chPtr != 1)
 		{
 			ChgLtl2Lrg(tmp);
@@ -186,10 +190,10 @@ void DecCom(void)
 			}
 			else
 			{
-				PutStr("command not found",1);
+				PutStr("command not found", 1);
 			}
 		}
-		PutStr(">",0);
+		PutStr(">", 0);
 		chCnt = 0;
 	}
 }
@@ -198,12 +202,12 @@ long CmpCom(char *str)
 {
 	char *cmpStr, *tmpStr, err;
 	gComNo = 0;
-	while(TBL_END != MonCom[gComNo].comStr)
+	while (TBL_END != MonCom[gComNo].comStr)
 	{
 		err=0;
 		tmpStr = str;
 		cmpStr = MonCom[gComNo].comStr;
-		while(*tmpStr!=0)
+		while (*tmpStr)
 		{
 			if (*tmpStr == *cmpStr)
 			{
@@ -218,9 +222,9 @@ long CmpCom(char *str)
 		}
 		if ((!err) && (*cmpStr == 0))
 		{
-			return(0);
+			return 0;
 		}
 		gComNo++;
 	}
-	return(1);
+	return 1;
 }
Index: git/makefile
===================================================================
--- git.orig/makefile
+++ git/makefile
@@ -42,6 +42,14 @@ FILENAME_ADD = _TXRZ-G2L1
 DEVICE   = RZG2L
 DDR_TYPE = DDR3L
 DDR_SIZE = 512MB
+else ifeq ("$(BOARD)", "TXRZ-G2L2")
+#--------------------------------------
+# Ka-Ro RZ/G2L based TXRZ-G2L2 CoM
+#--------------------------------------
+FILENAME_ADD = _TXRZ-G2L2
+DEVICE   = RZG2L
+DDR_TYPE = DDR3L
+DDR_SIZE = 1GB
 else ifeq ("$(BOARD)", "QSRZ-G2L0")
 #--------------------------------------
 # Ka-Ro RZ/G2L based QSRZ-G2L0 CoM
@@ -322,6 +330,11 @@ ifeq ("$(BOARD)", "TXRZ-G2L1")
 else
 	CFLAGS += -DTXRZ_G2L1=0
 endif
+ifeq ("$(BOARD)", "TXRZ-G2L2")
+	CFLAGS += -DTXRZ_G2L2=1
+else
+	CFLAGS += -DTXRZ_G2L2=0
+endif
 ifeq ("$(BOARD)", "QSRZ-G2L0")
 	CFLAGS += -DQSRZ_G2L0=1
 else
Index: git/ddr/common/ddr.c
===================================================================
--- git.orig/ddr/common/ddr.c
+++ git/ddr/common/ddr.c
@@ -44,6 +44,9 @@
 #include "param_mc_txrz-g2l0.c"
 #elif (TXRZ_G2L1 == 1)
 #include "param_mc_txrz-g2l1.c"
+#elif (TXRZ_G2L2 == 1)
+/* same as g2l0 */
+#include "param_mc_txrz-g2l0.c"
 #elif (QSRZ_G2L0 == 1)
 #include "param_mc_qsrz-g2l0.c"
 #elif (QSRZ_G2L1 == 1)
@@ -59,6 +62,9 @@
 #include "param_swizzle_txrz-g2l0.c"
 #elif (TXRZ_G2L1 == 1)
 #include "param_swizzle_txrz-g2l1.c"
+#elif (TXRZ_G2L2 == 1)
+/* same as g2l0 */
+#include "param_swizzle_txrz-g2l0.c"
 #elif (QSRZ_G2L0 == 1)
 #include "param_swizzle_qsrz-g2l0.c"
 #elif (QSRZ_G2L1 == 1)
@@ -77,9 +83,9 @@
 #define	_MIN(a, b)		((a) < (b) ? (a) : (b))
 #define	_MAX(a, b)		((a) > (b) ? (a) : (b))
 
-#define	MAX_BYTE_LANES		(2U)
-#define	MAX_BEST_VREF_SAVED	(30U)
-#define	VREF_SETP			(1U)
+#define	MAX_BYTE_LANES		2U
+#define	MAX_BEST_VREF_SAVED	30U
+#define	VREF_SETP		1U
 
 #define	INFO(...)
 #define ARRAY_SIZE(X) (sizeof(X)/(sizeof(X[0])))
