Index: linux-qsmp-2550/include/dt-bindings/regulator/nxp,pca9450-regulator.h
===================================================================
--- /dev/null
+++ linux-qsmp-2550/include/dt-bindings/regulator/nxp,pca9450-regulator.h
@@ -0,0 +1,43 @@
+/* SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause) */
+/*
+ * Copyright (C) 2023, STMicroelectronics - All Rights Reserved
+ */
+
+#ifndef __DT_BINDINGS_REGULATOR_PCA9450_REGULATOR_H
+#define __DT_BINDINGS_REGULATOR_PCA9450_REGULATOR_H
+
+/* SCMI voltage domains identifiers */
+
+/* SOC Internal regulators */
+#define VOLTD_SCMI_VDDIO1		0
+#define VOLTD_SCMI_VDDIO2		1
+#define VOLTD_SCMI_VDDIO3		2
+#define VOLTD_SCMI_VDDIO4		3
+#define VOLTD_SCMI_VDDIO		4
+#define VOLTD_SCMI_UCPD			5
+#define VOLTD_SCMI_USB33		6
+#define VOLTD_SCMI_ADC			7
+#define VOLTD_SCMI_GPU			8
+#define VOLTD_SCMI_VREFBUF		9
+
+/* PCA9450 regulators */
+#define VOLTD_SCMI_PCA9450_BUCK1	10
+#define VOLTD_SCMI_PCA9450_BUCK2	11
+#define VOLTD_SCMI_PCA9450_BUCK3	12
+#define VOLTD_SCMI_PCA9450_BUCK4	13
+#define VOLTD_SCMI_PCA9450_BUCK5	14
+#define VOLTD_SCMI_PCA9450_BUCK6	15
+#define VOLTD_SCMI_PCA9450_LDO1		17
+#define VOLTD_SCMI_PCA9450_LDO2		18
+#define VOLTD_SCMI_PCA9450_LDO3		19
+#define VOLTD_SCMI_PCA9450_LDO4		20
+#define VOLTD_SCMI_PCA9450_LDO5		21
+
+/* External regulators */
+#define VOLTD_SCMI_REGU0		26
+#define VOLTD_SCMI_REGU1		27
+#define VOLTD_SCMI_REGU2		28
+#define VOLTD_SCMI_REGU3		29
+#define VOLTD_SCMI_REGU4		30
+
+#endif /*__DT_BINDINGS_REGULATOR_PCA9450_REGULATOR_H */
