Index: linux-txmp-2550/arch/arm64/boot/dts/st/stm32mp25xxai-pinctrl.dtsi
===================================================================
--- linux-txmp-2550.orig/arch/arm64/boot/dts/st/stm32mp25xxai-pinctrl.dtsi
+++ linux-txmp-2550/arch/arm64/boot/dts/st/stm32mp25xxai-pinctrl.dtsi
@@ -4,6 +4,8 @@
  * Author: Alexandre Torgue <alexandre.torgue@foss.st.com> for STMicroelectronics.
  */
 
+#include <dt-bindings/pinctrl/stm32-pinfunc.h>
+
 &pinctrl {
 	st,package = <STM32MP_PKG_AI>;
 
Index: linux-txmp-2550/arch/arm64/boot/dts/st/stm32mp25xxak-pinctrl.dtsi
===================================================================
--- linux-txmp-2550.orig/arch/arm64/boot/dts/st/stm32mp25xxak-pinctrl.dtsi
+++ linux-txmp-2550/arch/arm64/boot/dts/st/stm32mp25xxak-pinctrl.dtsi
@@ -4,6 +4,8 @@
  * Author: Alexandre Torgue <alexandre.torgue@foss.st.com> for STMicroelectronics.
  */
 
+#include <dt-bindings/pinctrl/stm32-pinfunc.h>
+
 &pinctrl {
 	st,package = <STM32MP_PKG_AK>;
 
Index: linux-txmp-2550/arch/arm64/boot/dts/st/stm32mp25xxal-pinctrl.dtsi
===================================================================
--- linux-txmp-2550.orig/arch/arm64/boot/dts/st/stm32mp25xxal-pinctrl.dtsi
+++ linux-txmp-2550/arch/arm64/boot/dts/st/stm32mp25xxal-pinctrl.dtsi
@@ -4,6 +4,8 @@
  * Author: Alexandre Torgue <alexandre.torgue@foss.st.com> for STMicroelectronics.
  */
 
+#include <dt-bindings/pinctrl/stm32-pinfunc.h>
+
 &pinctrl {
 	st,package = <STM32MP_PKG_AL>;
 
