From 5402dac13f318aae8fc3d58c0b9570b842fe8b1e Mon Sep 17 00:00:00 2001
From: Uwe Steinkohl <us@karo-electronics.de>
Date: Fri, 21 Apr 2023 17:39:28 +0200
Subject: [PATCH] clock support for CM33

---
 drivers/clk/renesas/r9a07g044-cpg.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/clk/renesas/r9a07g044-cpg.c b/drivers/clk/renesas/r9a07g044-cpg.c
index 47c16265f..4a16e574e 100644
--- a/drivers/clk/renesas/r9a07g044-cpg.c
+++ b/drivers/clk/renesas/r9a07g044-cpg.c
@@ -269,6 +269,9 @@ static const unsigned int r9a07g044_crit_mod_clks[] __initconst = {
 	MOD_CLK_BASE + R9A07G044_GIC600_GICCLK,
 	MOD_CLK_BASE + R9A07G044_IA55_CLK,
 	MOD_CLK_BASE + R9A07G044_DMAC_ACLK,
+	MOD_CLK_BASE + R9A07G044_SCIF2_CLK_PCK,
+	MOD_CLK_BASE + R9A07G044_OSTM2_PCLK,
+	MOD_CLK_BASE + R9A07G044_SCIF1_CLK_PCK,
 };
 
 const struct rzg2l_cpg_info r9a07g044_cpg_info = {
-- 
2.17.1

