Index: linux-qsmp-1570/drivers/net/phy/smsc.c
===================================================================
--- linux-qsmp-1570.orig/drivers/net/phy/smsc.c
+++ linux-qsmp-1570/drivers/net/phy/smsc.c
@@ -185,9 +185,36 @@ static int smsc_phy_suspend(struct phy_d
 {
 	struct device *dev = &phydev->mdio.dev;
 
+	if (phydev->suspended)
+		return 0;
+
 	/* do not power down PHY when PHY enable power/wakeup */
-	if (!device_may_wakeup(dev))
-		return genphy_suspend(phydev);
+	if (device_may_wakeup(dev))
+		return 0;
+
+	if (phy_interrupt_is_valid(phydev)) {
+		phydev->interrupts = PHY_INTERRUPT_DISABLED;
+		if (phydev->drv->config_intr)
+			phydev->drv->config_intr(phydev);
+	}
+
+	return genphy_suspend(phydev);
+}
+
+static int smsc_phy_resume(struct phy_device *phydev)
+{
+	genphy_resume(phydev);
+
+	if (!phydev->suspended)
+		return 0;
+
+	usleep_range(25000, 26000);
+
+	if (phy_interrupt_is_valid(phydev)) {
+		phydev->interrupts = PHY_INTERRUPT_ENABLED;
+		if (phydev->drv->config_intr)
+			phydev->drv->config_intr(phydev);
+	}
 
 	return 0;
 }
@@ -499,8 +526,8 @@ static struct phy_driver smsc_phy_driver
 	.get_strings	= smsc_get_strings,
 	.get_stats	= smsc_get_stats,
 
-	.suspend	= genphy_suspend,
-	.resume		= genphy_resume,
+	.suspend	= smsc_phy_suspend,
+	.resume		= smsc_phy_resume,
 }, {
 	.phy_id		= 0x0007c110,
 	.phy_id_mask	= 0xfffffff0,
@@ -556,7 +583,7 @@ static struct phy_driver smsc_phy_driver
 	.get_stats	= smsc_get_stats,
 
 	.suspend	= smsc_phy_suspend,
-	.resume		= genphy_resume,
+	.resume		= smsc_phy_resume,
 } };
 
 module_phy_driver(smsc_phy_driver);
