Index: linux-qsmp-1510/drivers/gpu/drm/stm/ltdc.c
===================================================================
--- linux-qsmp-1510.orig/drivers/gpu/drm/stm/ltdc.c
+++ linux-qsmp-1510/drivers/gpu/drm/stm/ltdc.c
@@ -982,7 +982,7 @@ static void ltdc_crtc_mode_set_nofb(stru
 	if (bus_flags & DRM_BUS_FLAG_DE_LOW)
 		val |= GCR_DEPOL;
 
-	if (bus_flags & DRM_BUS_FLAG_PIXDATA_DRIVE_NEGEDGE)
+	if (bus_flags & DRM_BUS_FLAG_PIXDATA_DRIVE_POSEDGE)
 		val |= GCR_PCPOL;
 
 	if (connector && connector->state->dithering == DRM_MODE_DITHERING_ON)
