Index: u-boot/common/fdt_support.c
===================================================================
--- u-boot.orig/common/fdt_support.c
+++ u-boot/common/fdt_support.c
@@ -664,6 +664,9 @@ int fdt_shrink_to_minimum(void *blob, ui
 
 	/* Change the fdt header to reflect the correct size */
 	fdt_set_totalsize(blob, actualsize);
+
+	if (env_get("fdtsize"))
+		env_set_hex("fdtsize", actualsize);
 
 	if (fdt_memrsv) {
 		/* Add the new reservation */
