Index: u-boot/lib/display_options.c
===================================================================
--- u-boot.orig/lib/display_options.c
+++ u-boot/lib/display_options.c
@@ -8,6 +8,7 @@
 #include <console.h>
 #include <div64.h>
 #include <version.h>
+#include <watchdog.h>
 #include <linux/ctype.h>
 #include <asm/io.h>
 
@@ -208,6 +209,8 @@ int print_buffer(ulong addr, const void
 #ifndef CONFIG_SPL_BUILD
 		if (ctrlc())
 			return -1;
+
+		WATCHDOG_RESET();
 #endif
 	}
 
