Index: u-boot/common/cli_readline.c
===================================================================
--- u-boot.orig/common/cli_readline.c
+++ u-boot/common/cli_readline.c
@@ -272,6 +272,12 @@ static int cread_line(const char *const
 			first = 0;
 		}
 
+#ifdef CONFIG_SHOW_ACTIVITY
+		while (!tstc()) {	/* while no incoming data */
+			show_activity(0);
+			WATCHDOG_RESET();
+		}
+#endif
 		ichar = getcmd_getch();
 
 		/* ichar=0x0 when error occurs in U-Boot getc */
