Index: git/core/arch/arm/plat-stm32mp2/drivers/sub.mk
===================================================================
--- git.orig/core/arch/arm/plat-stm32mp2/drivers/sub.mk
+++ git/core/arch/arm/plat-stm32mp2/drivers/sub.mk
@@ -1,6 +1,5 @@
 srcs-$(CFG_STM32_PWR) += stm32mp25_pwr_regul.c
-srcs-$(CFG_STPMIC2) += stm32mp2_pmic.c
-srcs-$(CFG_STPMIC2) += stpmic2.c
+srcs-$(CFG_STPMIC2) += stm32mp2_pmic.c stpmic2.c
 srcs-y += stm32mp25_syscfg.c
 srcs-$(CFG_STM32_PWR) += stm32mp25_pwr.c
 srcs-$(CFG_STM32_PWR_IRQ) += stm32mp25_pwr_irq.c
Index: git/core/drivers/pca9450.c
===================================================================
--- /dev/null
+++ git/core/drivers/pca9450.c
@@ -0,0 +1,488 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright 2024 Lothar Waßmann <LW@KARO-electronics.de>
+ * based on:
+ *     core/arch/arm/plat-stm32mp2/drivers/stpmic2.c
+ *         Copyright (c) 2022-2023, STMicroelectronics - All Rights Reserved
+ */
+
+#include <assert.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <string.h>
+#include <trace.h>
+#include <util.h>
+#include <drivers/stm32_i2c.h>
+#include <kernel/panic.h>
+#include "pca9450.h"
+
+#define VOLTAGE_INDEX_INVALID		((size_t)~0)
+
+static const struct regulator_limits {
+       uint16_t min;
+       uint16_t max;
+       uint16_t step;
+       uint8_t mask;
+} pca9450_regulator_limits[] = {
+	{  6000, 21875,  125, 0x7f, }, /* BUCK[1..3] */
+	{  6000, 34000,  250, 0x7f, }, /* BUCK[4..6] */
+	{ 16000, 33000, 1000, 0x07, }, /* LDO[1] */
+	{  8000, 11500,  500, 0x1f, }, /* LDO[2] */
+	{  8000, 33000, 1000, 0x1f, }, /* LDO[3..4] */
+	{ 18000, 33000, 1000, 0x0f, }, /* LDO[5] */
+};
+
+struct regul_struct {
+	const char *name;
+	size_t limits;
+	uint16_t *volt_table;
+	size_t volt_table_size;
+	uint16_t volt_cr;
+	uint16_t alt_volt_cr;
+	uint16_t en_cr;
+	uint16_t pd_reg;
+	uint8_t en_shift;
+	uint8_t pd_val;
+};
+
+#define DEFINE_BUCK123(_index) {				\
+	.name			= "buck" # _index,		\
+	.en_cr			= BUCK ## _index ## CTRL,	\
+	.en_shift		= BUCK_EN_SHIFT,		\
+	.volt_cr		= BUCK ## _index ## OUT_DVS0,	\
+	.alt_volt_cr		= BUCK ## _index ## OUT_DVS1,	\
+	.limits			= 0,				\
+	.pd_reg			= BUCK ## _index ## CTRL,	\
+	.pd_val			= BUCK_AD_BIT,			\
+}
+
+#define DEFINE_BUCK456(_index) {				\
+	.name			= "buck" # _index,		\
+	.en_cr			= BUCK ## _index ## CTRL,	\
+	.en_shift		= BUCK_EN_SHIFT,		\
+	.volt_cr		= BUCK ## _index ## OUT,	\
+	.limits			= 1,				\
+	.pd_reg			= BUCK ## _index ## CTRL,	\
+	.pd_val			= BUCK_AD_BIT,			\
+}
+
+#define DEFINE_LDO(_index, _limits) {				\
+	.name			= "ldo" # _index,		\
+	.en_cr			= LDO ## _index ## CTRL,	\
+	.en_shift		= LDO_EN_SHIFT,			\
+	.volt_cr		= LDO ## _index ## CTRL,	\
+	.limits			= _limits,			\
+	.pd_reg			= LDO_AD_CTRL,			\
+	.pd_val			= LDO ## _index ## _AD_BIT,	\
+}
+
+static uint16_t ldo1_volt_table[8] = {
+	1600,
+	1700,
+	1800,
+	1900,
+	3000,
+	3100,
+	3200,
+	3300,
+};
+
+static uint16_t *volt_tables[] = {
+	NULL,
+	NULL,
+	ldo1_volt_table,
+	NULL,
+	NULL,
+	NULL,
+};
+
+/* Table of Regulators in PMIC SoC */
+static struct regul_struct regul_table[] = {
+	[PCA9450_BUCK1] = DEFINE_BUCK123(1),
+	[PCA9450_BUCK2] = DEFINE_BUCK123(2),
+	[PCA9450_BUCK3] = DEFINE_BUCK123(3),
+	[PCA9450_BUCK4] = DEFINE_BUCK456(4),
+	[PCA9450_BUCK5] = DEFINE_BUCK456(5),
+	[PCA9450_BUCK6] = DEFINE_BUCK456(6),
+	[PCA9450_LDO1] = DEFINE_LDO(1, 2),
+	[PCA9450_LDO2] = DEFINE_LDO(2, 3),
+	[PCA9450_LDO3] = DEFINE_LDO(3, 4),
+	[PCA9450_LDO4] = DEFINE_LDO(4, 4),
+	[PCA9450_LDO5] = DEFINE_LDO(5, 5),
+};
+
+static inline const struct regulator_limits *get_limits(const struct regul_struct *regul)
+{
+	return &pca9450_regulator_limits[regul->limits];
+}
+
+static inline uint8_t mv_to_regval(const struct regulator_limits *limits, uint16_t mv)
+{
+	return (mv * 10 - limits->min) / limits->step;
+}
+
+static inline uint16_t regval_to_mv(const struct regulator_limits *limits, uint8_t regval)
+{
+	uint16_t mv = (regval & limits->mask) * limits->step + limits->min;
+
+	if (mv > limits->max)
+		mv = limits->max;
+	return mv / 10;
+}
+
+TEE_Result pca9450_register_read(struct pmic_handle_s *pmic,
+				 uint16_t register_id, uint8_t *value)
+{
+	int res = stm32_i2c_read_write_membyte(pmic->pmic_i2c_handle,
+					       pmic->pmic_i2c_addr,
+					       register_id, value,
+					       false /* !write */);
+	if (res) {
+		EMSG("Failed to read reg:0x%04x", register_id);
+		return TEE_ERROR_COMMUNICATION;
+	}
+
+	FMSG("REG:0x%04x v=0x%02x", register_id, *value);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result pca9450_register_write(struct pmic_handle_s *pmic,
+				  uint16_t register_id, uint8_t value)
+{
+	int ret;
+
+	ret = stm32_i2c_read_write_membyte(pmic->pmic_i2c_handle,
+					   pmic->pmic_i2c_addr,
+					   register_id, &value,
+					   true /* write */);
+	if (ret) {
+		EMSG("Failed to write reg:0x%4x: %d", register_id, ret);
+		return TEE_ERROR_COMMUNICATION;
+	}
+
+	FMSG("REG:0x%04x written v=0x%02x", register_id, value);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result pca9450_register_update(struct pmic_handle_s *pmic,
+				   uint16_t register_id, uint8_t value,
+				   uint8_t mask)
+{
+	TEE_Result res;
+	uint8_t val;
+
+	res = pca9450_register_read(pmic, register_id, &val);
+	if (res)
+		return res;
+
+	val = (val & ~mask) | (value & mask);
+
+	FMSG("REG:0x%04x v=0x%02x mask=0x%02x -> 0x%02x", register_id, value,
+	     mask, val);
+
+	return pca9450_register_write(pmic, register_id, val);
+}
+
+TEE_Result pca9450_regulator_set_state(struct pmic_handle_s *pmic,
+				       uint8_t id, bool enable)
+{
+	const struct regul_struct *regul = &regul_table[id];
+	const uint8_t en_mask = REG_ENABLE_MASK << regul->en_shift;
+	const uint8_t en_val = (enable ? REG_ON : REG_OFF) << regul->en_shift;
+
+	return pca9450_register_update(pmic, regul->en_cr, en_val, en_mask);
+}
+
+TEE_Result pca9450_regulator_get_state(struct pmic_handle_s *pmic,
+				       uint8_t id, bool *enabled)
+{
+	TEE_Result res;
+	const struct regul_struct *regul = &regul_table[id];
+	uint8_t val;
+
+	res = pca9450_register_read(pmic, regul->en_cr, &val);
+	if (res)
+		return res;
+
+	*enabled = !!(val & (REG_ENABLE_MASK << regul->en_shift));
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result pca9450_regulator_levels_mv(struct pmic_handle_s *pmic __unused,
+				       uint8_t id, const uint16_t **levels,
+				       size_t *levels_count)
+{
+	const struct regul_struct *regul = &regul_table[id];
+
+	if (!regul)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (levels_count)
+		*levels_count = regul->volt_table_size;
+	if (levels)
+		*levels = regul->volt_table;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result pca9450_regulator_get_voltage(struct pmic_handle_s *pmic,
+					 uint8_t id, uint16_t *val)
+{
+	TEE_Result res;
+	const struct regul_struct *regul = &regul_table[id];
+	const struct regulator_limits *limits = get_limits(regul);
+	uint8_t mask = limits->mask;
+	uint8_t idx;
+
+	if (regul->volt_table_size == 0)
+		return TEE_ERROR_GENERIC;
+
+	res = pca9450_register_read(pmic, regul->volt_cr, &idx);
+	if (res)
+		return res;
+
+	idx &= mask;
+	if (idx > regul->volt_table_size)
+		idx = regul->volt_table_size - 1;
+
+	*val = regul->volt_table[idx];
+	return TEE_SUCCESS;
+}
+
+static size_t voltage_to_index(const struct regul_struct *regul,
+			       uint16_t millivolts)
+{
+	size_t i;
+
+	assert(regul->volt_table);
+	for (i = 0; i < regul->volt_table_size; i++)
+		if (regul->volt_table[i] == millivolts)
+			return i;
+
+	return VOLTAGE_INDEX_INVALID;
+}
+
+TEE_Result pca9450_regulator_set_voltage(struct pmic_handle_s *pmic,
+					 uint8_t id, uint16_t mv)
+{
+	const struct regul_struct *regul = &regul_table[id];
+	const struct regulator_limits *limits = get_limits(regul);
+	size_t index;
+	uint8_t mask = limits->mask;
+
+	if (mv * 10 > limits->max || mv * 10 < limits->min)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	index = voltage_to_index(regul, mv);
+	if (index == VOLTAGE_INDEX_INVALID)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return pca9450_register_update(pmic, regul->volt_cr,
+				       index, mask);
+}
+
+TEE_Result pca9450_regulator_get_prop(struct pmic_handle_s *pmic, uint8_t id,
+				      enum pca9450_prop_id prop, uint8_t *arg)
+{
+	TEE_Result res;
+	const struct regul_struct *regul = &regul_table[id];
+	uint8_t val;
+
+	IMSG("%s: get prop 0x%04x", regul->name, prop);
+
+	switch (prop) {
+	case PCA9450_PULL_DOWN:
+		res = pca9450_register_read(pmic, regul->pd_reg, &val);
+		if (res)
+			return res;
+		*arg = val;
+		break;
+	default:
+		EMSG("Invalid prop 0x%04x", prop);
+		panic();
+	}
+
+	return res;
+}
+
+TEE_Result pca9450_regulator_set_prop(struct pmic_handle_s *pmic, uint8_t id,
+				      enum pca9450_prop_id prop, uint8_t arg)
+{
+	const struct regul_struct *regul = &regul_table[id];
+
+	IMSG("%s: set prop 0x%04x arg=%u", regul->name, prop, arg);
+
+	switch (prop) {
+	case PCA9450_PULL_DOWN:
+		return pca9450_register_update(pmic, regul->pd_reg,
+					       arg ? regul->pd_val : 0,
+					       regul->pd_val);
+	default:
+		EMSG("Invalid prop 0x%04x", prop);
+		panic();
+	}
+
+	return TEE_SUCCESS;
+}
+
+void pca9450_dump_regulators(struct pmic_handle_s *pmic)
+{
+	size_t i;
+
+	assert(pmic);
+
+	for (i = 0; i < ARRAY_SIZE(regul_table); i++) {
+		uint16_t val;
+		bool state;
+		char __maybe_unused const *name;
+
+		if (!regul_table[i].volt_cr)
+			continue;
+
+		pca9450_regulator_get_voltage(pmic, i, &val);
+		pca9450_regulator_get_state(pmic, i, &state);
+
+		name = regul_table[i].name;
+		IMSG("PMIC regul %s: %s, %dmv", name,
+		     state ? "EN" : "DIS", val);
+	}
+}
+
+TEE_Result pca9450_get_chip_id(struct pmic_handle_s *pmic, uint8_t *val)
+{
+	return pca9450_register_read(pmic, DEVICE_ID, val);
+}
+
+/*
+ *
+ * Low Power
+ *
+ */
+TEE_Result pca9450_lp_set_state(struct pmic_handle_s *pmic, uint8_t id,
+				bool enable)
+{
+	const struct regul_struct *regul = &regul_table[id];
+	const uint8_t en_mask = REG_ENABLE_MASK << regul->en_shift;
+	const uint8_t en_val = (enable ? REG_STBY_ON : REG_STBY_OFF) << regul->en_shift;
+
+	return pca9450_register_update(pmic, regul->en_cr, en_val, en_mask);
+}
+
+TEE_Result pca9450_lp_set_voltage(struct pmic_handle_s *pmic, uint8_t id,
+				  uint16_t mv)
+{
+	TEE_Result res;
+	const struct regul_struct *regul = &regul_table[id];
+	const struct regulator_limits *limits = get_limits(regul);
+	size_t index;
+
+	if (!regul->alt_volt_cr)
+		return TEE_SUCCESS;
+	if (!regul->volt_table_size)
+		return TEE_SUCCESS;
+
+	index = voltage_to_index(regul, mv);
+	if (index == VOLTAGE_INDEX_INVALID)
+		return TEE_ERROR_GENERIC;
+
+	res = pca9450_register_update(pmic, regul->alt_volt_cr,
+				      index, limits->mask);
+	if (res)
+		return res;
+
+	res = pca9450_register_update(pmic, regul->en_cr, 0*BUCK_DVS_CTRL, BUCK_DVS_CTRL);
+	return res;
+}
+
+/*
+ *
+ * Interrupt handling
+ *
+ */
+TEE_Result pca9450_set_irq_mask(struct pmic_handle_s *pmic, uint8_t num,
+				bool state)
+{
+	uint8_t mask = BIT(num);
+	uint8_t val = state ? mask : 0;
+
+	return pca9450_register_update(pmic, INT1_MSK, val, mask);
+}
+
+enum itr_return pca9450_handle_irq(struct pmic_handle_s *pmic)
+{
+	enum itr_return res = ITRR_NONE;
+	uint8_t val;
+
+	FMSG("PCA9450 irq handler");
+
+	if (pca9450_register_read(pmic, INT1, &val))
+		panic();
+
+	if (val) {
+		int it_id;
+
+		FMSG("PCA9450 irq pending irq=0x%02x", val);
+
+		pmic->irq_count++;
+
+		for (it_id = 0; it_id < 8; it_id++) {
+			if (val & BIT(it_id)) {
+				if (pca9450_irq_callback(pmic, it_id) == ITRR_HANDLED)
+					res = ITRR_HANDLED;
+			}
+		}
+	}
+
+	return res;
+}
+
+void pca9450_init(void)
+{
+	size_t id;
+
+	for (id = 0; id < ARRAY_SIZE(regul_table); id++) {
+		struct regul_struct *regul = &regul_table[id];
+		const struct regulator_limits *limits = get_limits(regul);
+		size_t num_voltages = limits->mask + 1;
+		size_t idx;
+		uint16_t *volt_table = volt_tables[regul->limits];
+
+		if (!volt_table)
+			volt_table = calloc(num_voltages, sizeof(*volt_table));
+		if (!volt_table) {
+			EMSG("Failed to allocate volt_table for %s\n",
+			     regul->name);
+			panic();
+		}
+
+		FMSG("%s@%d: %s volt_table=%p\n", __func__, __LINE__,
+		     regul->name, volt_table);
+
+		for (idx = 0; idx < num_voltages; idx++) {
+			/* skip tables that have already been set up */
+			if (!volt_table[idx]) {
+				int ret;
+
+				ret = regval_to_mv(limits, idx);
+				if (ret < 0) {
+					EMSG("Regulator reg value %08zx out of range 0..%u\n",
+					     idx, limits->mask + 1U);
+					panic();
+				}
+				if ((uint16_t)ret > limits->max)
+					break;
+				volt_table[idx] = ret;
+			}
+			FMSG("%s@%d: %s[%zu]=%umV\n", __func__, __LINE__,
+			     regul->name, idx, volt_table[idx]);
+		}
+		regul->volt_table = volt_table;
+		regul->volt_table_size = idx;
+		IMSG("%s@%d: %s volt_table=%p %zu voltages [%u..%u]mV\n",
+		     __func__, __LINE__, regul->name, regul->volt_table,
+		     regul->volt_table_size, volt_table[0], volt_table[idx - 1]);
+	}
+}
Index: git/core/drivers/pca9450.h
===================================================================
--- /dev/null
+++ git/core/drivers/pca9450.h
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2024, Lothar Waßmann <LW@KARO-electronics.de>
+ * based on: include/drivers/st/stpmic2.h
+ *    Copyright (c) 2022-2023, STMicroelectronics - All Rights Reserved
+ */
+
+#ifndef __PCA9450_H__
+#define __PCA9450_H__
+
+#include <tee_api_types.h>
+#include <kernel/interrupt.h>
+#include <sys/queue.h>
+
+#ifndef SD_VSEL
+#define SD_VSEL 0
+#endif
+
+enum regulator_id {
+	PCA9450_BUCK1,
+	PCA9450_BUCK2,
+	PCA9450_BUCK3,
+	PCA9450_BUCK4,
+	PCA9450_BUCK5,
+	PCA9450_BUCK6,
+	PCA9450_LDO1,
+	PCA9450_LDO2,
+	PCA9450_LDO3,
+	PCA9450_LDO4,
+	PCA9450_LDO5,
+};
+
+enum pca9450_variant {
+	PCA9450A,
+	PCA9450BC,
+};
+
+#define CHIP_ID_MASK		0xf0
+#define CHIP_ID_SHIFT		4
+#define CHIP_ID_PCA9450A	1
+#define CHIP_ID_PCA9450BC	3
+
+/* Status Registers */
+#define DEVICE_ID		0x00
+#define INT1			0x01
+#define INT1_MSK		0x02
+#define STATUS1			0x03
+#define RESET_CTRL		0x08
+#define WDOG_B_CFG_MSK		(3 << WDOG_B_CFG_SHIFT)
+#define WDOG_B_CFG_SHIFT	6
+#define PMIC_RST_CFG_MSK	(3 << PMIC_RST_CFG_SHIFT)
+#define PMIC_RST_CFG_SHIFT	4
+
+#define CONFIG1			0x09
+#define CONFIG2			0x0a
+#define BUCK123_DVS		0x0c
+#define BUCK1OUT_LIMIT		0x0d
+#define BUCK2OUT_LIMIT		0x0e
+#define BUCK3OUT_LIMIT		0x0f
+#define BUCK1CTRL		0x10
+#define BUCK1OUT_DVS0		0x11
+#define BUCK1OUT_DVS1		0x12
+#define BUCK2CTRL		0x13
+#define BUCK2OUT_DVS0		0x14
+#define BUCK2OUT_DVS1		0x15
+#define BUCK3CTRL		0x16
+#define BUCK3OUT_DVS0		0x17
+#define BUCK3OUT_DVS1		0x18
+#define BUCK4CTRL		0x19
+#define BUCK4OUT		0x1a
+#define BUCK5CTRL		0x1b
+#define BUCK5OUT		0x1c
+#define BUCK6CTRL		0x1d
+#define BUCK6OUT		0x1e
+#define LDO_AD_CTRL		0x20
+#define LDO1CTRL		0x21
+#define LDO2CTRL		0x22
+#define LDO3CTRL		0x23
+#define LDO4CTRL		0x24
+#if SD_VSEL == 0
+#define LDO5CTRL		0x25
+#else
+#define LDO5CTRL		0x26
+#endif
+#define LOADSW_CTRL		0x2a
+#define VRFLT1_STS		0x2b
+#define VRFLT2_STS		0x2c
+#define VRFLT1_MASK		0x2d
+#define VRFLT2_MASK		0x2e
+
+#define BUCK_EN_SHIFT		0
+#define LDO_EN_SHIFT		6
+#define BUCK_DVS_CTRL		BIT(4)
+
+#define REG_ENABLE_MASK		0x3
+#define REG_OFF			0x0
+#define REG_STBY_ON		0x3 /* actually 0x1! */
+#define REG_STBY_OFF		0x3 /* actually 0x2! */
+#define REG_ON			0x3
+
+#define BUCK_AD_BIT		BIT(3)
+#define LDO1_AD_BIT		BIT(7)
+#define LDO2_AD_BIT		BIT(6)
+#define LDO3_AD_BIT		BIT(5)
+#define LDO4_AD_BIT		BIT(4)
+#define LDO5_AD_BIT		BIT(3)
+
+enum pca9450_prop_id {
+	PCA9450_PULL_DOWN,
+	PCA9450_WDOG_B_RESET,
+	PCA9450_PMIC_RST_B,
+};
+
+#ifdef CFG_STM32_PWR_IRQ
+struct pmic_it_handle_s {
+	uint8_t pmic_it;
+	uint8_t notif_id;
+
+	SLIST_ENTRY(pmic_it_handle_s) link;
+};
+#endif
+
+struct pmic_handle_s {
+	struct i2c_handle_s *pmic_i2c_handle;
+	uint16_t pmic_i2c_addr;
+	uint16_t irq_count;
+#ifdef CFG_STM32_PWR_IRQ
+	SLIST_HEAD(, pmic_it_handle_s) it_list;
+#endif
+};
+
+TEE_Result pca9450_register_read(struct pmic_handle_s *pmic,
+				 uint16_t register_id, uint8_t *value);
+TEE_Result pca9450_register_write(struct pmic_handle_s *pmic,
+				  uint16_t register_id, uint8_t value);
+TEE_Result pca9450_register_update(struct pmic_handle_s *pmic,
+				   uint16_t register_id, uint8_t value,
+				   uint8_t mask);
+
+TEE_Result pca9450_regulator_set_state(struct pmic_handle_s *pmic,
+				       uint8_t id, bool enable);
+TEE_Result pca9450_regulator_get_state(struct pmic_handle_s *pmic,
+				       uint8_t id, bool *enabled);
+
+TEE_Result pca9450_regulator_levels_mv(struct pmic_handle_s *pmic,
+				       uint8_t id, const uint16_t **levels,
+				       size_t *levels_count);
+TEE_Result pca9450_regulator_get_voltage(struct pmic_handle_s *pmic,
+					 uint8_t id, uint16_t *val);
+TEE_Result pca9450_regulator_set_voltage(struct pmic_handle_s *pmic,
+					 uint8_t id, uint16_t millivolts);
+
+TEE_Result pca9450_regulator_get_prop(struct pmic_handle_s *pmic, uint8_t id,
+				      enum pca9450_prop_id prop, uint8_t *arg);
+TEE_Result pca9450_regulator_set_prop(struct pmic_handle_s *pmic, uint8_t id,
+				      enum pca9450_prop_id prop, uint8_t arg);
+
+void pca9450_dump_regulators(struct pmic_handle_s *pmic);
+TEE_Result pca9450_get_chip_id(struct pmic_handle_s *pmic, uint8_t *val);
+TEE_Result pca9450_get_product_id(struct pmic_handle_s *pmic, uint8_t *val);
+
+/* Low Power handling */
+TEE_Result pca9450_lp_set_state(struct pmic_handle_s *pmic, uint8_t id,
+				bool enable);
+TEE_Result pca9450_lp_set_voltage(struct pmic_handle_s *pmic, uint8_t id,
+				  uint16_t mv);
+
+/* Interrupt Handling */
+enum itr_return pca9450_irq_callback(struct pmic_handle_s *pmic, uint8_t it_id);
+TEE_Result pca9450_set_irq_mask(struct pmic_handle_s *pmic, uint8_t num,
+				bool state);
+TEE_Result pca9450_irq_gen(struct pmic_handle_s *pmic, uint8_t num);
+enum itr_return pca9450_handle_irq(struct pmic_handle_s *pmic);
+
+void pca9450_init(void);
+#endif /* __PCA9450_H */
Index: git/core/drivers/pmic-pca9450.c
===================================================================
--- /dev/null
+++ git/core/drivers/pmic-pca9450.c
@@ -0,0 +1,778 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright 2024 Lothar Waßmann <LW@KARO-electronics.de>
+ * based on:
+ *     core/arch/arm/plat-stm32mp2/drivers/stm32mp2_pmic.c
+ *         Copyright (c) 2022-2023, STMicroelectronics
+ */
+
+#include <config.h>
+#include <drivers/regulator.h>
+#include <drivers/stm32_exti.h>
+#include <drivers/stm32_i2c.h>
+#include <drivers/stm32mp25_pwr.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <kernel/notif.h>
+#include <kernel/panic.h>
+#include <kernel/pm.h>
+#include <libfdt.h>
+#include <platform_config.h>
+#include <stdbool.h>
+#include <stm32_util.h>
+#include <stm32mp_pm.h>
+#include <trace.h>
+#include <util.h>
+#include "pca9450.h"
+
+#define PMIC_I2C_TRIALS			1
+#define PMIC_I2C_TIMEOUT_BUSY_MS	5
+
+/* Mutex for protecting PMIC accesses */
+static struct mutex pmic_mu = MUTEX_INITIALIZER;
+
+/*
+ * Low power configurations:
+ *
+ * STM32_PM_DEFAULT
+ *   "default" sub nodes in device-tree
+ *   is applied at probe, and re-applied at PM resume.
+ *   should support STOP1, LP-STOP1, STOP2, LP-STOP2
+ *
+ * STM32_PM_LPLV
+ *   "lplv" sub nodes in device-tree
+ *   should support STOP1, LPLV-STOP1, STOP2, LPLV-STOP2
+ *
+ * STM32_PM_STANDBY
+ *   "standby" sub nodes in device-tree
+ *   should support STANDBY1-DDR-SR
+ *   is applied in pm suspend call back
+ *
+ * STM32_PM_OFF
+ *   "off" sub nodes in device-tree
+ *   should support STANDBY-DDR-OFF mode
+ *   and should be applied before shutdown
+ *
+ */
+#define STM32_PM_DEFAULT		0
+#define STM32_PM_LPLV			1
+#define STM32_PM_STANDBY		2
+#define STM32_PM_OFF			3
+#define STM32_PM_NB_SOC_MODES		4
+
+static const char *plat_get_lp_mode_name(int mode __unused)
+{
+	switch (mode) {
+	case STM32_PM_DEFAULT:
+		return "default";
+	case STM32_PM_LPLV:
+		return "lplv";
+	case STM32_PM_STANDBY:
+		return "standby";
+	case STM32_PM_OFF:
+		return "off";
+	default:
+		EMSG("Invalid lp mode %d", mode);
+		panic();
+	}
+}
+
+#define PCA9450_LP_STATE_OFF		BIT(0)
+#define PCA9450_LP_STATE_ON		BIT(1)
+#define PCA9450_LP_STATE_UNMODIFIED	BIT(2)
+#define PCA9450_LP_STATE_SET_VOLT	BIT(3)
+
+struct regu_handle_s {
+	struct pmic_handle_s *pmic;
+	struct regul_desc desc;
+	uint32_t id;
+	uint16_t lp_mv[STM32_PM_NB_SOC_MODES];
+	uint8_t lp_state[STM32_PM_NB_SOC_MODES];
+};
+
+static TEE_Result pmic_set_state(const struct regul_desc *desc, bool enable)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+
+	FMSG("%u:%s: set state to %u", id, desc->node_name, enable);
+
+	return pca9450_regulator_set_state(pmic, id, enable);
+}
+
+static TEE_Result pmic_get_state(const struct regul_desc *desc, bool *enabled)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+
+	FMSG("%u:%s: get state\n", id, desc->node_name);
+
+	return pca9450_regulator_get_state(pmic, id, enabled);
+}
+
+static TEE_Result pmic_get_voltage(const struct regul_desc *desc, uint16_t *mv)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+	uint16_t val;
+	TEE_Result res;
+
+	res = pca9450_regulator_get_voltage(pmic, id, &val);
+	if (res)
+		return res;
+
+	*mv = val;
+
+	FMSG("%s: get volt %u.%03u mV", desc->node_name, val / 1000U, val % 1000U);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pmic_verify_voltage(const struct regul_desc *desc, uint16_t mv)
+{
+	TEE_Result res;
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+	const uint16_t *levels;
+	size_t count;
+	size_t n;
+
+	res = pca9450_regulator_levels_mv(pmic, id, &levels,
+					  &count);
+	if (res)
+		return res;
+
+	for (n = 0; n < count; n++)
+		if (levels[n] == mv)
+			return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+
+}
+
+static TEE_Result pmic_set_voltage(const struct regul_desc *desc, uint16_t mv)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+	TEE_Result res;
+
+	FMSG("%s: set volt to %u.%03u mV", desc->node_name,
+	     mv / 1000U, mv % 1000U);
+
+	res = pca9450_regulator_set_voltage(pmic, id, mv);
+	return res;
+}
+
+static TEE_Result pmic_list_voltages(const struct regul_desc *desc,
+				     uint16_t **levels, size_t *count)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+
+	FMSG("%s: list volt", desc->node_name);
+
+	return pca9450_regulator_levels_mv(pmic, id, (const uint16_t **)levels,
+					   count);
+}
+
+/*
+ * Set flag from regulator framework.
+ * Only handle regultor standard flags
+ */
+static TEE_Result pmic_set_flag(const struct regul_desc *desc, uint16_t flag)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+	TEE_Result res = TEE_SUCCESS;
+
+	FMSG("%s: set_flag 0x%04x", desc->node_name, flag);
+
+	switch (flag) {
+	case REGUL_PULL_DOWN:
+		res = pca9450_regulator_set_prop(pmic, id,
+						 PCA9450_PULL_DOWN, 0);
+		break;
+	default:
+		EMSG("Invalid flag 0x%04x", flag);
+		panic();
+	}
+
+	return res;
+}
+
+static void driver_lock(const struct regul_desc *desc __unused)
+{
+	if (thread_get_id_may_fail() != THREAD_ID_INVALID)
+		mutex_lock(&pmic_mu);
+}
+
+static void driver_unlock(const struct regul_desc *desc __unused)
+{
+	if (thread_get_id_may_fail() != THREAD_ID_INVALID)
+		mutex_unlock(&pmic_mu);
+}
+
+static TEE_Result __maybe_unused
+regu_prepare_suspend(const struct regul_desc *desc, uint8_t mode)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+	uint8_t state = regu->lp_state[mode];
+	uint16_t mv = regu->lp_mv[mode];
+	TEE_Result res;
+
+	IMSG("%s: suspend state:%#08x %u mv", desc->node_name,
+	     state, mv);
+
+	if (state & PCA9450_LP_STATE_UNMODIFIED)
+		return TEE_SUCCESS;
+
+	if (state & PCA9450_LP_STATE_OFF) {
+		res = pca9450_lp_set_state(pmic, id, false);
+		if (res)
+			return res;
+	}
+
+	if (state & PCA9450_LP_STATE_ON) {
+		res = pca9450_lp_set_state(pmic, id, true);
+		if (res)
+			return res;
+
+		if (state & PCA9450_LP_STATE_SET_VOLT) {
+			res = pca9450_lp_set_voltage(pmic, id, mv);
+			if (res)
+				return res;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static const struct regul_ops pmic_ops = {
+	.set_state = pmic_set_state,
+	.get_state = pmic_get_state,
+	.set_voltage = pmic_set_voltage,
+	.get_voltage = pmic_get_voltage,
+	.list_voltages = pmic_list_voltages,
+	.set_flag = pmic_set_flag,
+	.lock = driver_lock,
+	.unlock = driver_unlock,
+};
+
+#define DEFINE_REGU(name) { \
+	.node_name = name, \
+	.driver_data = NULL, \
+	.ops = &pmic_ops, \
+	.ramp_delay_uv_per_us = 2200, \
+	.enable_ramp_delay_us = 1000, \
+}
+
+static const struct regul_desc pmic_reguls[] = {
+	[PCA9450_BUCK1] = DEFINE_REGU("buck1"),
+	[PCA9450_BUCK2] = DEFINE_REGU("buck2"),
+	[PCA9450_BUCK3] = DEFINE_REGU("buck3"),
+	[PCA9450_BUCK4] = DEFINE_REGU("buck4"),
+	[PCA9450_BUCK5] = DEFINE_REGU("buck5"),
+	[PCA9450_BUCK6] = DEFINE_REGU("buck6"),
+	[PCA9450_LDO1] = DEFINE_REGU("ldo1"),
+	[PCA9450_LDO2] = DEFINE_REGU("ldo2"),
+	[PCA9450_LDO3] = DEFINE_REGU("ldo3"),
+	[PCA9450_LDO4] = DEFINE_REGU("ldo4"),
+	[PCA9450_LDO5] = DEFINE_REGU("ldo5"),
+};
+DECLARE_KEEP_PAGER(pmic_reguls);
+
+struct regul_property {
+	const char *name;
+	uint16_t prop;
+};
+
+static const struct regul_property prop_table[] = {
+	{
+		.name = "wdog-b-reset",
+		.prop = PCA9450_WDOG_B_RESET,
+	},
+	{
+		.name = "pmic-rst-b",
+		.prop = PCA9450_PMIC_RST_B,
+	},
+};
+
+static TEE_Result __unused pca9450_set_prop(const struct regul_desc *desc,
+					    enum pca9450_prop_id prop, uint16_t value)
+{
+	const struct regu_handle_s *regu = desc->driver_data;
+	struct pmic_handle_s *pmic = regu->pmic;
+	uint32_t id = regu->id;
+
+	assert(desc);
+
+	FMSG("%s: set_prop 0x%x val=%u\n", desc->node_name, prop, value);
+
+	return pca9450_regulator_set_prop(pmic, id, prop, value);
+}
+
+static TEE_Result parse_properties(const void *fdt, int node,
+				   struct pmic_handle_s *pmic)
+
+{
+	const struct regul_property *p;
+
+	for (p = prop_table; p < (prop_table + ARRAY_SIZE(prop_table)); p++) {
+		const char *val;
+		int len;
+		int arg = -1;
+		TEE_Result res;
+
+		switch (p->prop) {
+		case PCA9450_WDOG_B_RESET:
+			// fallthrough
+		case PCA9450_PMIC_RST_B:
+			val = fdt_get_string(fdt, node, &len);
+
+			if (!val)
+				arg = 0;
+			else if (strcmp(val, "all") == 0)
+				arg = 3;
+			else if (strcmp(val, "cold") == 0)
+				arg = 2;
+			else if (strcmp(val, "warm") == 0)
+				arg = 1;
+			else if (strcmp(val, "disabled") == 0)
+				arg = 0;
+			else
+				EMSG("%s: Invalid %s property value: '%s'",
+				     fdt_get_name(fdt, node, NULL), p->name, val);
+			if (arg == -1)
+				continue;
+			if (p->prop == PCA9450_WDOG_B_RESET)
+				res = pca9450_register_update(pmic, RESET_CTRL,
+							       WDOG_B_CFG_MSK,
+							       arg << WDOG_B_CFG_SHIFT);
+			else
+				res = pca9450_register_update(pmic, RESET_CTRL,
+							       PMIC_RST_CFG_MSK,
+							       arg << PMIC_RST_CFG_SHIFT);
+			if (res)
+				return res;
+			break;
+
+		default:
+			FMSG("%s: Ignoring property '%s'",
+			     fdt_get_name(fdt, node, NULL), p->name);
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result parse_low_power_mode(const void *fdt,
+				 const struct regul_desc *desc,
+				 int node, int mode)
+{
+	struct regu_handle_s *regu = (void *)desc->driver_data;
+	const fdt32_t *cuint;
+
+	regu->lp_state[mode] = 0;
+
+	if (fdt_getprop(fdt, node, "regulator-off-in-suspend", NULL)) {
+		FMSG("%s: mode:%d OFF", desc->node_name, mode);
+		regu->lp_state[mode] |= PCA9450_LP_STATE_OFF;
+	} else if (fdt_getprop(fdt, node, "regulator-on-in-suspend", NULL)) {
+		FMSG("%s: mode:%d ON", desc->node_name, mode);
+		regu->lp_state[mode] |= PCA9450_LP_STATE_ON;
+	} else {
+		regu->lp_state[mode] |= PCA9450_LP_STATE_UNMODIFIED;
+	}
+
+	cuint = fdt_getprop(fdt, node, "regulator-suspend-microvolt", NULL);
+	if (cuint) {
+		uint16_t mv = fdt32_to_cpu(*cuint) / 1000;
+		TEE_Result res;
+
+		IMSG("%s: mode:%d suspend mv=%u", desc->node_name,
+		     mode, mv);
+
+		res = pmic_verify_voltage(desc, mv);
+		if (res != TEE_SUCCESS) {
+			EMSG("Invalid suspend voltage %u", mv);
+			return res;
+		}
+		regu->lp_state[mode] |= PCA9450_LP_STATE_SET_VOLT;
+		regu->lp_mv[mode] = mv;
+	}
+	return TEE_SUCCESS;
+}
+
+static TEE_Result parse_low_power_modes(const void *fdt,
+					const struct regul_desc *desc, int node)
+{
+	int mode;
+
+	for (mode = 0; mode < STM32_PM_NB_SOC_MODES; mode++) {
+		const char *lp_mode_name = plat_get_lp_mode_name(mode);
+
+		if (lp_mode_name) {
+			int n;
+			TEE_Result res;
+
+			/* Get the configs from regulator_state_node subnode */
+			n = fdt_subnode_offset(fdt, node, lp_mode_name);
+			if (n >= 0) {
+				res = parse_low_power_mode(fdt, desc, n, mode);
+				if (res)
+					return res;
+			}
+		}
+	}
+	return TEE_SUCCESS;
+}
+
+#ifdef CFG_PM
+static TEE_Result pca9450_regu_pm(enum pm_op op, uint32_t pm_hint,
+				  const struct pm_callback_handle *pm_handle)
+{
+	struct regul_desc *desc = (struct regul_desc *)pm_handle->handle;
+	TEE_Result res;
+	unsigned int pwrlvl = PM_HINT_PLATFORM_STATE(pm_hint);
+	uint8_t mode;
+
+	if (op == PM_OP_SUSPEND) {
+		/* configure PMIC level according MAX PM domain OFF */
+		switch (pwrlvl) {
+		case PM_D1_LEVEL:
+		case PM_D2_LEVEL:
+			mode = STM32_PM_LPLV;
+			break;
+		case PM_D2_LPLV_LEVEL:
+			mode = STM32_PM_STANDBY;
+			break;
+		case PM_MAX_LEVEL:
+			mode = STM32_PM_OFF;
+			break;
+		default:
+			mode = STM32_PM_DEFAULT;
+			break;
+		}
+	} else if (op == PM_OP_RESUME) {
+		mode = STM32_PM_DEFAULT;
+	} else {
+		EMSG("INVALID PM OP 0x%08x", op);
+		panic();
+	}
+
+	res = regu_prepare_suspend(desc, mode);
+	return res;
+}
+#endif
+
+static TEE_Result register_pmic_regulator(const void *fdt,
+					  struct pmic_handle_s *pmic,
+					  const char *regu_name, int node)
+{
+	TEE_Result res;
+	struct regu_handle_s *regu;
+	struct regul_desc *desc;
+	size_t id;
+
+	IMSG("PCA9450 register %s", regu_name);
+
+	for (id = 0; id < ARRAY_SIZE(pmic_reguls); id++)
+		if (!strcmp(pmic_reguls[id].node_name, regu_name))
+			break;
+
+	assert(id < ARRAY_SIZE(pmic_reguls));
+
+	regu = calloc(1, sizeof(*regu));
+	if (!regu)
+		panic();
+
+	regu->pmic = pmic;
+	regu->id = id;
+	desc = &regu->desc;
+
+	memcpy(desc, pmic_reguls + id, sizeof(*desc));
+	desc->driver_data = regu;
+
+	res = regulator_register(desc, node);
+	if (res) {
+		EMSG("Failed to register %s", regu_name);
+		goto out;
+	}
+
+	res = parse_low_power_modes(fdt, desc, node);
+	if (res) {
+		EMSG("Failed to parse low power modes for %s", regu_name);
+		goto out;
+	}
+
+	res = regu_prepare_suspend(desc, STM32_PM_DEFAULT);
+	if (res) {
+		EMSG("Failed to prepare suspend mode for %s", regu_name);
+		goto out;
+	}
+
+#ifdef CFG_PM
+	register_pm_core_service_cb(pca9450_regu_pm, desc, desc->node_name);
+#endif
+
+	return TEE_SUCCESS;
+
+out:
+	free(regu);
+	return res;}
+
+static TEE_Result parse_regulator_fdt_nodes(const void *fdt, int node,
+					    struct pmic_handle_s *pmic)
+{
+	int regulators_node;
+	int regu_node;
+
+	regulators_node = fdt_subnode_offset(fdt, node, "regulators");
+	if (regulators_node < 0)
+		panic();
+
+	fdt_for_each_subnode(regu_node, fdt, regulators_node) {
+		TEE_Result res = TEE_ERROR_GENERIC;
+		int status = _fdt_get_status(fdt, regu_node);
+		const char *regu_name;
+
+		assert(status >= 0);
+		if (status == DT_STATUS_DISABLED)
+			continue;
+
+		regu_name = fdt_get_name(fdt, regu_node, NULL);
+		assert(regu_name);
+
+		res = register_pmic_regulator(fdt, pmic, regu_name, regu_node);
+		if (res) {
+			EMSG("Failed to register %s", regu_name);
+			return res;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+#ifdef CFG_STM32_PWR_IRQ
+
+enum itr_return pca9450_irq_callback(struct pmic_handle_s *pmic, uint8_t it_id)
+{
+	struct pmic_it_handle_s *prv = NULL;
+
+	FMSG("PCA9450 it id %d", (int)it_id);
+
+	SLIST_FOREACH(prv, &pmic->it_list, link)
+		if (prv->pmic_it == it_id) {
+			FMSG("PCA9450 send notif %u", prv->notif_id);
+
+			notif_send_it(prv->notif_id);
+
+			return ITRR_HANDLED;
+		}
+
+	return ITRR_NONE;
+}
+
+static enum itr_return pca9450_irq_handler(struct itr_handler *handler)
+{
+	struct pmic_handle_s *pmic = handler->data;
+
+	FMSG("PCA9450 irq");
+
+	return pca9450_handle_irq(pmic);
+}
+
+static TEE_Result initialize_pca9450_irq(const void *fdt, int node,
+					 struct pmic_handle_s *pmic)
+{
+	TEE_Result res;
+	struct itr_handler *hdl;
+	const fdt32_t *cuint;
+	const uint32_t *notif_ids;
+	int nb_notif;
+
+	IMSG("Init pca9450 irq");
+
+	SLIST_INIT(&pmic->it_list);
+
+	cuint = fdt_getprop(fdt, node, "st,wakeup-pin-number", NULL);
+	if (cuint) {
+		size_t it = fdt32_to_cpu(*cuint) - 1;
+
+		res = stm32mp25_pwr_itr_alloc_add(it, pca9450_irq_handler,
+						  PWR_WKUP_FLAG_FALLING |
+						  PWR_WKUP_FLAG_THREADED,
+						  pmic, &hdl);
+		if (res)
+			return res;
+
+		stm32mp25_pwr_itr_enable(hdl->it);
+	}
+
+	notif_ids = fdt_getprop(fdt, node, "st,notif-it-id", &nb_notif);
+	if (nb_notif > 0) {
+		struct pmic_it_handle_s *prv;
+		unsigned int i;
+		const uint32_t *pmic_its;
+		int nb_it;
+
+		pmic_its = fdt_getprop(fdt, node, "st,pmic-it-id", &nb_it);
+		if (!pmic_its)
+			return TEE_ERROR_ITEM_NOT_FOUND;
+
+		if (nb_it != nb_notif)
+			panic("st,notif-it-id incorrect description");
+
+		for (i = 0; i < (nb_notif / sizeof(uint32_t)); i++) {
+			uint8_t pmic_it;
+
+			prv = calloc(1, sizeof(*prv));
+			if (!prv)
+				panic("pmic: Could not allocate pmic it");
+
+			pmic_it = fdt32_to_cpu(pmic_its[i]);
+
+			prv->pmic_it = pmic_it;
+			prv->notif_id = fdt32_to_cpu(notif_ids[i]);
+
+			SLIST_INSERT_HEAD(&pmic->it_list, prv, link);
+
+			/* Enable requested interrupt */
+			if (pca9450_set_irq_mask(pmic, pmic_it, false))
+				return TEE_ERROR_GENERIC;
+
+			IMSG("PCA9450 forwards pmic_it:%u as notif:%u",
+			     prv->pmic_it, prv->notif_id);
+		}
+	}
+
+	/* Unmask all over-current interrupts */
+	res = pca9450_register_write(pmic, INT1_MSK, 0x00);
+
+	return res;
+}
+#endif
+
+static void initialize_pca9450_i2c(const void *fdt, int node,
+				   struct pmic_handle_s *pmic,
+				   struct i2c_handle_s *i2c)
+{
+	const fdt32_t *cuint = NULL;
+	uint32_t pmic_i2c_addr;
+
+	cuint = fdt_getprop(fdt, node, "reg", NULL);
+	if (!cuint)
+		panic("PMIC configuration failed on reg property");
+
+	pmic_i2c_addr = fdt32_to_cpu(*cuint) << 1;
+	if (pmic_i2c_addr > UINT16_MAX)
+		panic("PMIC configuration failed on i2c address translation");
+
+	if (!stm32_i2c_is_device_ready(i2c, pmic_i2c_addr,
+				       PMIC_I2C_TRIALS,
+				       PMIC_I2C_TIMEOUT_BUSY_MS)) {
+		panic("PCA9450 I2C init failed");
+	}
+
+	pmic->pmic_i2c_handle = i2c;
+	pmic->pmic_i2c_addr = pmic_i2c_addr;
+}
+
+/*
+ * PMIC and resource initialization
+ */
+static void initialize_pca9450(const void *fdt, int node,
+			       struct pmic_handle_s *pmic,
+			       struct i2c_handle_s *i2c)
+{
+	uint8_t chip_id;
+	const char *variant;
+
+	FMSG("Initialize pca9450");
+
+	initialize_pca9450_i2c(fdt, node, pmic, i2c);
+
+	if (pca9450_get_chip_id(pmic, &chip_id))
+		panic("Failed to access PMIC");
+
+	switch ((chip_id & CHIP_ID_MASK) >> CHIP_ID_SHIFT) {
+	case CHIP_ID_PCA9450A:
+		variant = "A";
+		break;
+	case CHIP_ID_PCA9450BC:
+		variant = "B/C";
+		break;
+	default:
+		EMSG("Unsupported PCA9450 chip variant 0x%02x", chip_id);
+		panic();
+	}
+	IMSG("PMIC PCA9450%s 0x%02x", variant, chip_id);
+
+	pca9450_init();
+
+	pca9450_register_write(pmic, BUCK123_DVS, 0x00);
+
+	pca9450_dump_regulators(pmic);
+}
+
+static TEE_Result pmic_pca9450_probe(const void *fdt, int node,
+				     const void *compat_data __unused)
+{
+	struct i2c_handle_s *i2c = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct pmic_handle_s *pmic = NULL;
+
+	FMSG("Probe pca9450");
+
+	res = i2c_dt_get_by_subnode(fdt, node, &i2c);
+	if (res)
+		return res;
+
+	pmic = calloc(1, sizeof(*pmic));
+	if (!pmic)
+		panic();
+
+	initialize_pca9450(fdt, node, pmic, i2c);
+
+#ifdef CFG_STM32_PWR_IRQ
+	res = initialize_pca9450_irq(fdt, node, pmic);
+	if (res) {
+		free(pmic);
+		return res;
+	}
+#endif
+
+	res = parse_properties(fdt, node, pmic);
+	if (res) {
+		EMSG("Failed to parse properties");
+		free(pmic);
+		return res;
+	}
+
+	res = parse_regulator_fdt_nodes(fdt, node, pmic);
+	if (res)
+		panic();
+
+	return TEE_SUCCESS;
+}
+
+static const struct dt_device_match pmic_pca9450_match_table[] = {
+	{ .compatible = "nxp,pca9450" },
+	{ }
+};
+
+DEFINE_DT_DRIVER(pmic_pca9450_dt_driver) = {
+	.name = "pmic_pca9450",
+	.match_table = pmic_pca9450_match_table,
+	.probe = pmic_pca9450_probe,
+};
Index: git/core/drivers/sub.mk
===================================================================
--- git.orig/core/drivers/sub.mk
+++ git/core/drivers/sub.mk
@@ -72,6 +72,7 @@ srcs-$(CFG_VERSAL_GPIO) += versal_gpio.c
 srcs-$(CFG_VERSAL_MBOX) += versal_mbox.c
 srcs-$(CFG_VERSAL_PM) += versal_pm.c
 srcs-$(CFG_STM32MP15_HUK) += stm32mp15_huk.c
+srcs-$(CFG_PMIC_PCA9450) += pmic-pca9450.c pca9450.c
 
 subdirs-y += counter
 subdirs-y += crypto
Index: git/core/include/dt-bindings/regulator/nxp,pca9450-regulator.h
===================================================================
--- /dev/null
+++ git/core/include/dt-bindings/regulator/nxp,pca9450-regulator.h
@@ -0,0 +1,43 @@
+/* SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause) */
+/*
+ * Copyright (C) 2023, STMicroelectronics - All Rights Reserved
+ */
+
+#ifndef __DT_BINDINGS_REGULATOR_PCA9450_REGULATOR_H
+#define __DT_BINDINGS_REGULATOR_PCA9450_REGULATOR_H
+
+/* SCMI voltage domains identifiers */
+
+/* SOC Internal regulators */
+#define VOLTD_SCMI_VDDIO1		0
+#define VOLTD_SCMI_VDDIO2		1
+#define VOLTD_SCMI_VDDIO3		2
+#define VOLTD_SCMI_VDDIO4		3
+#define VOLTD_SCMI_VDDIO		4
+#define VOLTD_SCMI_UCPD			5
+#define VOLTD_SCMI_USB33		6
+#define VOLTD_SCMI_ADC			7
+#define VOLTD_SCMI_GPU			8
+#define VOLTD_SCMI_VREFBUF		9
+
+/* PCA9450 regulators */
+#define VOLTD_SCMI_PCA9450_BUCK1	10
+#define VOLTD_SCMI_PCA9450_BUCK2	11
+#define VOLTD_SCMI_PCA9450_BUCK3	12
+#define VOLTD_SCMI_PCA9450_BUCK4	13
+#define VOLTD_SCMI_PCA9450_BUCK5	14
+#define VOLTD_SCMI_PCA9450_BUCK6	15
+#define VOLTD_SCMI_PCA9450_LDO1		17
+#define VOLTD_SCMI_PCA9450_LDO2		18
+#define VOLTD_SCMI_PCA9450_LDO3		19
+#define VOLTD_SCMI_PCA9450_LDO4		20
+#define VOLTD_SCMI_PCA9450_LDO5		21
+
+/* External regulators */
+#define VOLTD_SCMI_REGU0		26
+#define VOLTD_SCMI_REGU1		27
+#define VOLTD_SCMI_REGU2		28
+#define VOLTD_SCMI_REGU3		29
+#define VOLTD_SCMI_REGU4		30
+
+#endif /*__DT_BINDINGS_REGULATOR_PCA9450_REGULATOR_H */
