diff --git a/arch/arm/config.mk b/arch/arm/config.mk
index db54b6dcea..fc92247bf1 100644
--- a/arch/arm/config.mk
+++ b/arch/arm/config.mk
@@ -16,7 +16,7 @@ endif
 LDFLAGS_FINAL += --gc-sections
 PLATFORM_RELFLAGS += -ffunction-sections -fdata-sections \
 		     -fno-common -ffixed-r9
-PLATFORM_RELFLAGS += $(call cc-option, -msoft-float) \
+PLATFORM_RELFLAGS += $(call cc-option, -mfloat-abi=hard -mfpu=vfpv4) \
       $(call cc-option,-mshort-load-bytes,$(call cc-option,-malignment-traps,))
 
 PLATFORM_CPPFLAGS += -D__ARM__
