Index: u-boot-imx8mm-5.4.24.2.1.0/drivers/usb/gadget/f_mass_storage.c
===================================================================
--- u-boot-imx8mm-5.4.24.2.1.0.orig/drivers/usb/gadget/f_mass_storage.c
+++ u-boot-imx8mm-5.4.24.2.1.0/drivers/usb/gadget/f_mass_storage.c
@@ -244,6 +244,7 @@
 #include <malloc.h>
 #include <common.h>
 #include <g_dnl.h>
+#include <watchdog.h>
 
 #include <linux/err.h>
 #include <linux/usb/ch9.h>
@@ -682,6 +683,8 @@ static int sleep_thread(struct fsg_commo
 			if (ctrlc())
 				return -EPIPE;
 
+			WATCHDOG_RESET();
+
 			/* Check cable connection */
 			if (!g_dnl_board_usb_cable_connected())
 				return -EIO;
