Index: git/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c
===================================================================
--- git.orig/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c
+++ git/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c
@@ -621,6 +621,10 @@ static void rzg2l_mipi_dsi_enable(struct
 
 	rzg2l_mipi_dsi_set_display_timing(mipi_dsi);
 
+	ret = rzg2l_mipi_dsi_start_hs_clock(mipi_dsi);
+	if (ret < 0)
+		return;
+
 	if (mipi_dsi->panel) {
 		/*
 		 * Workaround
@@ -650,9 +654,6 @@ static void rzg2l_mipi_dsi_enable(struct
 		}
 	}
 
-	ret = rzg2l_mipi_dsi_start_hs_clock(mipi_dsi);
-	if (ret < 0)
-		return;
 start_video:
 	ret = rzg2l_mipi_dsi_start_video(mipi_dsi);
 	if (ret < 0)
