diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index e229de55030b..134868728605 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -292,6 +292,7 @@ struct cpg_param resolution_param_parallel[TABLE_PARALLEL_MAX] = {
 	{ 29605, 1,  88, 13673431, 6, 6, 0, 0x16, 1, 0}, /* WVGA 29.605MHz */
 	{ 40000, 1,  70,        0, 7, 3, 0, 0x16, 1, 0}, /* SVGA 40.00MHz */
 	{ 65000, 1,  81,  4194304, 5, 3, 0, 0x16, 1, 0}, /* XGA 65.00MHz */
+	{ 68200, 1,  68,  3355443, 1, 1, 0, 0x16, 1, 0}, /* 68.20MHz */
 	{ 71000, 1,  71,        0, 6, 2, 0, 0x16, 1, 0}, /* WXGA 1280x800 71.0MHz */
 	{ 74176, 1,  86,  9037327, 7, 2, 0, 0x16, 1, 0}, /* 720p 74.176MHz */
 	{ 74250, 1,  86, 10485760, 7, 2, 0, 0x16, 1, 0}, /* 720p 74.25MHz */
