diff --git a/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c b/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c
index b5a4a13e2af9..db30acc99573 100644
--- a/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c
+++ b/drivers/gpu/drm/rcar-du/rzg2l_mipi_dsi.c
@@ -738,8 +738,8 @@ static int rzg2l_mipi_dsi_host_attach(struct mipi_dsi_host *host,
 {
 	struct rzg2l_mipi_dsi *mipi_dsi = host_to_rzg2l_mipi_dsi(host);
 
-	if (device->lanes > mipi_dsi->num_data_lanes)
-		return -EINVAL;
+	// if (device->lanes > mipi_dsi->num_data_lanes)
+	// 	return -EINVAL;
 
 	mipi_dsi->lanes = device->lanes;
 	mipi_dsi->format = device->format;
