Index: linux-txmp-2550/drivers/gpu/drm/stm/ltdc.c
===================================================================
--- linux-txmp-2550.orig/drivers/gpu/drm/stm/ltdc.c
+++ linux-txmp-2550/drivers/gpu/drm/stm/ltdc.c
@@ -865,7 +865,7 @@ static void ltdc_crtc_atomic_enable(stru
 	if (bus_flags & DRM_BUS_FLAG_DE_LOW)
 		val |= GCR_DEPOL;
 
-	if (bus_flags & DRM_BUS_FLAG_PIXDATA_DRIVE_NEGEDGE)
+	if (bus_flags & DRM_BUS_FLAG_PIXDATA_DRIVE_POSEDGE)
 		val |= GCR_PCPOL;
 
 	if (connector && connector->state->dithering == DRM_MODE_DITHERING_ON)
