Index: linux-qsmp-2550/drivers/gpu/drm/stm/lvds.c
===================================================================
--- linux-qsmp-2550.orig/drivers/gpu/drm/stm/lvds.c
+++ linux-qsmp-2550/drivers/gpu/drm/stm/lvds.c
@@ -348,6 +348,25 @@ const enum lvds_pixel lvds_bitmap_jeida_
 };
 
 /*
+ * Expected JEIDA-RGB666 data to be sent in LSB format
+ * CHAN0   {ONE, ONE, ZERO, ZERO, ZERO, ONE, ONE}
+ * CHAN1   {G0,  R5,  R4,   R3,   R2,   R1,  R0}
+ * CHAN2   {B1,  B0,  G5,   G4,   G3,   G2,  G1}
+ * CHAN3   {DE,  VS,  HS,   B5,   B4,   B3,  B2}
+ */
+const enum lvds_pixel lvds_bitmap_jeida_rgb666[4][7] = {
+	{
+		PIX_ONE, PIX_ONE, PIX_ZER, PIX_ZER, PIX_ZER, PIX_ONE, PIX_ONE
+	}, {
+		PIX_G_0, PIX_R_5, PIX_R_4, PIX_R_3, PIX_R_2, PIX_R_1, PIX_R_0
+	}, {
+		PIX_B_1, PIX_B_0, PIX_G_5, PIX_G_4, PIX_G_3, PIX_G_2, PIX_G_1
+	}, {
+		PIX_D_E, PIX_V_S, PIX_H_S, PIX_B_5, PIX_B_4, PIX_B_3, PIX_B_2
+	}
+};
+
+/*
  * Expected VESA-RGB888 data to be sent in LSB format
  *	    bit6 ............................bit0
  * CHAN0   {ONE, ONE, ZERO, ZERO, ZERO, ONE, ONE}
@@ -766,8 +785,8 @@ static int lvds_pixel_clk_register(struc
 static void lvds_config_data_mapping(struct stm_lvds *lvds)
 {
 	const struct drm_display_info *info;
-	u32 lvds_dmlcr[5], lvds_dmmcr[5]; /* 4 data lanes + 1 clk lane */
-	int i;
+	size_t i, num_lanes;
+	const enum lvds_pixel (*bitmap)[7];
 
 	info = &(&lvds->connector)->display_info;
 	if (!info->num_bus_formats || !info->bus_formats) {
@@ -780,48 +799,33 @@ static void lvds_config_data_mapping(str
  */
 
 	switch (info->bus_formats[0]) {
-	case MEDIA_BUS_FMT_RGB666_1X7X3_SPWG: /* VESA-RGB666 */
-		DRM_WARN("Pixel format with data mapping not yet compatible.\n");
+	case MEDIA_BUS_FMT_RGB666_1X7X3_SPWG: /* JEIDA-RGB666 */
+		bitmap = &lvds_bitmap_jeida_rgb666[0];
+		num_lanes = ARRAY_SIZE(lvds_bitmap_jeida_rgb666);
 		break;
 	case MEDIA_BUS_FMT_RGB888_1X7X4_SPWG: /* VESA-RGB888 */
-		for (i = 0; i < 5; i++) {
-			lvds_dmlcr[i] = ((lvds_bitmap_vesa_rgb888[i][0])
-					 + (lvds_bitmap_vesa_rgb888[i][1] << 5)
-					 + (lvds_bitmap_vesa_rgb888[i][2] << 10)
-					 + (lvds_bitmap_vesa_rgb888[i][3] << 15));
-			lvds_dmmcr[i] = ((lvds_bitmap_vesa_rgb888[i][4])
-					 + (lvds_bitmap_vesa_rgb888[i][5] << 5)
-					 + (lvds_bitmap_vesa_rgb888[i][6] << 10));
-		}
+		bitmap = &lvds_bitmap_vesa_rgb888[0];
+		num_lanes = ARRAY_SIZE(lvds_bitmap_vesa_rgb888);
 		break;
 	case MEDIA_BUS_FMT_RGB888_1X7X4_JEIDA: /* JEIDA-RGB888 */
-		for (i = 0; i < 5; i++) {
-			lvds_dmlcr[i] = ((lvds_bitmap_jeida_rgb888[i][0])
-					+ (lvds_bitmap_jeida_rgb888[i][1] << 5)
-					+ (lvds_bitmap_jeida_rgb888[i][2] << 10)
-					+ (lvds_bitmap_jeida_rgb888[i][3] << 15));
-			lvds_dmmcr[i] = ((lvds_bitmap_jeida_rgb888[i][4])
-					+ (lvds_bitmap_jeida_rgb888[i][5] << 5)
-					+ (lvds_bitmap_jeida_rgb888[i][6] << 10));
-		}
+		bitmap = &lvds_bitmap_jeida_rgb888[0];
+		num_lanes = ARRAY_SIZE(lvds_bitmap_jeida_rgb888);
 		break;
 	default:
 		dev_warn(lvds->dev, "Unsupported LVDS bus format 0x%04x\n",
 			 info->bus_formats[0]);
+		return;
 	}
 
-	/* Write registers at the end of computations */
-	/* TODO use for loop */
-	lvds_write(lvds, LVDS_DMLCR0, lvds_dmlcr[0]);
-	lvds_write(lvds, LVDS_DMMCR0, lvds_dmmcr[0]);
-	lvds_write(lvds, LVDS_DMLCR1, lvds_dmlcr[1]);
-	lvds_write(lvds, LVDS_DMMCR1, lvds_dmmcr[1]);
-	lvds_write(lvds, LVDS_DMLCR2, lvds_dmlcr[2]);
-	lvds_write(lvds, LVDS_DMMCR2, lvds_dmmcr[2]);
-	lvds_write(lvds, LVDS_DMLCR3, lvds_dmlcr[3]);
-	lvds_write(lvds, LVDS_DMMCR3, lvds_dmmcr[3]);
-	lvds_write(lvds, LVDS_DMLCR4, lvds_dmlcr[4]);
-	lvds_write(lvds, LVDS_DMMCR4, lvds_dmmcr[4]);
+	for (i = 0; i < num_lanes; i++) {
+		u32 lvds_dmlcr = bitmap[i][0] + (bitmap[i][1] << 5) +
+			(bitmap[i][2] << 10) + (bitmap[i][3] << 15);
+		u32 lvds_dmmcr = bitmap[i][4] + (bitmap[i][5] << 5) +
+			(bitmap[i][6] << 10);
+
+		lvds_write(lvds, LVDS_DMLCR0 + i * 8, lvds_dmlcr);
+		lvds_write(lvds, LVDS_DMMCR0 + i * 8, lvds_dmmcr);
+	}
 }
 
 static int lvds_config_mode(struct stm_lvds *lvds)
