Index: karo-tx-linux/arch/arm/boot/dts/stm32mp151.dtsi
===================================================================
--- karo-tx-linux.orig/arch/arm/boot/dts/stm32mp151.dtsi
+++ karo-tx-linux/arch/arm/boot/dts/stm32mp151.dtsi
@@ -1370,7 +1370,7 @@
 			compatible = "st,stm32mp1-dwmac", "snps,dwmac-4.20a";
 			reg = <0x5800a000 0x2000>;
 			reg-names = "stmmaceth";
-			interrupts-extended = <&intc GIC_SPI 61 IRQ_TYPE_LEVEL_HIGH>;
+			interrupts = <GIC_SPI 61 IRQ_TYPE_LEVEL_HIGH>;
 			interrupt-names = "macirq";
 			clock-names = "stmmaceth",
 				      "mac-clk-tx",
